{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "75172cb8-5c22-40d3-9fa7-e3087373ff53",
   "metadata": {},
   "source": [
    "# Note6_R 繪圖出潛在降解生物的環境分布"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f05710ad-0d19-4343-b5ca-34c49e21be2e",
   "metadata": {},
   "source": [
    "# loose 標準的 facet stacked bar char"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "37a53715-edef-4e1b-8631-b725aac7dd9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ggplot2)\n",
    "library(readr)\n",
    "library(gridExtra)\n",
    "library(patchwork)\n",
    "library(grid)\n",
    "library(cowplot)\n",
    "library(stringr)\n",
    "library(forcats)\n",
    "\n",
    "\n",
    "# convert the table to a data frame\n",
    "HMM_Final <- read.csv(\"B50_HMMFinal_loose_ForBarChart.csv\", header=TRUE, sep = ',')\n",
    "\n",
    "# Check unique value of dataframe\n",
    "unique(HMM_Final$ecosystem)\n",
    "unique(HMM_Final$ecosystem_category)\n",
    "unique(HMM_Final$taxonomy)\n",
    "\n",
    "# Replace all \"Non-marine Saline and Alkaline\" values in column \"ecosystem_category\" with \"Salt Lake\"\n",
    "HMM_Final$ecosystem_category[HMM_Final$ecosystem_category == \"Non-marine Saline and Alkaline\"] <- \"Salt Lake\"\n",
    "\n",
    "# Replace \"Alphaproteobacteria\" with \"α-Proteobacteria\"; \"Gammaproteobacteria\" with 'γ-Proteobacteria'\n",
    "HMM_Final$taxonomy <- gsub(\"Alphaproteobacteria\", \"α-Proteobacteria\", HMM_Final$taxonomy)\n",
    "HMM_Final$taxonomy <- gsub(\"Gammaproteobacteria\", \"γ-Proteobacteria\", HMM_Final$taxonomy)\n",
    "\n",
    "# Change taxonomy order\n",
    "taxonomy_order <- c( \"Unclassified Bacteria\", 'Other Bacterial Phylum','Myxococcia', 'Thermodesulfobacterota',\n",
    "                     \"Chloroflexi\", \"α-Proteobacteria\", \"γ-Proteobacteria\", \"Actinobacteria\")\n",
    "\n",
    "HMM_Final$taxonomy <- fct_relevel(HMM_Final$taxonomy, taxonomy_order)\n",
    "HMM_Final\n",
    "\n",
    "# change ecosystem order\n",
    "ecosystem_order <- c(\"Aquatic\", \"Terrestrial\", \"Host-associated\", \"Engineered\")\n",
    "HMM_Final$ecosystem <- fct_relevel(HMM_Final$ecosystem, ecosystem_order)\n",
    "\n",
    "# create the color vector\n",
    "class_colors <- c(\"α-Proteobacteria\" = \"#2e75b6\",\n",
    "                  \"γ-Proteobacteria\" = \"#9dc3e6\", \n",
    "                  \"Actinobacteria\" = \"#c00000\",\n",
    "                  \"Chloroflexi\" = \"#a9d18e\",\n",
    "                  'Thermodesulfobacterota' = '#ffd966',\n",
    "                  \"Myxococcia\" = \"#7e33b8\",\n",
    "                  \"Other Bacterial Phylum\" = \"#404040\",\n",
    "                  \"Unclassified Bacteria\" = \"#cbcbcb\"\n",
    ")\n",
    "\n",
    "# ggplot No legend\n",
    "p <- ggplot(HMM_Final, aes(x = str_wrap(ecosystem_category, width = 30), y = NumberOfMAGs, fill = taxonomy)) +\n",
    "  geom_bar(stat = \"identity\",\n",
    "           width = ifelse(HMM_Final$ecosystem == \"Terrestrial\", 0.4, 0.8),\n",
    "           position = \"stack\") +\n",
    "  labs(title = \"\",\n",
    "       x =\"\",\n",
    "       y = \"\")+\n",
    "  scale_fill_manual(values = class_colors,\n",
    "                    limits = c(\"Actinobacteria\", \"Chloroflexi\", \"α-Proteobacteria\", \"γ-Proteobacteria\",\n",
    "                              'Myxococcia', 'Thermodesulfobacterota', 'Other Bacterial Phylum', \"Unclassified Bacteria\"))+\n",
    "  theme_bw()+\n",
    "  theme(plot.title = element_text(size = 14, face = \"bold\", hjust = 0.5),\n",
    "        axis.title.y = element_text(size = 11, margin = margin(r = 10), face = \"bold\"),\n",
    "        strip.text = element_text(size = 11, face = \"bold\"),\n",
    "        axis.text.x = element_text(size = 10, angle = 45, hjust = 1, color =\"black\"),\n",
    "        axis.text.y = element_text(size = 10),\n",
    "        legend.text = element_text(size = 10),\n",
    "        legend.position = \"none\",\n",
    "        legend.title = element_blank(),\n",
    "        panel.grid.minor.y = element_blank(),\n",
    "        panel.grid.major.x = element_blank(),\n",
    "        plot.margin = unit(c(-0.5, 0.1, -0.5, 0.1), \"cm\"),\n",
    "        text = element_text(family = \"Arial\"))+\n",
    "  scale_y_continuous(expand = c(0.03, 0), breaks = seq(0, 300, 50)) +\n",
    "  facet_wrap(~ecosystem, scales = \"free_x\", nrow = 1)+\n",
    "  guides(fill = guide_legend(ncol = 4))\n",
    "p\n",
    "\n",
    "# save plot with ggsave\n",
    "# unit = 'cm': 高圖參數: (width = 18.18, height = 26.67, dpi = 300) / 矮圖參數: (width = 18.18, height = 7, dpi = 300)\n",
    "ggsave(\"MAGsHits_StackedBarChart_loose.tiff\", p, width = 18.18, height = 7, dpi = 300, units = \"cm\")\n",
    "\n",
    "\n",
    "#-----------------------\n",
    "# ggplot with legend\n",
    "pl <- ggplot(HMM_Final, aes(x = str_wrap(ecosystem_category, width = 30), y = NumberOfMAGs, fill = taxonomy)) +\n",
    "  geom_bar(stat = \"identity\",\n",
    "           width = ifelse(HMM_Final$ecosystem == \"Terrestrial\", 0.4, 0.8),\n",
    "           position = \"stack\") +\n",
    "  labs(title = \"\",\n",
    "       x =\"\",\n",
    "       y = \"Number of estrogen-degrading MAGs\")+\n",
    "  scale_fill_manual(values = class_colors,\n",
    "                    limits = c(\"Actinobacteria\", \"Chloroflexi\", \"α-Proteobacteria\", \"γ-Proteobacteria\",\n",
    "                               'Myxococcia', 'Thermodesulfobacterota', 'Other Bacterial Phylum', \"Unclassified Bacteria\"))+\n",
    "  theme_bw()+\n",
    "  theme(plot.title = element_text(size = 14, face = \"bold\", hjust = 0.5),\n",
    "        axis.title.y = element_text(size = 12, margin = margin(r = 10), face = \"bold\"),\n",
    "        strip.text = element_text(size = 12, face = \"bold\"),\n",
    "        axis.text.x = element_text(size = 10, angle = 45, hjust = 1, color =\"black\"),\n",
    "        axis.text.y = element_text(size = 10),\n",
    "        legend.text = element_text(size = 10),\n",
    "        legend.position = \"bottom\",\n",
    "        legend.title = element_blank(),\n",
    "        panel.grid.minor.y = element_blank(),\n",
    "        panel.grid.major.x = element_blank(),\n",
    "        plot.margin = unit(c(0.5, 0.1, 0.5, 0.1), \"cm\"),\n",
    "        text = element_text(family = \"Arial\"))+\n",
    "  scale_y_continuous(expand = c(0.03, 0), breaks = seq(0, 300, 50)) +\n",
    "  facet_wrap(~ecosystem, scales = \"free_x\", nrow = 1)+\n",
    "  guides(fill = guide_legend(ncol = 4))\n",
    "pl\n",
    "# save plot for legend\n",
    "ggsave(\"MAGsHits_BarChart_loose_legend.tiff\", pl, width = 25.4, height = 26.67, dpi = 300, units = \"cm\")\n",
    "\n",
    "# done\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55130df1-deb0-4bcd-adef-0a21c9d20c6d",
   "metadata": {},
   "source": [
    "# loose 標準的 PercentageBarChart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b28a6b39-160f-47ad-8e5e-8c029e58fd01",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ggplot2)\n",
    "library(readr)\n",
    "library(gridExtra)\n",
    "library(patchwork)\n",
    "library(grid)\n",
    "library(cowplot)\n",
    "library(stringr)\n",
    "library(forcats)\n",
    "library(dplyr)\n",
    "\n",
    "# convert the table to a data frame\n",
    "HMM_Final <- read.csv(\"B50_HMMFinal_loose_ForBarChart.csv\", header=TRUE, sep = ',')\n",
    "\n",
    "# Replace all \"Non-marine Saline and Alkaline\" values in column \"ecosystem_category\" with \"Salt Lake\"\n",
    "HMM_Final$ecosystem_category[HMM_Final$ecosystem_category == \"Non-marine Saline and Alkaline\"] <- \"Salt Lake\"\n",
    "\n",
    "# Add new column with percentage values\n",
    "HMM_Final <- HMM_Final %>%\n",
    "  group_by(ecosystem_category) %>%\n",
    "  mutate(percent = NumberOfMAGs / sum(NumberOfMAGs) * 100)\n",
    "HMM_Final\n",
    "\n",
    "\n",
    "# Check unique value of dataframe\n",
    "unique(HMM_Final$ecosystem)\n",
    "unique(HMM_Final$ecosystem_category)\n",
    "unique(HMM_Final$taxonomy)\n",
    "\n",
    "# Change taxonomy order\n",
    "taxonomy_order <- c( \"Unclassified Bacteria\", 'Other Bacterial Phylum','Myxococcia', 'Thermodesulfobacterota',\n",
    "                     \"Chloroflexi\", \"Alphaproteobacteria\", \"Gammaproteobacteria\", \"Actinobacteria\")\n",
    "\n",
    "HMM_Final$taxonomy <- fct_relevel(HMM_Final$taxonomy, taxonomy_order)\n",
    "HMM_Final\n",
    "\n",
    "# change ecosystem order\n",
    "ecosystem_order <- c(\"Aquatic\", \"Terrestrial\", \"Host-associated\", \"Engineered\")\n",
    "HMM_Final$ecosystem <- fct_relevel(HMM_Final$ecosystem, ecosystem_order)\n",
    "\n",
    "# create the color vector\n",
    "class_colors <- c(\"Alphaproteobacteria\" = \"#2e75b6\",\n",
    "                  \"Gammaproteobacteria\" = \"#9dc3e6\", \n",
    "                  \"Actinobacteria\" = \"#c00000\",\n",
    "                  \"Chloroflexi\" = \"#a9d18e\",\n",
    "                  'Thermodesulfobacterota' = '#ffd966',\n",
    "                  \"Myxococcia\" = \"#7e33b8\",\n",
    "                  \"Other Bacterial Phylum\" = \"#404040\",\n",
    "                  \"Unclassified Bacteria\" = \"#cbcbcb\"\n",
    ")\n",
    "\n",
    "# ggplot No legend\n",
    "p <- ggplot(HMM_Final, aes(x = str_wrap(ecosystem_category, width = 30), y = percent, fill = taxonomy)) +\n",
    "  geom_bar(stat = \"identity\",\n",
    "           width = ifelse(HMM_Final$ecosystem == \"Terrestrial\", 0.4, 0.8),\n",
    "           position = \"stack\") +\n",
    "  labs(title = \"\",\n",
    "       x =\"\",\n",
    "       y = \"\")+\n",
    "  scale_fill_manual(values = class_colors,\n",
    "                    limits = c(\"Actinobacteria\", \"Chloroflexi\", \"Alphaproteobacteria\", \"Gammaproteobacteria\",\n",
    "                               'Myxococcia', 'Thermodesulfobacterota', 'Other Bacterial Phylum', \"Unclassified Bacteria\"))+\n",
    "  theme_bw()+\n",
    "  theme(plot.title = element_text(size = 14, face = \"bold\", hjust = 0.5),\n",
    "        axis.title.y = element_text(size = 11, margin = margin(r = 10), face = \"bold\"),\n",
    "        strip.text = element_text(size = 11, face = \"bold\"),\n",
    "        axis.text.x = element_text(size = 10, angle = 45, hjust = 1, color =\"black\"),\n",
    "        axis.text.y = element_blank(),\n",
    "        axis.ticks.y = element_blank(),\n",
    "        legend.text = element_text(size = 10),\n",
    "        legend.position = \"none\",\n",
    "        legend.title = element_blank(),\n",
    "        panel.grid.minor.y = element_blank(),\n",
    "        panel.grid.major.x = element_blank(),\n",
    "        plot.margin = unit(c(-0.5, 0.1, -0.5, 0.8), \"cm\"),\n",
    "        text = element_text(family = \"Arial\"))+\n",
    "  scale_y_continuous(expand = c(0.01, 0)) +\n",
    "  facet_wrap(~ecosystem, scales = \"free_x\", nrow = 1)+\n",
    "  guides(fill = guide_legend(ncol = 4))\n",
    "p\n",
    "\n",
    "# save plot with ggsave\n",
    "# unit = 'cm': 高圖參數: (width = 18.18, height = 26.67, dpi = 300) / 矮圖參數: (width = 18.18, height = 7, dpi = 300)\n",
    "ggsave(\"MAGsHits_PercentageBarChart_loose.tiff\", p, width = 18.18, height = 7, dpi = 300, units = \"cm\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2475f116-30ac-44d1-befa-9da206c0cb11",
   "metadata": {},
   "source": [
    "# loose 標準的 World map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b21700b-a478-4d85-a355-9eb682d26ff9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# download the shapefile from Natural Earth Data\n",
    "download.file(\"https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/110m/cultural/ne_110m_admin_0_countries.zip\", \"ne_110m_admin_0_countries.zip\")\n",
    "\n",
    "# unzip the file\n",
    "unzip(\"ne_110m_admin_0_countries.zip\")\n",
    "\n",
    "# load the shapefile as an sf object\n",
    "library(sf)\n",
    "world <- st_read(\"ne_110m_admin_0_countries.shp\")\n",
    "\n",
    "# clean up\n",
    "unlink(\"ne_110m_admin_0_countries.zip\")\n",
    "unlink(\"ne_110m_admin_0_countries.shp\")\n",
    "unlink(\"ne_110m_admin_0_countries.shx\")\n",
    "unlink(\"ne_110m_admin_0_countries.dbf\")\n",
    "\n",
    "#-----------------------------------------------\n",
    "# Load required packages\n",
    "library(ggplot2)\n",
    "library(sf)\n",
    "library(maps)\n",
    "library(dplyr)\n",
    "library(forcats)\n",
    "library(cowplot)\n",
    "library(scales)\n",
    "\n",
    "# read table\n",
    "df <- read.csv(\"B50_HMMFinal_loose_ForMaps.csv\")\n",
    "df\n",
    "\n",
    "# Replace all \"Non-marine Saline and Alkaline\" values in column \"ecosystem_category\" with \"Salt Lake\"\n",
    "df$Ecosystem_Category[df$Ecosystem_Category == \"Non-marine Saline and Alkaline\"] <- \"Salt Lake\"\n",
    "\n",
    "# Get unique ecosystem names\n",
    "ecosystems <- unique(df$Ecosystem)\n",
    "\n",
    "# Create empty list to store dataframes\n",
    "df_list <- list()\n",
    "\n",
    "# Loop through each ecosystem and create a new dataframe for each\n",
    "for (i in ecosystems) {\n",
    "  temp_df <- df[df$Ecosystem == i, ]\n",
    "  df_list[[i]] <- temp_df\n",
    "}\n",
    "\n",
    "# View list of dataframes\n",
    "df_list\n",
    "\n",
    "# rename seperate list\n",
    "Aquatic_df <- do.call(rbind, df_list[1])\n",
    "Engineered_df <- do.call(rbind, df_list[2])\n",
    "Host_associate_df <- do.call(rbind, df_list[3])\n",
    "Terrestrial_df <- do.call(rbind, df_list[4])\n",
    "\n",
    "# Convert the dataframe to an sf object and set coordinates\n",
    "Aquatic_sf <- st_as_sf(Aquatic_df, coords = c(\"Longitude\", \"Latitude\"), crs = 4326)\n",
    "Engineered_sf <- st_as_sf(Engineered_df, coords = c(\"Longitude\", \"Latitude\"), crs = 4326)\n",
    "Host_associate_sf <- st_as_sf(Host_associate_df, coords = c(\"Longitude\", \"Latitude\"), crs = 4326)\n",
    "Terrestrial_sf <- st_as_sf(Terrestrial_df, coords = c(\"Longitude\", \"Latitude\"), crs = 4326)\n",
    "\n",
    "#---------------------------\n",
    "# plot Aquatic map\n",
    "\n",
    "# change ecosystem_category order\n",
    "ecosystem_category_order <- c(\"Marine\", \"Freshwater\", \"Salt Lake\", \"Sediment\", 'Thermal springs')\n",
    "Aquatic_sf$Ecosystem_Category <- fct_relevel(Aquatic_sf$Ecosystem_Category, ecosystem_category_order)\n",
    "\n",
    "# create the color vector\n",
    "Aquatic_colors <- c(\"Marine\" = \"#1e4c9c\",\n",
    "                    \"Freshwater\" = \"#4286bf\",\n",
    "                    \"Salt Lake\" = \"#83b4d3\", \n",
    "                    \"Sediment\" = \"#404040\",\n",
    "                    'Thermal springs'= '#7f7f7f'\n",
    ")\n",
    "\n",
    "# Plot\n",
    "Aquatic_map <-\n",
    "  ggplot(data=world)+\n",
    "  geom_sf(fill='#dadada', color='#dadada')+\n",
    "  geom_sf(data=Aquatic_sf, aes(color = Ecosystem_Category,\n",
    "                               size = count))+\n",
    "  theme_void()+\n",
    "  theme(\n",
    "    legend.text = element_text(size = 12),\n",
    "    legend.title = element_blank(),\n",
    "    legend.position = 'bottom',\n",
    "    plot.margin = unit(c(0, 0.3, 0, 0.1), \"cm\"),\n",
    "    legend.margin = margin(t = -0.5, r = 0, b = 0, l = 0, unit = \"cm\")\n",
    "  )+\n",
    "  scale_size_continuous(breaks = c(1, 8, 15), labels = c(\"1\", \"8\", \"15\"), name = \"Count\")+\n",
    "  scale_color_manual(values = Aquatic_colors,\n",
    "                     label = c('Marine', 'Freshwater', 'Salt Lake', 'Sediment', 'Thermal springs'),\n",
    "                     guide = guide_legend(nrow = 2))\n",
    "\n",
    "Aquatic_map\n",
    "\n",
    "# save plot\n",
    "ggsave(\"loose_Aquatic_Maps_Main.tiff\", Aquatic_map, width = 18.18, height = 8, dpi = 300, units = \"cm\")\n",
    "\n",
    "#--------------------------------------------------------------------------------\n",
    "# Load required packages\n",
    "library(ggplot2)\n",
    "library(sf)\n",
    "library(maps)\n",
    "library(dplyr)\n",
    "library(forcats)\n",
    "library(cowplot)\n",
    "library(scales)\n",
    "\n",
    "# read table\n",
    "df <- read.csv(\"B50_HMMFinal_loose_ForMaps.csv\")\n",
    "df\n",
    "\n",
    "# Replace all \"Non-marine Saline and Alkaline\" values in column \"ecosystem_category\" with \"Salt Lake\"\n",
    "df$Ecosystem_Category[df$Ecosystem_Category == \"Non-marine Saline and Alkaline\"] <- \"Salt Lake\"\n",
    "\n",
    "# Get unique ecosystem names\n",
    "ecosystems <- unique(df$Ecosystem)\n",
    "\n",
    "# Create empty list to store dataframes\n",
    "df_list <- list()\n",
    "\n",
    "# Loop through each ecosystem and create a new dataframe for each\n",
    "for (i in ecosystems) {\n",
    "  temp_df <- df[df$Ecosystem == i, ]\n",
    "  df_list[[i]] <- temp_df\n",
    "}\n",
    "\n",
    "# View list of dataframes\n",
    "df_list\n",
    "\n",
    "# rename seperate list\n",
    "Aquatic_df <- do.call(rbind, df_list[1])\n",
    "Engineered_df <- do.call(rbind, df_list[2])\n",
    "Host_associate_df <- do.call(rbind, df_list[3])\n",
    "Terrestrial_df <- do.call(rbind, df_list[4])\n",
    "\n",
    "# Convert the dataframe to an sf object and set coordinates\n",
    "Aquatic_sf <- st_as_sf(Aquatic_df, coords = c(\"Longitude\", \"Latitude\"), crs = 4326)\n",
    "Engineered_sf <- st_as_sf(Engineered_df, coords = c(\"Longitude\", \"Latitude\"), crs = 4326)\n",
    "Host_associate_sf <- st_as_sf(Host_associate_df, coords = c(\"Longitude\", \"Latitude\"), crs = 4326)\n",
    "Terrestrial_sf <- st_as_sf(Terrestrial_df, coords = c(\"Longitude\", \"Latitude\"), crs = 4326)\n",
    "\n",
    "#---------------------------\n",
    "# plot Aquatic map\n",
    "\n",
    "# change ecosystem_category order\n",
    "ecosystem_category_order <- c(\"Marine\", \"Freshwater\", \"Salt Lake\", \"Sediment\", 'Thermal springs')\n",
    "Aquatic_sf$Ecosystem_Category <- fct_relevel(Aquatic_sf$Ecosystem_Category, ecosystem_category_order)\n",
    "\n",
    "# create the color vector\n",
    "Aquatic_colors <- c(\"Marine\" = \"#1e4c9c\",\n",
    "                    \"Freshwater\" = \"#4286bf\",\n",
    "                    \"Salt Lake\" = \"#83b4d3\", \n",
    "                    \"Sediment\" = \"#404040\",\n",
    "                    'Thermal springs'= '#7f7f7f'\n",
    ")\n",
    "\n",
    "# Plot\n",
    "Aquatic_map <-\n",
    "  ggplot(data=world)+\n",
    "  geom_sf(fill='#dadada', color='#dadada')+\n",
    "  geom_sf(data=Aquatic_sf, aes(color = Ecosystem_Category,\n",
    "                               size = count))+\n",
    "  theme_void()+\n",
    "  theme(\n",
    "    legend.text = element_text(size = 12),\n",
    "    legend.title = element_blank(),\n",
    "    legend.position = 'bottom',\n",
    "    plot.margin = unit(c(0, 0.3, 0, 0.1), \"cm\"),\n",
    "    legend.margin = margin(t = -0.5, r = 0, b = 0, l = 0, unit = \"cm\")\n",
    "  )+\n",
    "  scale_size_continuous(breaks = c(1, 10, 20), labels = c(\"1\", \"10\", \"20\"), name = \"Count\", limits = c(1, 20), guide = \"none\")+\n",
    "  scale_color_manual(values = Aquatic_colors,\n",
    "                     label = c('Marine', 'Freshwater', 'Salt Lake', 'Sediment', 'Thermal springs'),\n",
    "                     guide = guide_legend(nrow = 2))\n",
    "\n",
    "Aquatic_map\n",
    "\n",
    "# save plot\n",
    "ggsave(\"loose_Aquatic_Maps_SI.tiff\", Aquatic_map, width = 18.18, height = 8, dpi = 300, units = \"cm\")\n",
    "\n",
    "\n",
    "\n",
    "#---------------------------\n",
    "# plot Terrestrial map\n",
    "\n",
    "# change ecosystem_category order\n",
    "ecosystem_category_order <- c(\"Soil\", \"Deep subsurface\", \"Peat\")\n",
    "Terrestrial_sf$Ecosystem_Category <- fct_relevel(Terrestrial_sf$Ecosystem_Category, ecosystem_category_order)\n",
    "\n",
    "# create the color vector\n",
    "Terrestrial_colors <- c(\"Soil\" = \"#0d7532\",\n",
    "                        \"Deep subsurface\" = \"#39a646\",\n",
    "                        \"Peat\" = \"#8ed967\"\n",
    ")\n",
    "\n",
    "# Plot\n",
    "Terrestrial_map <-\n",
    "  ggplot(data=world)+\n",
    "  geom_sf(fill='#dadada', color='#dadada')+\n",
    "  geom_sf(data=Terrestrial_sf, aes(color = Ecosystem_Category,\n",
    "                                   size = count))+\n",
    "  theme_void()+\n",
    "  theme(\n",
    "    legend.text = element_text(size = 12),\n",
    "    legend.title = element_blank(),\n",
    "    legend.position = 'bottom',\n",
    "    plot.margin = unit(c(0, 0.3, 0, 0.1), \"cm\"),\n",
    "    legend.margin = margin(t = -0.5, r = 0, b = 0, l = 0, unit = \"cm\")\n",
    "  )+\n",
    "  scale_size_continuous(breaks = c(1, 10, 20), labels = c(\"1\", \"10\", \"20\"), name = \"Count\", limits = c(1, 20), guide = \"none\")+\n",
    "  scale_color_manual(values = Terrestrial_colors,\n",
    "                     label = c(\"Soil\", \"Deep subsurface\", \"Peat\"),\n",
    "                     guide = guide_legend(nrow = 2))\n",
    "Terrestrial_map\n",
    "\n",
    "# save plot\n",
    "ggsave(\"loose_Terrestrial_map_SI.tiff\", Terrestrial_map, width = 18.18, height = 8, dpi = 300, units = \"cm\")\n",
    "\n",
    "\n",
    "#--------------------\n",
    "# plot Host_associate map\n",
    "\n",
    "# change ecosystem_category order\n",
    "ecosystem_category_order <- c(\"Porifera\", \"Plants\", \"Fungi\", 'Mammals', 'Annelida', 'Invertebrates', 'Microbial')\n",
    "Host_associate_sf$Ecosystem_Category <- fct_relevel(Host_associate_sf$Ecosystem_Category, ecosystem_category_order)\n",
    "\n",
    "# create the color vector\n",
    "Host_associate_colors <- c(\"Porifera\" = \"#f69a21\",\n",
    "                           \"Plants\" = \"#ffca51\",\n",
    "                           \"Fungi\" = \"#fde596\",\n",
    "                           \"Mammals\" = \"#fff7c5\",\n",
    "                           \"Annelida\" = \"#404040\",\n",
    "                           \"Invertebrates\" = \"#7f7f7f\",\n",
    "                           \"Microbial\" = \"#ccc59d\"\n",
    ")\n",
    "\n",
    "# Plot\n",
    "Host_associate_map <-\n",
    "  ggplot(data=world)+\n",
    "  geom_sf(fill='#dadada', color='#dadada')+\n",
    "  geom_sf(data=Host_associate_sf, aes(color = Ecosystem_Category,\n",
    "                                      size = count))+\n",
    "  theme_void()+\n",
    "  theme(\n",
    "    legend.text = element_text(size = 12),\n",
    "    legend.title = element_blank(),\n",
    "    legend.position = 'bottom',\n",
    "    plot.margin = unit(c(0, 0.3, 0, 0.1), \"cm\"),\n",
    "    legend.margin = margin(t = -0.5, r = 0, b = 0, l = 0, unit = \"cm\")\n",
    "  )+\n",
    "  scale_size_continuous(breaks = c(1, 10, 20), labels = c(\"1\", \"10\", \"20\"), name = \"Count\", limits = c(1, 20), guide = \"none\")+\n",
    "  scale_color_manual(values = Host_associate_colors,\n",
    "                     label = c(\"Porifera\", \"Plants\", \"Fungi\", 'Mammals', 'Annelida', 'Invertebrates', 'Microbial'),\n",
    "                     guide = guide_legend(nrow = 2))\n",
    "Host_associate_map\n",
    "\n",
    "# save plot\n",
    "ggsave(\"loose_Host_associate_map_SI.tiff\", Host_associate_map, width = 18.18, height = 8, dpi = 300, units = \"cm\")\n",
    "\n",
    "\n",
    "#----------------------------\n",
    "# plot Engineered map\n",
    "\n",
    "# change ecosystem_category order\n",
    "ecosystem_category_order <- c(\"Wastewater\", \"Solid waste\", \"Built environment\", 'Lab enrichment', 'Modeled', 'Bioreactor', 'Lab synthesis')\n",
    "Engineered_sf$Ecosystem_Category <- fct_relevel(Engineered_sf$Ecosystem_Category, ecosystem_category_order)\n",
    "\n",
    "# create the color vector\n",
    "Engineered_colors <- c(\"Wastewater\" = \"#a3231a\",\n",
    "                       \"Solid waste\" = \"#d92e16\",\n",
    "                       \"Built environment\" = \"#f56a45\",\n",
    "                       \"Lab enrichment\" = \"#ffb799\",\n",
    "                       \"Modeled\" = \"#ffece3\",\n",
    "                       \"Bioreactor\" = \"#404040\",\n",
    "                       \"Lab synthesis\" = \"#7f7f7f\"\n",
    ")\n",
    "\n",
    "# Plot\n",
    "Engineered_map <-\n",
    "  ggplot(data=world)+\n",
    "  geom_sf(fill='#dadada', color='#dadada')+\n",
    "  geom_sf(data=Engineered_sf, aes(color = Ecosystem_Category,\n",
    "                                  size = count))+\n",
    "  theme_void()+\n",
    "  theme(\n",
    "    legend.text = element_text(size = 12),\n",
    "    legend.title = element_blank(),\n",
    "    legend.position = 'bottom',\n",
    "    plot.margin = unit(c(0, 0.3, 0, 0.1), \"cm\"),\n",
    "    legend.margin = margin(t = -0.5, r = 0, b = 0, l = 0, unit = \"cm\")\n",
    "  )+\n",
    "  scale_size_continuous(breaks = c(1, 10, 20), labels = c(\"1\", \"10\", \"20\"), name = \"Count\", limits = c(1, 20), guide = \"none\")+\n",
    "  scale_color_manual(values = Engineered_colors,\n",
    "                     label = c(\"Wastewater\", \"Solid waste\", \"Built environment\", 'Lab enrichment', 'Modeled', 'Bioreactor', 'Lab synthesis'),\n",
    "                     guide = guide_legend(nrow = 2))\n",
    "Engineered_map\n",
    "\n",
    "# save plot\n",
    "ggsave(\"loose_Engineered_map_SI.tiff\", Engineered_map, width = 18.18, height = 8, dpi = 300, units = \"cm\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab1ccaed-4824-4296-b93b-300fa93e2a50",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# 不執行"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cc4cd5f-4dea-48bf-8cb5-5247b6a2d52e",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Crtical 標準的 facet stacked bar char"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a3eb544-9508-4f90-bce5-b3e8dfe74994",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ggplot2)\n",
    "library(readr)\n",
    "library(gridExtra)\n",
    "library(patchwork)\n",
    "library(grid)\n",
    "library(cowplot)\n",
    "library(stringr)\n",
    "library(forcats)\n",
    "\n",
    "\n",
    "# convert the table to a data frame\n",
    "HMM_Final <- read.csv(\"B50_HMMFinal_critical_ForBarChart.csv\", header=TRUE, sep = ',')\n",
    "\n",
    "# Check unique value of dataframe\n",
    "unique(HMM_Final$ecosystem)\n",
    "unique(HMM_Final$ecosystem_type)\n",
    "unique(HMM_Final$taxonomy)\n",
    "\n",
    "# Change taxonomy order\n",
    "# create the taxonomy order vactor\n",
    "taxonomy_order <- c(\"p_Actinobacteriota_c_Actinobacteria\", \"p_Actinobacteriota_c_Acidimicrobiia\", \n",
    "                    \"p_Chloroflexota_c_Dehalococcoidia\", \"p_Chloroflexota_c_Chloroflexia\",\n",
    "                    \"p_Myxococcota\", \"p_Desulfobacterota\", \"p_Other bacterial phyla\", \"p_Unclassified Bacteria\",\n",
    "                    \"p_Proteobacteria_c_Alphaproteobacteria\", \"p_Proteobacteria_c_Gammaproteobacteria\")\n",
    "\n",
    "HMM_Final$taxonomy <- fct_relevel(HMM_Final$taxonomy, taxonomy_order)\n",
    "HMM_Final\n",
    "\n",
    "# change ecosystem order\n",
    "ecosystem_order <- c(\"Aquatic\", \"Terrestrial\", \"Host-associated\", \"Engineered\")\n",
    "HMM_Final$ecosystem <- fct_relevel(HMM_Final$ecosystem, ecosystem_order)\n",
    "HMM_Final\n",
    "\n",
    "# create the color vector\n",
    "class_colors <- c(\"p_Proteobacteria_c_Alphaproteobacteria\" = \"#2e75b6\",\n",
    "                  \"p_Proteobacteria_c_Gammaproteobacteria\" = \"#9dc3e6\",\n",
    "                  \"p_Actinobacteriota_c_Actinobacteria\" = \"#ff1000\", \n",
    "                  \"p_Actinobacteriota_c_Acidimicrobiia\" = \"#c00000\",\n",
    "                  'p_Actinobacteriota_c_Thermoleophilia'= '#F05E54',\n",
    "                  \"p_Chloroflexota_c_Dehalococcoidia\" = \"#ceffad\",\n",
    "                  \"p_Chloroflexota_c_Chloroflexia\" = \"#a9d18e\",\n",
    "                  \"p_Myxococcota\" = \"#ffd966\",\n",
    "                  \"p_Desulfobacterota\" = \"#fe660c\",\n",
    "                  \"p_Other bacterial phyla\" = \"#7e33b8\",\n",
    "                  \"p_Unclassified Bacteria\" = \"#767171\"\n",
    ")\n",
    "\n",
    "# ggplot\n",
    "p <- ggplot(HMM_Final, aes(x = str_wrap(ecosystem_type, width = 20), y = NumberOfMAGs, fill = taxonomy)) +\n",
    "  geom_bar(stat = \"identity\",\n",
    "           width = ifelse(HMM_Final$ecosystem == \"Engineered\" | HMM_Final$ecosystem == \"Terrestrial\", 0.27, 0.8),\n",
    "           position = \"stack\") +\n",
    "  labs(title = \"Metagenomic analysis\",\n",
    "       x =\"\",\n",
    "       y = \"estrogen degraders\")+\n",
    "  scale_fill_manual(values = class_colors,\n",
    "                    limits = c(\"p_Proteobacteria_c_Alphaproteobacteria\", \"p_Proteobacteria_c_Gammaproteobacteria\",\n",
    "                               \"p_Actinobacteriota_c_Acidimicrobiia\", \"p_Myxococcota\")) +\n",
    "  theme_bw()+\n",
    "  theme(plot.title = element_text(size = 12, face = \"bold\", hjust = 0.5),\n",
    "        axis.title.y = element_text(size = 10, margin = margin(r = 10), face = \"bold\"),\n",
    "        strip.text = element_text(size = 10, face = \"bold\"),\n",
    "        axis.text.x = element_text(size = 8, angle = 45, hjust = 1, color =\"black\"),\n",
    "        axis.text.y = element_text(size = 8),\n",
    "        legend.text = element_text(size = 8),\n",
    "        legend.position = \"bottom\",\n",
    "        legend.title = element_blank(),\n",
    "        panel.grid.minor.y = element_blank(),\n",
    "        panel.grid.major.x = element_blank(),\n",
    "        plot.margin = unit(c(0.5, 0.1, 0.5, 0.1), \"cm\"),\n",
    "        text = element_text(family = \"Arial\"))+\n",
    "  facet_wrap(~ecosystem, scales = \"free_x\", nrow = 1)+\n",
    "  guides(fill = guide_legend(ncol = 2))\n",
    "p\n",
    "\n",
    "# save your plot with ggsave\n",
    "ggsave(\"MAGsHits_distribution_critical_facet.tiff\", p, width = 7.16, height = 10.5, dpi = 300)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "753542d7-40f2-4897-b73e-e2c8f0463729",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## >=10 hits 標準的 facet stacked bar char"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a12fd218-3147-4a1e-8acf-94a4127d6f7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ggplot2)\n",
    "library(readr)\n",
    "library(gridExtra)\n",
    "library(patchwork)\n",
    "library(grid)\n",
    "library(cowplot)\n",
    "library(stringr)\n",
    "library(forcats)\n",
    "\n",
    "\n",
    "# convert the table to a data frame\n",
    "HMM_Final <- read_csv(\"B50_HMMFinal_10Hits_ForBarChart.csv\")\n",
    "\n",
    "# Check unique value of dataframe\n",
    "unique(HMM_Final$ecosystem)\n",
    "unique(HMM_Final$ecosystem_type)\n",
    "unique(HMM_Final$taxonomy)\n",
    "\n",
    "# Change taxonomy order\n",
    "taxonomy_order <- c(\"p_Myxococcota\", \"p_Desulfobacterota\", \"p_Other bacterial phyla\", \"p_Unclassified Bacteria\",\n",
    "                    \"p_Chloroflexota_c_Dehalococcoidia\", \"p_Chloroflexota_c_Chloroflexia\",\n",
    "                    \"p_Proteobacteria_c_Alphaproteobacteria\", \"p_Proteobacteria_c_Gammaproteobacteria\",\n",
    "                    \"p_Actinobacteriota_c_Actinobacteria\", \"p_Actinobacteriota_c_Acidimicrobiia\")\n",
    "\n",
    "HMM_Final$taxonomy <- fct_relevel(HMM_Final$taxonomy, taxonomy_order)\n",
    "HMM_Final\n",
    "\n",
    "# change ecosystem order\n",
    "ecosystem_order <- c(\"Aquatic\", \"Terrestrial\", \"Host-associated\", \"Engineered\")\n",
    "HMM_Final$ecosystem <- fct_relevel(HMM_Final$ecosystem, ecosystem_order)\n",
    "HMM_Final\n",
    "\n",
    "# create the color vector\n",
    "class_colors <- c(\"p_Proteobacteria_c_Alphaproteobacteria\" = \"#2e75b6\",\n",
    "                  \"p_Proteobacteria_c_Gammaproteobacteria\" = \"#9dc3e6\",\n",
    "                  \"p_Actinobacteriota_c_Actinobacteria\" = \"#ff1000\", \n",
    "                  \"p_Actinobacteriota_c_Acidimicrobiia\" = \"#c00000\",\n",
    "                  'p_Actinobacteriota_c_Thermoleophilia'= '#F05E54',\n",
    "                  \"p_Chloroflexota_c_Dehalococcoidia\" = \"#ceffad\",\n",
    "                  \"p_Chloroflexota_c_Chloroflexia\" = \"#a9d18e\",\n",
    "                  \"p_Myxococcota\" = \"#ffd966\",\n",
    "                  \"p_Desulfobacterota\" = \"#fe660c\",\n",
    "                  \"p_Other bacterial phyla\" = \"#7e33b8\",\n",
    "                  \"p_Unclassified Bacteria\" = \"#767171\"\n",
    ")\n",
    "\n",
    "# ggplot\n",
    "p <- ggplot(HMM_Final, aes(x = str_wrap(ecosystem_type, width = 30), y = NumberOfMAGs, fill = taxonomy)) +\n",
    "  geom_bar(stat = \"identity\",\n",
    "           width = ifelse(HMM_Final$ecosystem == \"Terrestrial\", 0.3, 0.7),\n",
    "           position = \"stack\") +\n",
    "  labs(title = \"Metagenomic analysis\",\n",
    "       x =\"\",\n",
    "       y = \"estrogen degraders\")+\n",
    "  scale_fill_manual(values = class_colors,\n",
    "                    limits = c(\"p_Actinobacteriota_c_Actinobacteria\", \"p_Actinobacteriota_c_Acidimicrobiia\", \n",
    "                               \"p_Proteobacteria_c_Alphaproteobacteria\", \"p_Proteobacteria_c_Gammaproteobacteria\", \n",
    "                               \"p_Unclassified Bacteria\", \"p_Myxococcota\")) +\n",
    "  theme_bw()+\n",
    "  theme(plot.title = element_text(size = 12, face = \"bold\", hjust = 0.5),\n",
    "        axis.title.y = element_text(size = 10, margin = margin(r = 10), face = \"bold\"),\n",
    "        strip.text = element_text(size = 10, face = \"bold\"),\n",
    "        axis.text.x = element_text(size = 8, angle = 45, hjust = 1, color =\"black\"),\n",
    "        axis.text.y = element_text(size = 8),\n",
    "        legend.text = element_text(size = 8),\n",
    "        legend.position = \"bottom\",\n",
    "        legend.title = element_blank(),\n",
    "        panel.grid.minor.y = element_blank(),\n",
    "        panel.grid.major.x = element_blank(),\n",
    "        plot.margin = unit(c(0.5, 0.1, 0.5, 0.1), \"cm\"),\n",
    "        text = element_text(family = \"Arial\"))+\n",
    "  facet_wrap(~ecosystem, scales = \"free_x\", nrow = 1)+\n",
    "  guides(fill = guide_legend(ncol = 3))\n",
    "p\n",
    "\n",
    "# save your plot with ggsave\n",
    "ggsave(\"MAGsHits_distribution_10Hits_facet.tiff\", p, width = 7.16, height = 10.5, dpi = 300)\n",
    "\n",
    "\n",
    "# Check legend\n",
    "#p <- ggplot(HMM_Final, aes(x = str_wrap(ecosystem_type, width = 30), y = NumberOfMAGs, fill = taxonomy)) +\n",
    "#  geom_bar(stat = \"identity\")\n",
    "#p\n",
    "#"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b7cdffa-b4a2-4052-a8fa-a8ade8b36767",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## reference 標準的 facet stacked bar char"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da47a62a-6480-4c3e-b352-c27c99fab919",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ggplot2)\n",
    "library(readr)\n",
    "library(gridExtra)\n",
    "library(patchwork)\n",
    "library(grid)\n",
    "library(cowplot)\n",
    "library(stringr)\n",
    "library(forcats)\n",
    "\n",
    "\n",
    "# convert the table to a data frame\n",
    "HMM_Final <- read_csv(\"B50_HMMFinal_reference_ForBarChart.csv\")\n",
    "\n",
    "# Check unique value of dataframe\n",
    "unique(HMM_Final$ecosystem)\n",
    "unique(HMM_Final$ecosystem_type)\n",
    "unique(HMM_Final$taxonomy)\n",
    "\n",
    "# Change taxonomy order\n",
    "taxonomy_order <- c(\"p_Myxococcota\", \"p_Desulfobacterota\", \"p_Other bacterial phyla\", \"p_Unclassified Bacteria\",\n",
    "                    \"p_Chloroflexota_c_Dehalococcoidia\", \"p_Chloroflexota_c_Chloroflexia\",\n",
    "                    \"p_Actinobacteriota_c_Acidimicrobiia\", \"p_Actinobacteriota_c_Actinobacteria\",\n",
    "                    \"p_Proteobacteria_c_Alphaproteobacteria\", \"p_Proteobacteria_c_Gammaproteobacteria\")\n",
    "\n",
    "HMM_Final$taxonomy <- fct_relevel(HMM_Final$taxonomy, taxonomy_order)\n",
    "HMM_Final\n",
    "\n",
    "# change ecosystem order\n",
    "ecosystem_order <- c(\"Aquatic\", \"Terrestrial\", \"Host-associated\", \"Engineered\")\n",
    "HMM_Final$ecosystem <- fct_relevel(HMM_Final$ecosystem, ecosystem_order)\n",
    "HMM_Final\n",
    "\n",
    "# create the color vector\n",
    "class_colors <- c(\"p_Proteobacteria_c_Alphaproteobacteria\" = \"#2e75b6\",\n",
    "                  \"p_Proteobacteria_c_Gammaproteobacteria\" = \"#9dc3e6\",\n",
    "                  \"p_Actinobacteriota_c_Actinobacteria\" = \"#ff1000\", \n",
    "                  \"p_Actinobacteriota_c_Acidimicrobiia\" = \"#c00000\",\n",
    "                  'p_Actinobacteriota_c_Thermoleophilia'= '#F05E54',\n",
    "                  \"p_Chloroflexota_c_Dehalococcoidia\" = \"#ceffad\",\n",
    "                  \"p_Chloroflexota_c_Chloroflexia\" = \"#a9d18e\",\n",
    "                  \"p_Myxococcota\" = \"#ffd966\",\n",
    "                  \"p_Desulfobacterota\" = \"#fe660c\",\n",
    "                  \"p_Other bacterial phyla\" = \"#7e33b8\",\n",
    "                  \"p_Unclassified Bacteria\" = \"#767171\"\n",
    ")\n",
    "\n",
    "# ggplot\n",
    "p <- ggplot(HMM_Final, aes(x = str_wrap(ecosystem_type, width = 30), y = NumberOfMAGs, fill = taxonomy)) +\n",
    "  geom_bar(stat = \"identity\",\n",
    "           width = ifelse(HMM_Final$ecosystem == \"Terrestrial\", 0.3, 0.7),\n",
    "           position = \"stack\") +\n",
    "  labs(title = \"Metagenomic analysis\",\n",
    "       x =\"\",\n",
    "       y = \"estrogen degraders\")+\n",
    "  scale_fill_manual(values = class_colors,\n",
    "                    limits = c( \n",
    "                      \"p_Proteobacteria_c_Alphaproteobacteria\", \"p_Proteobacteria_c_Gammaproteobacteria\",\n",
    "                      \"p_Actinobacteriota_c_Actinobacteria\", \"p_Actinobacteriota_c_Acidimicrobiia\", \n",
    "                               \"p_Unclassified Bacteria\", \"p_Myxococcota\")) +\n",
    "  theme_bw()+\n",
    "  theme(plot.title = element_text(size = 12, face = \"bold\", hjust = 0.5),\n",
    "        axis.title.y = element_text(size = 10, margin = margin(r = 10), face = \"bold\"),\n",
    "        strip.text = element_text(size = 10, face = \"bold\"),\n",
    "        axis.text.x = element_text(size = 8, angle = 45, hjust = 1, color =\"black\"),\n",
    "        axis.text.y = element_text(size = 8),\n",
    "        legend.text = element_text(size = 8),\n",
    "        legend.position = \"bottom\",\n",
    "        legend.title = element_blank(),\n",
    "        panel.grid.minor.y = element_blank(),\n",
    "        panel.grid.major.x = element_blank(),\n",
    "        plot.margin = unit(c(0.5, 0.1, 0.5, 0.1), \"cm\"),\n",
    "        text = element_text(family = \"Arial\"))+\n",
    "  facet_wrap(~ecosystem, scales = \"free_x\", nrow = 1)+\n",
    "  facet_grid(~ecosystem, scales = \"free_x\")+\n",
    "  guides(fill = guide_legend(ncol = 3))\n",
    "p\n",
    "# save your plot with ggsave\n",
    "ggsave(\"MAGsHits_distribution_reference_facet.tiff\", p, width = 7.16, height = 10.5, dpi = 300)\n",
    "\n",
    "\n",
    "# Check legend\n",
    "#p <- ggplot(HMM_Final, aes(x = str_wrap(ecosystem_type, width = 30), y = NumberOfMAGs, fill = taxonomy)) +\n",
    "  geom_bar(stat = \"identity\")\n",
    "#p\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
