{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9d674811-f1f2-4e5e-b287-de699c4991b1",
   "metadata": {},
   "source": [
    "# Note3_關於proteo的MAGs分析"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea0cfcd8-be71-40c5-8050-41a25c748d1a",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 第一部分: 含有bitscore的篩選分析"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd805917-cd27-4a37-a6e3-3519ee3ca699",
   "metadata": {},
   "source": [
    "### 1.1 使用 positive control 來了解 hmm profiles 的分析情形 (e-value、coverage、bit-score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ade9e632-28f8-4809-91e8-29da6eb08f02",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>P_edc13525_I_</th>\n",
       "      <th>P_edc13530</th>\n",
       "      <th>P_edc13535</th>\n",
       "      <th>P_edc13545</th>\n",
       "      <th>P_edc13550</th>\n",
       "      <th>P_edc13555</th>\n",
       "      <th>P_edc13560</th>\n",
       "      <th>P_edc13565</th>\n",
       "      <th>P_edc13570_I_</th>\n",
       "      <th>P_edc13575</th>\n",
       "      <th>P_edc13580_I_</th>\n",
       "      <th>P_edc13585</th>\n",
       "      <th>P_edc13590</th>\n",
       "      <th>P_edc13595</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P_Altererythrobacter_estronivo_MH-B5_1_edc_bit_score</th>\n",
       "      <td>622.1</td>\n",
       "      <td>216.7</td>\n",
       "      <td>664.5</td>\n",
       "      <td>387.9</td>\n",
       "      <td>672.7</td>\n",
       "      <td>644.7</td>\n",
       "      <td>555.2</td>\n",
       "      <td>512.9</td>\n",
       "      <td>501.1</td>\n",
       "      <td>175.4</td>\n",
       "      <td>2001.2</td>\n",
       "      <td>714.2</td>\n",
       "      <td>213.2</td>\n",
       "      <td>214.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_Novosphingobium_sp_ES21_edc_bit_score</th>\n",
       "      <td>613.9</td>\n",
       "      <td>207.5</td>\n",
       "      <td>663.5</td>\n",
       "      <td>373.5</td>\n",
       "      <td>657.8</td>\n",
       "      <td>616.4</td>\n",
       "      <td>547.0</td>\n",
       "      <td>459.4</td>\n",
       "      <td>477.4</td>\n",
       "      <td>178.0</td>\n",
       "      <td>1968.2</td>\n",
       "      <td>706.2</td>\n",
       "      <td>185.3</td>\n",
       "      <td>202.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_Novosphingobium_tardaugens_NBRC16725_edc_bit_score</th>\n",
       "      <td>626.4</td>\n",
       "      <td>209.7</td>\n",
       "      <td>680.8</td>\n",
       "      <td>394.8</td>\n",
       "      <td>671.6</td>\n",
       "      <td>648.2</td>\n",
       "      <td>556.8</td>\n",
       "      <td>513.9</td>\n",
       "      <td>499.1</td>\n",
       "      <td>191.8</td>\n",
       "      <td>1997.6</td>\n",
       "      <td>717.6</td>\n",
       "      <td>215.0</td>\n",
       "      <td>221.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_Sphingobium_estronivorans_AXB_edc_bit_score</th>\n",
       "      <td>614.0</td>\n",
       "      <td>216.0</td>\n",
       "      <td>652.0</td>\n",
       "      <td>379.4</td>\n",
       "      <td>656.3</td>\n",
       "      <td>631.3</td>\n",
       "      <td>556.3</td>\n",
       "      <td>510.7</td>\n",
       "      <td>508.9</td>\n",
       "      <td>177.3</td>\n",
       "      <td>1971.9</td>\n",
       "      <td>720.8</td>\n",
       "      <td>218.0</td>\n",
       "      <td>194.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_Sphingomonas_sp_KC8_edc_bit_score</th>\n",
       "      <td>609.7</td>\n",
       "      <td>195.2</td>\n",
       "      <td>672.4</td>\n",
       "      <td>370.1</td>\n",
       "      <td>645.2</td>\n",
       "      <td>613.0</td>\n",
       "      <td>519.0</td>\n",
       "      <td>527.2</td>\n",
       "      <td>491.1</td>\n",
       "      <td>183.0</td>\n",
       "      <td>1937.0</td>\n",
       "      <td>702.1</td>\n",
       "      <td>222.9</td>\n",
       "      <td>204.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Min_BitScore</th>\n",
       "      <td>609.7</td>\n",
       "      <td>195.2</td>\n",
       "      <td>652.0</td>\n",
       "      <td>370.1</td>\n",
       "      <td>645.2</td>\n",
       "      <td>613.0</td>\n",
       "      <td>519.0</td>\n",
       "      <td>459.4</td>\n",
       "      <td>477.4</td>\n",
       "      <td>175.4</td>\n",
       "      <td>1937.0</td>\n",
       "      <td>702.1</td>\n",
       "      <td>185.3</td>\n",
       "      <td>194.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    P_edc13525_I_  P_edc13530  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_bit_...          622.1       216.7   \n",
       "P_Novosphingobium_sp_ES21_edc_bit_score                     613.9       207.5   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_bit_...          626.4       209.7   \n",
       "P_Sphingobium_estronivorans_AXB_edc_bit_score               614.0       216.0   \n",
       "P_Sphingomonas_sp_KC8_edc_bit_score                         609.7       195.2   \n",
       "Min_BitScore                                                609.7       195.2   \n",
       "\n",
       "                                                    P_edc13535  P_edc13545  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_bit_...       664.5       387.9   \n",
       "P_Novosphingobium_sp_ES21_edc_bit_score                  663.5       373.5   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_bit_...       680.8       394.8   \n",
       "P_Sphingobium_estronivorans_AXB_edc_bit_score            652.0       379.4   \n",
       "P_Sphingomonas_sp_KC8_edc_bit_score                      672.4       370.1   \n",
       "Min_BitScore                                             652.0       370.1   \n",
       "\n",
       "                                                    P_edc13550  P_edc13555  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_bit_...       672.7       644.7   \n",
       "P_Novosphingobium_sp_ES21_edc_bit_score                  657.8       616.4   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_bit_...       671.6       648.2   \n",
       "P_Sphingobium_estronivorans_AXB_edc_bit_score            656.3       631.3   \n",
       "P_Sphingomonas_sp_KC8_edc_bit_score                      645.2       613.0   \n",
       "Min_BitScore                                             645.2       613.0   \n",
       "\n",
       "                                                    P_edc13560  P_edc13565  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_bit_...       555.2       512.9   \n",
       "P_Novosphingobium_sp_ES21_edc_bit_score                  547.0       459.4   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_bit_...       556.8       513.9   \n",
       "P_Sphingobium_estronivorans_AXB_edc_bit_score            556.3       510.7   \n",
       "P_Sphingomonas_sp_KC8_edc_bit_score                      519.0       527.2   \n",
       "Min_BitScore                                             519.0       459.4   \n",
       "\n",
       "                                                    P_edc13570_I_  P_edc13575  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_bit_...          501.1       175.4   \n",
       "P_Novosphingobium_sp_ES21_edc_bit_score                     477.4       178.0   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_bit_...          499.1       191.8   \n",
       "P_Sphingobium_estronivorans_AXB_edc_bit_score               508.9       177.3   \n",
       "P_Sphingomonas_sp_KC8_edc_bit_score                         491.1       183.0   \n",
       "Min_BitScore                                                477.4       175.4   \n",
       "\n",
       "                                                    P_edc13580_I_  P_edc13585  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_bit_...         2001.2       714.2   \n",
       "P_Novosphingobium_sp_ES21_edc_bit_score                    1968.2       706.2   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_bit_...         1997.6       717.6   \n",
       "P_Sphingobium_estronivorans_AXB_edc_bit_score              1971.9       720.8   \n",
       "P_Sphingomonas_sp_KC8_edc_bit_score                        1937.0       702.1   \n",
       "Min_BitScore                                               1937.0       702.1   \n",
       "\n",
       "                                                    P_edc13590  P_edc13595  \n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_bit_...       213.2       214.4  \n",
       "P_Novosphingobium_sp_ES21_edc_bit_score                  185.3       202.9  \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_bit_...       215.0       221.8  \n",
       "P_Sphingobium_estronivorans_AXB_edc_bit_score            218.0       194.7  \n",
       "P_Sphingomonas_sp_KC8_edc_bit_score                      222.9       204.2  \n",
       "Min_BitScore                                             185.3       194.7  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>P_edc13525_I_</th>\n",
       "      <th>P_edc13530</th>\n",
       "      <th>P_edc13535</th>\n",
       "      <th>P_edc13545</th>\n",
       "      <th>P_edc13550</th>\n",
       "      <th>P_edc13555</th>\n",
       "      <th>P_edc13560</th>\n",
       "      <th>P_edc13565</th>\n",
       "      <th>P_edc13570_I_</th>\n",
       "      <th>P_edc13575</th>\n",
       "      <th>P_edc13580_I_</th>\n",
       "      <th>P_edc13585</th>\n",
       "      <th>P_edc13590</th>\n",
       "      <th>P_edc13595</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P_Altererythrobacter_estronivo_MH-B5_1_edc_Evalue</th>\n",
       "      <td>3.900000e-190</td>\n",
       "      <td>8.900000e-68</td>\n",
       "      <td>2.500000e-203</td>\n",
       "      <td>1.300000e-119</td>\n",
       "      <td>1.400000e-205</td>\n",
       "      <td>6.100000e-197</td>\n",
       "      <td>2.900000e-170</td>\n",
       "      <td>4.700000e-157</td>\n",
       "      <td>8.600000e-154</td>\n",
       "      <td>4.000000e-55</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.800000e-218</td>\n",
       "      <td>1.100000e-66</td>\n",
       "      <td>4.300000e-67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_Novosphingobium_sp_ES21_edc_Evalue</th>\n",
       "      <td>1.300000e-187</td>\n",
       "      <td>6.200000e-65</td>\n",
       "      <td>4.900000e-203</td>\n",
       "      <td>3.200000e-115</td>\n",
       "      <td>4.500000e-201</td>\n",
       "      <td>2.400000e-188</td>\n",
       "      <td>9.200000e-168</td>\n",
       "      <td>8.100000e-141</td>\n",
       "      <td>1.400000e-146</td>\n",
       "      <td>6.200000e-56</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.500000e-216</td>\n",
       "      <td>4.200000e-58</td>\n",
       "      <td>1.500000e-63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_Novosphingobium_tardaugens_NBRC16725_edc_Evalue</th>\n",
       "      <td>1.900000e-191</td>\n",
       "      <td>1.100000e-65</td>\n",
       "      <td>2.600000e-208</td>\n",
       "      <td>9.000000e-122</td>\n",
       "      <td>2.800000e-205</td>\n",
       "      <td>4.700000e-198</td>\n",
       "      <td>8.500000e-171</td>\n",
       "      <td>2.100000e-157</td>\n",
       "      <td>3.100000e-153</td>\n",
       "      <td>3.300000e-60</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.400000e-219</td>\n",
       "      <td>2.800000e-67</td>\n",
       "      <td>2.200000e-69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_Sphingobium_estronivorans_AXB_edc_Evalue</th>\n",
       "      <td>1.100000e-187</td>\n",
       "      <td>1.500000e-67</td>\n",
       "      <td>1.600000e-199</td>\n",
       "      <td>5.200000e-117</td>\n",
       "      <td>1.300000e-200</td>\n",
       "      <td>6.800000e-193</td>\n",
       "      <td>1.300000e-170</td>\n",
       "      <td>2.200000e-156</td>\n",
       "      <td>3.500000e-156</td>\n",
       "      <td>1.100000e-55</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.800000e-220</td>\n",
       "      <td>3.800000e-68</td>\n",
       "      <td>5.300000e-61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_Sphingomonas_sp_KC8_edc_Evalue</th>\n",
       "      <td>2.400000e-186</td>\n",
       "      <td>3.900000e-61</td>\n",
       "      <td>1.000000e-205</td>\n",
       "      <td>3.800000e-114</td>\n",
       "      <td>3.300000e-197</td>\n",
       "      <td>2.600000e-187</td>\n",
       "      <td>3.300000e-159</td>\n",
       "      <td>2.200000e-161</td>\n",
       "      <td>9.900000e-151</td>\n",
       "      <td>2.000000e-57</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.400000e-214</td>\n",
       "      <td>1.200000e-69</td>\n",
       "      <td>6.700000e-64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Max_Evalue</th>\n",
       "      <td>2.400000e-186</td>\n",
       "      <td>3.900000e-61</td>\n",
       "      <td>1.600000e-199</td>\n",
       "      <td>3.800000e-114</td>\n",
       "      <td>3.300000e-197</td>\n",
       "      <td>2.600000e-187</td>\n",
       "      <td>3.300000e-159</td>\n",
       "      <td>8.100000e-141</td>\n",
       "      <td>1.400000e-146</td>\n",
       "      <td>4.000000e-55</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.400000e-214</td>\n",
       "      <td>4.200000e-58</td>\n",
       "      <td>5.300000e-61</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   P_edc13525_I_  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_Evalue  3.900000e-190   \n",
       "P_Novosphingobium_sp_ES21_edc_Evalue               1.300000e-187   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_Evalue  1.900000e-191   \n",
       "P_Sphingobium_estronivorans_AXB_edc_Evalue         1.100000e-187   \n",
       "P_Sphingomonas_sp_KC8_edc_Evalue                   2.400000e-186   \n",
       "Max_Evalue                                         2.400000e-186   \n",
       "\n",
       "                                                     P_edc13530  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_Evalue  8.900000e-68   \n",
       "P_Novosphingobium_sp_ES21_edc_Evalue               6.200000e-65   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_Evalue  1.100000e-65   \n",
       "P_Sphingobium_estronivorans_AXB_edc_Evalue         1.500000e-67   \n",
       "P_Sphingomonas_sp_KC8_edc_Evalue                   3.900000e-61   \n",
       "Max_Evalue                                         3.900000e-61   \n",
       "\n",
       "                                                      P_edc13535  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_Evalue  2.500000e-203   \n",
       "P_Novosphingobium_sp_ES21_edc_Evalue               4.900000e-203   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_Evalue  2.600000e-208   \n",
       "P_Sphingobium_estronivorans_AXB_edc_Evalue         1.600000e-199   \n",
       "P_Sphingomonas_sp_KC8_edc_Evalue                   1.000000e-205   \n",
       "Max_Evalue                                         1.600000e-199   \n",
       "\n",
       "                                                      P_edc13545  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_Evalue  1.300000e-119   \n",
       "P_Novosphingobium_sp_ES21_edc_Evalue               3.200000e-115   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_Evalue  9.000000e-122   \n",
       "P_Sphingobium_estronivorans_AXB_edc_Evalue         5.200000e-117   \n",
       "P_Sphingomonas_sp_KC8_edc_Evalue                   3.800000e-114   \n",
       "Max_Evalue                                         3.800000e-114   \n",
       "\n",
       "                                                      P_edc13550  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_Evalue  1.400000e-205   \n",
       "P_Novosphingobium_sp_ES21_edc_Evalue               4.500000e-201   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_Evalue  2.800000e-205   \n",
       "P_Sphingobium_estronivorans_AXB_edc_Evalue         1.300000e-200   \n",
       "P_Sphingomonas_sp_KC8_edc_Evalue                   3.300000e-197   \n",
       "Max_Evalue                                         3.300000e-197   \n",
       "\n",
       "                                                      P_edc13555  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_Evalue  6.100000e-197   \n",
       "P_Novosphingobium_sp_ES21_edc_Evalue               2.400000e-188   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_Evalue  4.700000e-198   \n",
       "P_Sphingobium_estronivorans_AXB_edc_Evalue         6.800000e-193   \n",
       "P_Sphingomonas_sp_KC8_edc_Evalue                   2.600000e-187   \n",
       "Max_Evalue                                         2.600000e-187   \n",
       "\n",
       "                                                      P_edc13560  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_Evalue  2.900000e-170   \n",
       "P_Novosphingobium_sp_ES21_edc_Evalue               9.200000e-168   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_Evalue  8.500000e-171   \n",
       "P_Sphingobium_estronivorans_AXB_edc_Evalue         1.300000e-170   \n",
       "P_Sphingomonas_sp_KC8_edc_Evalue                   3.300000e-159   \n",
       "Max_Evalue                                         3.300000e-159   \n",
       "\n",
       "                                                      P_edc13565  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_Evalue  4.700000e-157   \n",
       "P_Novosphingobium_sp_ES21_edc_Evalue               8.100000e-141   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_Evalue  2.100000e-157   \n",
       "P_Sphingobium_estronivorans_AXB_edc_Evalue         2.200000e-156   \n",
       "P_Sphingomonas_sp_KC8_edc_Evalue                   2.200000e-161   \n",
       "Max_Evalue                                         8.100000e-141   \n",
       "\n",
       "                                                   P_edc13570_I_  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_Evalue  8.600000e-154   \n",
       "P_Novosphingobium_sp_ES21_edc_Evalue               1.400000e-146   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_Evalue  3.100000e-153   \n",
       "P_Sphingobium_estronivorans_AXB_edc_Evalue         3.500000e-156   \n",
       "P_Sphingomonas_sp_KC8_edc_Evalue                   9.900000e-151   \n",
       "Max_Evalue                                         1.400000e-146   \n",
       "\n",
       "                                                     P_edc13575  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_Evalue  4.000000e-55   \n",
       "P_Novosphingobium_sp_ES21_edc_Evalue               6.200000e-56   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_Evalue  3.300000e-60   \n",
       "P_Sphingobium_estronivorans_AXB_edc_Evalue         1.100000e-55   \n",
       "P_Sphingomonas_sp_KC8_edc_Evalue                   2.000000e-57   \n",
       "Max_Evalue                                         4.000000e-55   \n",
       "\n",
       "                                                   P_edc13580_I_  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_Evalue            0.0   \n",
       "P_Novosphingobium_sp_ES21_edc_Evalue                         0.0   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_Evalue            0.0   \n",
       "P_Sphingobium_estronivorans_AXB_edc_Evalue                   0.0   \n",
       "P_Sphingomonas_sp_KC8_edc_Evalue                             0.0   \n",
       "Max_Evalue                                                   0.0   \n",
       "\n",
       "                                                      P_edc13585  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_Evalue  2.800000e-218   \n",
       "P_Novosphingobium_sp_ES21_edc_Evalue               7.500000e-216   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_Evalue  2.400000e-219   \n",
       "P_Sphingobium_estronivorans_AXB_edc_Evalue         2.800000e-220   \n",
       "P_Sphingomonas_sp_KC8_edc_Evalue                   1.400000e-214   \n",
       "Max_Evalue                                         1.400000e-214   \n",
       "\n",
       "                                                     P_edc13590    P_edc13595  \n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_Evalue  1.100000e-66  4.300000e-67  \n",
       "P_Novosphingobium_sp_ES21_edc_Evalue               4.200000e-58  1.500000e-63  \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_Evalue  2.800000e-67  2.200000e-69  \n",
       "P_Sphingobium_estronivorans_AXB_edc_Evalue         3.800000e-68  5.300000e-61  \n",
       "P_Sphingomonas_sp_KC8_edc_Evalue                   1.200000e-69  6.700000e-64  \n",
       "Max_Evalue                                         4.200000e-58  5.300000e-61  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>P_edc13525_I_</th>\n",
       "      <th>P_edc13530</th>\n",
       "      <th>P_edc13535</th>\n",
       "      <th>P_edc13545</th>\n",
       "      <th>P_edc13550</th>\n",
       "      <th>P_edc13555</th>\n",
       "      <th>P_edc13560</th>\n",
       "      <th>P_edc13565</th>\n",
       "      <th>P_edc13570_I_</th>\n",
       "      <th>P_edc13575</th>\n",
       "      <th>P_edc13580_I_</th>\n",
       "      <th>P_edc13585</th>\n",
       "      <th>P_edc13590</th>\n",
       "      <th>P_edc13595</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P_Altererythrobacter_estronivo_MH-B5_1_edc_coverage</th>\n",
       "      <td>0.992424</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.942708</td>\n",
       "      <td>0.854478</td>\n",
       "      <td>0.976378</td>\n",
       "      <td>0.912393</td>\n",
       "      <td>0.973244</td>\n",
       "      <td>0.976064</td>\n",
       "      <td>0.993548</td>\n",
       "      <td>0.955556</td>\n",
       "      <td>0.995660</td>\n",
       "      <td>0.949367</td>\n",
       "      <td>0.940120</td>\n",
       "      <td>0.935897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_Novosphingobium_sp_ES21_edc_coverage</th>\n",
       "      <td>0.992424</td>\n",
       "      <td>0.993007</td>\n",
       "      <td>0.935401</td>\n",
       "      <td>0.854478</td>\n",
       "      <td>0.966234</td>\n",
       "      <td>0.916488</td>\n",
       "      <td>0.950820</td>\n",
       "      <td>0.975806</td>\n",
       "      <td>0.981013</td>\n",
       "      <td>0.955556</td>\n",
       "      <td>0.993929</td>\n",
       "      <td>0.949367</td>\n",
       "      <td>0.905882</td>\n",
       "      <td>0.893750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_Novosphingobium_tardaugens_NBRC16725_edc_coverage</th>\n",
       "      <td>0.992424</td>\n",
       "      <td>0.953020</td>\n",
       "      <td>0.947644</td>\n",
       "      <td>0.854478</td>\n",
       "      <td>0.976378</td>\n",
       "      <td>0.922246</td>\n",
       "      <td>0.973244</td>\n",
       "      <td>0.978552</td>\n",
       "      <td>0.993548</td>\n",
       "      <td>0.955556</td>\n",
       "      <td>0.995656</td>\n",
       "      <td>0.949367</td>\n",
       "      <td>0.934132</td>\n",
       "      <td>0.928105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_Sphingobium_estronivorans_AXB_edc_coverage</th>\n",
       "      <td>0.980100</td>\n",
       "      <td>0.946667</td>\n",
       "      <td>0.942708</td>\n",
       "      <td>0.854962</td>\n",
       "      <td>0.976378</td>\n",
       "      <td>0.922747</td>\n",
       "      <td>0.963455</td>\n",
       "      <td>0.981530</td>\n",
       "      <td>0.971698</td>\n",
       "      <td>0.962687</td>\n",
       "      <td>0.984563</td>\n",
       "      <td>0.949367</td>\n",
       "      <td>0.930233</td>\n",
       "      <td>0.915584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_Sphingomonas_sp_KC8_edc_coverage</th>\n",
       "      <td>0.987469</td>\n",
       "      <td>0.986111</td>\n",
       "      <td>0.940260</td>\n",
       "      <td>0.863636</td>\n",
       "      <td>0.976378</td>\n",
       "      <td>0.915948</td>\n",
       "      <td>0.976510</td>\n",
       "      <td>0.981627</td>\n",
       "      <td>0.910180</td>\n",
       "      <td>0.956204</td>\n",
       "      <td>0.983677</td>\n",
       "      <td>0.949367</td>\n",
       "      <td>0.936047</td>\n",
       "      <td>0.905660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Min_coverage</th>\n",
       "      <td>0.980100</td>\n",
       "      <td>0.946667</td>\n",
       "      <td>0.935401</td>\n",
       "      <td>0.854478</td>\n",
       "      <td>0.966234</td>\n",
       "      <td>0.912393</td>\n",
       "      <td>0.950820</td>\n",
       "      <td>0.975806</td>\n",
       "      <td>0.910180</td>\n",
       "      <td>0.955556</td>\n",
       "      <td>0.983677</td>\n",
       "      <td>0.949367</td>\n",
       "      <td>0.905882</td>\n",
       "      <td>0.893750</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    P_edc13525_I_  P_edc13530  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_cove...       0.992424    1.000000   \n",
       "P_Novosphingobium_sp_ES21_edc_coverage                   0.992424    0.993007   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_cove...       0.992424    0.953020   \n",
       "P_Sphingobium_estronivorans_AXB_edc_coverage             0.980100    0.946667   \n",
       "P_Sphingomonas_sp_KC8_edc_coverage                       0.987469    0.986111   \n",
       "Min_coverage                                             0.980100    0.946667   \n",
       "\n",
       "                                                    P_edc13535  P_edc13545  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_cove...    0.942708    0.854478   \n",
       "P_Novosphingobium_sp_ES21_edc_coverage                0.935401    0.854478   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_cove...    0.947644    0.854478   \n",
       "P_Sphingobium_estronivorans_AXB_edc_coverage          0.942708    0.854962   \n",
       "P_Sphingomonas_sp_KC8_edc_coverage                    0.940260    0.863636   \n",
       "Min_coverage                                          0.935401    0.854478   \n",
       "\n",
       "                                                    P_edc13550  P_edc13555  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_cove...    0.976378    0.912393   \n",
       "P_Novosphingobium_sp_ES21_edc_coverage                0.966234    0.916488   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_cove...    0.976378    0.922246   \n",
       "P_Sphingobium_estronivorans_AXB_edc_coverage          0.976378    0.922747   \n",
       "P_Sphingomonas_sp_KC8_edc_coverage                    0.976378    0.915948   \n",
       "Min_coverage                                          0.966234    0.912393   \n",
       "\n",
       "                                                    P_edc13560  P_edc13565  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_cove...    0.973244    0.976064   \n",
       "P_Novosphingobium_sp_ES21_edc_coverage                0.950820    0.975806   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_cove...    0.973244    0.978552   \n",
       "P_Sphingobium_estronivorans_AXB_edc_coverage          0.963455    0.981530   \n",
       "P_Sphingomonas_sp_KC8_edc_coverage                    0.976510    0.981627   \n",
       "Min_coverage                                          0.950820    0.975806   \n",
       "\n",
       "                                                    P_edc13570_I_  P_edc13575  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_cove...       0.993548    0.955556   \n",
       "P_Novosphingobium_sp_ES21_edc_coverage                   0.981013    0.955556   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_cove...       0.993548    0.955556   \n",
       "P_Sphingobium_estronivorans_AXB_edc_coverage             0.971698    0.962687   \n",
       "P_Sphingomonas_sp_KC8_edc_coverage                       0.910180    0.956204   \n",
       "Min_coverage                                             0.910180    0.955556   \n",
       "\n",
       "                                                    P_edc13580_I_  P_edc13585  \\\n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_cove...       0.995660    0.949367   \n",
       "P_Novosphingobium_sp_ES21_edc_coverage                   0.993929    0.949367   \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_cove...       0.995656    0.949367   \n",
       "P_Sphingobium_estronivorans_AXB_edc_coverage             0.984563    0.949367   \n",
       "P_Sphingomonas_sp_KC8_edc_coverage                       0.983677    0.949367   \n",
       "Min_coverage                                             0.983677    0.949367   \n",
       "\n",
       "                                                    P_edc13590  P_edc13595  \n",
       "P_Altererythrobacter_estronivo_MH-B5_1_edc_cove...    0.940120    0.935897  \n",
       "P_Novosphingobium_sp_ES21_edc_coverage                0.905882    0.893750  \n",
       "P_Novosphingobium_tardaugens_NBRC16725_edc_cove...    0.934132    0.928105  \n",
       "P_Sphingobium_estronivorans_AXB_edc_coverage          0.930233    0.915584  \n",
       "P_Sphingomonas_sp_KC8_edc_coverage                    0.936047    0.905660  \n",
       "Min_coverage                                          0.905882    0.893750  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n"
     ]
    }
   ],
   "source": [
    "# HMM actino positive control bit-score threshold\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "#------------------------------------------------------------------------\n",
    "# create a function can get the datafarme of each control's best bit-score, e-vale and coverage dataframes\n",
    "# hmm domtblout name should be ...aed_I_... NOT ..._I_aed_....\n",
    "def positive_control_df(control_dir, control_names, BitScore_df, Evalue_df, Coverage_df):\n",
    "    # Create an empty dictionary to store the best bit-scores; best e-vale; best coverage for each hmmsearch\n",
    "    best_bit_scores = {}\n",
    "    evalue_scores = {}\n",
    "    coverage_scores = {}\n",
    "\n",
    "    # Loop over the hmmsearches and parse the corresponding \"domtblout\" file\n",
    "    for control_name in control_names:\n",
    "        # Load the \"domtblout\" file into a pandas DataFrame\n",
    "        file_path = os.path.join(control_dir, control_name + \".domtblout\")\n",
    "        try:\n",
    "            df = pd.read_csv(file_path, comment=\"#\", sep='\\s+', header=None)\n",
    "        except pd.errors.EmptyDataError:\n",
    "            df = pd.DataFrame()\n",
    "\n",
    "        if df.empty:\n",
    "            pattern = r'P_.*_edc'\n",
    "            control_name = re.sub(pattern, 'P_edc', control_name)        \n",
    "            best_bit_scores[control_name] = None\n",
    "            evalue_scores[control_name] = None\n",
    "            coverage_scores[control_name] = None\n",
    "        else:        \n",
    "            # Assign column names to the DataFrame\n",
    "            df.columns = [\"target_name\", \"accession\", \"tlen\", \"query_name\", \"accession2\", \"qlen\", \"E-value\", \"score\", \"bias\",\n",
    "                          \"num_domains_index\", \"num_domains_total\", \"c-Evalue\", \"i-Evalue\", \"score2\", \"bias2\", \"hmm_from\", \"hmm_to\", \"ali_from\", \"ali_to\",\n",
    "                          \"env_from\", \"env_to\", \"acc\", \"description\"]\n",
    "            # Calculate the coverage for each hit\n",
    "            df[\"coverage\"] = (df[\"ali_to\"] - df[\"ali_from\"] + 1) / df[\"tlen\"]\n",
    "\n",
    "            # Filter the DataFrame by E-value and coverage, and sort by bit-score\n",
    "            significant_hits = df[(df[\"E-value\"] < 0.001) & (df[\"coverage\"] > 0.50)].sort_values(by=\"score\", ascending=False)\n",
    "\n",
    "            # replace strains name to aed..\n",
    "            pattern = r'P_.*_edc'\n",
    "            control_name = re.sub(pattern, 'P_edc', control_name)\n",
    "\n",
    "            # Extract the best bit-score and store it in the dictionary\n",
    "            if not significant_hits.empty:\n",
    "                best_bit_score = significant_hits.iloc[0][\"score\"]\n",
    "                best_bit_scores[control_name] = best_bit_score\n",
    "                evalue_score = significant_hits.iloc[0][\"E-value\"]\n",
    "                evalue_scores[control_name] = evalue_score\n",
    "                coverage_score = significant_hits.iloc[0][\"coverage\"]\n",
    "                coverage_scores[control_name] = coverage_score\n",
    "            else:\n",
    "                best_bit_scores[control_name] = None\n",
    "                evalue_scores[control_name] = None\n",
    "                coverage_scores[control_name] = None\n",
    "\n",
    "    # create the index names for three df\n",
    "    # Get the strain name\n",
    "    pattern = r'P_.*_edc'\n",
    "    StrainName = re.findall(pattern, control_names[0])\n",
    "    \n",
    "    # create index of bit score\n",
    "    bitscore_name = StrainName[0] + '_bit_score'\n",
    "    BitScore_Name = [bitscore_name]\n",
    "\n",
    "    # create index of Evalue\n",
    "    evalue_name = StrainName[0] + '_Evalue'\n",
    "    Evalue_Name = [evalue_name]\n",
    "    \n",
    "    # create index of coverage\n",
    "    coverage_name = StrainName[0] + '_coverage'\n",
    "    Coverage_Name = [coverage_name]\n",
    "    \n",
    "    # create a dataframe of bit score\n",
    "    BitScore_df = pd.DataFrame(best_bit_scores, index=BitScore_Name)\n",
    "\n",
    "    # create a dataframe of e-value\n",
    "    Evalue_df = pd.DataFrame(evalue_scores, index=Evalue_Name)\n",
    "\n",
    "    # create a dataframe of coverage\n",
    "    Coverage_df = pd.DataFrame(coverage_scores, index=Coverage_Name)\n",
    "    \n",
    "    return BitScore_df, Evalue_df, Coverage_df\n",
    "\n",
    "\n",
    "#---------------------------------------------------------------\n",
    "# Altererythrobacter_estronivo_MH-B5_1\n",
    "# Define the directory that contains the \"domtblout\" files\n",
    "control_dir = \"../data/raw/Proteo_HMM_PositiveControl/Altererythrobacter_estronivo_MH-B5_1/domtblout/\"\n",
    "\n",
    "# Define the names of the control_names (without the file extension); delet 13540\n",
    "control_names = ['P_Altererythrobacter_estronivo_MH-B5_1_edc13525_I_', 'P_Altererythrobacter_estronivo_MH-B5_1_edc13530', 'P_Altererythrobacter_estronivo_MH-B5_1_edc13535', 'P_Altererythrobacter_estronivo_MH-B5_1_edc13545', 'P_Altererythrobacter_estronivo_MH-B5_1_edc13550', 'P_Altererythrobacter_estronivo_MH-B5_1_edc13555', 'P_Altererythrobacter_estronivo_MH-B5_1_edc13560', 'P_Altererythrobacter_estronivo_MH-B5_1_edc13565', 'P_Altererythrobacter_estronivo_MH-B5_1_edc13570_I_', 'P_Altererythrobacter_estronivo_MH-B5_1_edc13575', 'P_Altererythrobacter_estronivo_MH-B5_1_edc13580_I_', 'P_Altererythrobacter_estronivo_MH-B5_1_edc13585', 'P_Altererythrobacter_estronivo_MH-B5_1_edc13590', 'P_Altererythrobacter_estronivo_MH-B5_1_edc13595']\n",
    "\n",
    "# Use positive_control_df function to get the df (control_dir, control_names, BitScore_df, Evalue_df, Coverage_df)\n",
    "B5_BitScore = pd.DataFrame()\n",
    "B5_Evalue = pd.DataFrame()\n",
    "B5_coverage = pd.DataFrame()\n",
    "\n",
    "B5_BitScore, B5_Evalue, B5_coverage = positive_control_df(control_dir, control_names, B5_BitScore, B5_Evalue, B5_coverage)\n",
    "\n",
    "\n",
    "#---------------------------------------------------------------\n",
    "# Novosphingobium_sp_ES21\n",
    "# Define the directory that contains the \"domtblout\" files\n",
    "control_dir = \"../data/raw/Proteo_HMM_PositiveControl/Novosphingobium_sp_ES21/domtblout/\"\n",
    "\n",
    "# Define the names of the control_names (without the file extension); delet 13540\n",
    "control_names = ['P_Novosphingobium_sp_ES21_edc13525_I_', 'P_Novosphingobium_sp_ES21_edc13530', 'P_Novosphingobium_sp_ES21_edc13535', 'P_Novosphingobium_sp_ES21_edc13545', 'P_Novosphingobium_sp_ES21_edc13550', 'P_Novosphingobium_sp_ES21_edc13555', 'P_Novosphingobium_sp_ES21_edc13560', 'P_Novosphingobium_sp_ES21_edc13565', 'P_Novosphingobium_sp_ES21_edc13570_I_', 'P_Novosphingobium_sp_ES21_edc13575', 'P_Novosphingobium_sp_ES21_edc13580_I_', 'P_Novosphingobium_sp_ES21_edc13585', 'P_Novosphingobium_sp_ES21_edc13590', 'P_Novosphingobium_sp_ES21_edc13595']\n",
    "\n",
    "    # Use positive_control_df function to get the df (control_dir, control_names, BitScore_df, Evalue_df, Coverage_df)\n",
    "ES21_BitScore = pd.DataFrame()\n",
    "ES21_Evalue = pd.DataFrame()\n",
    "ES21_coverage = pd.DataFrame()\n",
    "\n",
    "ES21_BitScore, ES21_Evalue, ES21_coverage = positive_control_df(control_dir, control_names, ES21_BitScore, ES21_Evalue, ES21_coverage)\n",
    "\n",
    "\n",
    "#---------------------------------------------------------------\n",
    "# Novosphingobium_tardaugens_NBRC16725\n",
    "# Define the directory that contains the \"domtblout\" files\n",
    "control_dir = \"../data/raw/Proteo_HMM_PositiveControl/Novosphingobium_tardaugens_NBRC16725/domtblout/\"\n",
    "\n",
    "# Define the names of the control_names (without the file extension); delet 13540\n",
    "control_names = ['P_Novosphingobium_tardaugens_NBRC16725_edc13525_I_', 'P_Novosphingobium_tardaugens_NBRC16725_edc13530', 'P_Novosphingobium_tardaugens_NBRC16725_edc13535', 'P_Novosphingobium_tardaugens_NBRC16725_edc13545', 'P_Novosphingobium_tardaugens_NBRC16725_edc13550', 'P_Novosphingobium_tardaugens_NBRC16725_edc13555', 'P_Novosphingobium_tardaugens_NBRC16725_edc13560', 'P_Novosphingobium_tardaugens_NBRC16725_edc13565', 'P_Novosphingobium_tardaugens_NBRC16725_edc13570_I_', 'P_Novosphingobium_tardaugens_NBRC16725_edc13575', 'P_Novosphingobium_tardaugens_NBRC16725_edc13580_I_', 'P_Novosphingobium_tardaugens_NBRC16725_edc13585', 'P_Novosphingobium_tardaugens_NBRC16725_edc13590', 'P_Novosphingobium_tardaugens_NBRC16725_edc13595']\n",
    "\n",
    "# Use positive_control_df function to get the df (control_dir, control_names, BitScore_df, Evalue_df, Coverage_df)\n",
    "NBRC16725_BitScore = pd.DataFrame()\n",
    "NBRC16725_Evalue = pd.DataFrame()\n",
    "NBRC16725_coverage = pd.DataFrame()\n",
    "\n",
    "NBRC16725_BitScore, NBRC16725_Evalue, NBRC16725_coverage = positive_control_df(control_dir, control_names, NBRC16725_BitScore, NBRC16725_Evalue, NBRC16725_coverage)\n",
    "\n",
    "\n",
    "#---------------------------------------------------------------\n",
    "# Sphingobium_estronivorans_AXB\n",
    "# Define the directory that contains the \"domtblout\" files\n",
    "control_dir = \"../data/raw/Proteo_HMM_PositiveControl/Sphingobium_estronivorans_AXB/domtblout/\"\n",
    "\n",
    "# Define the names of the control_names (without the file extension); delet 13540\n",
    "control_names = ['P_Sphingobium_estronivorans_AXB_edc13525_I_', 'P_Sphingobium_estronivorans_AXB_edc13530', 'P_Sphingobium_estronivorans_AXB_edc13535', 'P_Sphingobium_estronivorans_AXB_edc13545', 'P_Sphingobium_estronivorans_AXB_edc13550', 'P_Sphingobium_estronivorans_AXB_edc13555', 'P_Sphingobium_estronivorans_AXB_edc13560', 'P_Sphingobium_estronivorans_AXB_edc13565', 'P_Sphingobium_estronivorans_AXB_edc13570_I_', 'P_Sphingobium_estronivorans_AXB_edc13575', 'P_Sphingobium_estronivorans_AXB_edc13580_I_', 'P_Sphingobium_estronivorans_AXB_edc13585', 'P_Sphingobium_estronivorans_AXB_edc13590', 'P_Sphingobium_estronivorans_AXB_edc13595']\n",
    "\n",
    "# Use positive_control_df function to get the df (control_dir, control_names, BitScore_df, Evalue_df, Coverage_df)\n",
    "AXB_BitScore = pd.DataFrame()\n",
    "AXB_Evalue = pd.DataFrame()\n",
    "AXB_coverage = pd.DataFrame()\n",
    "\n",
    "AXB_BitScore, AXB_Evalue, AXB_coverage = positive_control_df(control_dir, control_names, AXB_BitScore, AXB_Evalue, AXB_coverage)\n",
    "\n",
    "\n",
    "#---------------------------------------------------------------\n",
    "# Sphingomonas_sp_KC8\n",
    "# Define the directory that contains the \"domtblout\" files\n",
    "control_dir = \"../data/raw/Proteo_HMM_PositiveControl/Sphingomonas_sp_KC8/domtblout/\"\n",
    "\n",
    "# Define the names of the control_names (without the file extension); delet 13540\n",
    "control_names = ['P_Sphingomonas_sp_KC8_edc13525_I_', 'P_Sphingomonas_sp_KC8_edc13530', 'P_Sphingomonas_sp_KC8_edc13535', 'P_Sphingomonas_sp_KC8_edc13545', 'P_Sphingomonas_sp_KC8_edc13550', 'P_Sphingomonas_sp_KC8_edc13555', 'P_Sphingomonas_sp_KC8_edc13560', 'P_Sphingomonas_sp_KC8_edc13565', 'P_Sphingomonas_sp_KC8_edc13570_I_', 'P_Sphingomonas_sp_KC8_edc13575', 'P_Sphingomonas_sp_KC8_edc13580_I_', 'P_Sphingomonas_sp_KC8_edc13585', 'P_Sphingomonas_sp_KC8_edc13590', 'P_Sphingomonas_sp_KC8_edc13595']\n",
    "\n",
    "# Use positive_control_df function to get the df (control_dir, control_names, BitScore_df, Evalue_df, Coverage_df)\n",
    "KC8_BitScore = pd.DataFrame()\n",
    "KC8_Evalue = pd.DataFrame()\n",
    "KC8_coverage = pd.DataFrame()\n",
    "\n",
    "KC8_BitScore, KC8_Evalue, KC8_coverage = positive_control_df(control_dir, control_names, KC8_BitScore, KC8_Evalue, KC8_coverage)\n",
    "\n",
    "\n",
    "#-------------------------------------------------------------------------------------------------------\n",
    "# Merge tabel and create a df of max and min\n",
    "\n",
    "# Got the min bit score of the df\n",
    "# merge 5 best bit score df\n",
    "Proteo_PositiveBitScore_df = pd.concat([B5_BitScore, ES21_BitScore, NBRC16725_BitScore, AXB_BitScore, KC8_BitScore], axis=0)\n",
    "\n",
    "Proteo_MinBitScore_df = Proteo_PositiveBitScore_df.min()\n",
    "Proteo_MinBitScore_df = Proteo_MinBitScore_df.iloc[0:]\n",
    "Proteo_MinBitScore_df = Proteo_MinBitScore_df.to_frame()\n",
    "Proteo_MinBitScore_df.columns = ['Min_BitScore']\n",
    "\n",
    "# transpose the DataFrame\n",
    "Proteo_MinBitScore_df = Proteo_MinBitScore_df.T\n",
    "\n",
    "# merge all nad Min table\n",
    "Proteo_PositiveBitScore_df = pd.concat([Proteo_PositiveBitScore_df, Proteo_MinBitScore_df], axis=0)\n",
    "Proteo_PositiveBitScore_df.to_csv('../data/processed/Final/ForReader/ControlData/Proteo_edc/Proteo_PositiveBitScore.csv')\n",
    "\n",
    "\n",
    "# Got the max e-value of the df\n",
    "# merge 5 e-value df\n",
    "Proteo_PositiveEvalue_df = pd.concat([B5_Evalue, ES21_Evalue, NBRC16725_Evalue, AXB_Evalue, KC8_Evalue], axis=0)\n",
    "\n",
    "Proteo_MaxEvalue_df = Proteo_PositiveEvalue_df.max()\n",
    "Proteo_MaxEvalue_df = Proteo_MaxEvalue_df.iloc[0:]\n",
    "Proteo_MaxEvalue_df = Proteo_MaxEvalue_df.to_frame()\n",
    "Proteo_MaxEvalue_df.columns = ['Max_Evalue']\n",
    "\n",
    "# transpose the DataFrame\n",
    "Proteo_MaxEvalue_df = Proteo_MaxEvalue_df.T\n",
    "\n",
    "# merge all nad Min table\n",
    "Proteo_PositiveEvalue_df = pd.concat([Proteo_PositiveEvalue_df, Proteo_MaxEvalue_df], axis=0)\n",
    "Proteo_PositiveEvalue_df.to_csv('../data/processed/Final/ForReader/ControlData/Proteo_edc/Proteo_PositiveEvalue.csv')\n",
    "\n",
    "\n",
    "# Got the min coverage of the df\n",
    "# merge 5 e-value df\n",
    "Proteo_Positivecoverage_df = pd.concat([B5_coverage, ES21_coverage, NBRC16725_coverage, AXB_coverage, KC8_coverage], axis=0)\n",
    "\n",
    "Proteo_Mincoverage_df = Proteo_Positivecoverage_df.min()\n",
    "Proteo_Mincoverage_df = Proteo_Mincoverage_df.iloc[0:]\n",
    "Proteo_Mincoverage_df = Proteo_Mincoverage_df.to_frame()\n",
    "Proteo_Mincoverage_df.columns = ['Min_coverage']\n",
    "\n",
    "# transpose the DataFrame\n",
    "Proteo_Mincoverage_df = Proteo_Mincoverage_df.T\n",
    "\n",
    "# merge all nad Min table\n",
    "Proteo_Positivecoverage_df = pd.concat([Proteo_Positivecoverage_df, Proteo_Mincoverage_df], axis=0)\n",
    "Proteo_Positivecoverage_df.to_csv('../data/processed/Final/ForReader/ControlData/Proteo_edc/Proteo_PositiveCoverage.csv')\n",
    "\n",
    "\n",
    "# Display three table\n",
    "display(Proteo_PositiveBitScore_df)\n",
    "display(Proteo_PositiveEvalue_df)\n",
    "display(Proteo_Positivecoverage_df)\n",
    "print('Done')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20ec0f18-988e-407d-9773-2e02a29e7673",
   "metadata": {},
   "source": [
    "### 1.2 使用 negative control 來獲得合適的bit score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "51acb95a-3119-4fab-a5dc-7eefaf2a5add",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Criteria_Bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P_edc13525_I_</th>\n",
       "      <td>127.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_edc13530</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_edc13535</th>\n",
       "      <td>397.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_edc13545</th>\n",
       "      <td>122.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_edc13550</th>\n",
       "      <td>188.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_edc13555</th>\n",
       "      <td>31.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_edc13560</th>\n",
       "      <td>211.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_edc13565</th>\n",
       "      <td>253.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_edc13570_I_</th>\n",
       "      <td>222.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_edc13575</th>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_edc13580_I_</th>\n",
       "      <td>1496.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_edc13585</th>\n",
       "      <td>254.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_edc13590</th>\n",
       "      <td>51.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_edc13595</th>\n",
       "      <td>99.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Criteria_Bitscore\n",
       "P_edc13525_I_              127.5\n",
       "P_edc13530                   0.0\n",
       "P_edc13535                 397.4\n",
       "P_edc13545                 122.5\n",
       "P_edc13550                 188.1\n",
       "P_edc13555                  31.5\n",
       "P_edc13560                 211.8\n",
       "P_edc13565                 253.4\n",
       "P_edc13570_I_              222.9\n",
       "P_edc13575                  45.0\n",
       "P_edc13580_I_             1496.4\n",
       "P_edc13585                 254.9\n",
       "P_edc13590                  51.3\n",
       "P_edc13595                  99.7"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>P_edc13525_I_</th>\n",
       "      <th>P_edc13530</th>\n",
       "      <th>P_edc13535</th>\n",
       "      <th>P_edc13545</th>\n",
       "      <th>P_edc13550</th>\n",
       "      <th>P_edc13555</th>\n",
       "      <th>P_edc13560</th>\n",
       "      <th>P_edc13565</th>\n",
       "      <th>P_edc13570_I_</th>\n",
       "      <th>P_edc13575</th>\n",
       "      <th>P_edc13580_I_</th>\n",
       "      <th>P_edc13585</th>\n",
       "      <th>P_edc13590</th>\n",
       "      <th>P_edc13595</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P_Comamonas_thiooxidans_CNB1_edc_bit_score</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>391.7</td>\n",
       "      <td>122.5</td>\n",
       "      <td>188.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>211.8</td>\n",
       "      <td>235.9</td>\n",
       "      <td>217.4</td>\n",
       "      <td>22.6</td>\n",
       "      <td>1496.4</td>\n",
       "      <td>254.9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>91.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_Novosphingobium_sp_strain_Chol11_edc_bit_score</th>\n",
       "      <td>127.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.5</td>\n",
       "      <td>200.1</td>\n",
       "      <td>249.1</td>\n",
       "      <td>222.9</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>228.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_Pseudomonas_putida_DOC21_cluster1_edc_bit_score</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>395.2</td>\n",
       "      <td>109.1</td>\n",
       "      <td>136.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>112.1</td>\n",
       "      <td>212.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>243.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>99.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P_Pseudomonas_stutzeri_Chol-1_edc_bit_score</th>\n",
       "      <td>105.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>397.4</td>\n",
       "      <td>114.8</td>\n",
       "      <td>135.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>203.1</td>\n",
       "      <td>253.4</td>\n",
       "      <td>221.4</td>\n",
       "      <td>35.9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>241.4</td>\n",
       "      <td>51.3</td>\n",
       "      <td>99.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Criteria_Bitscore</th>\n",
       "      <td>127.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>397.4</td>\n",
       "      <td>122.5</td>\n",
       "      <td>188.1</td>\n",
       "      <td>31.5</td>\n",
       "      <td>211.8</td>\n",
       "      <td>253.4</td>\n",
       "      <td>222.9</td>\n",
       "      <td>45.0</td>\n",
       "      <td>1496.4</td>\n",
       "      <td>254.9</td>\n",
       "      <td>51.3</td>\n",
       "      <td>99.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   P_edc13525_I_  P_edc13530  \\\n",
       "P_Comamonas_thiooxidans_CNB1_edc_bit_score                   0.0         0.0   \n",
       "P_Novosphingobium_sp_strain_Chol11_edc_bit_score           127.5         0.0   \n",
       "P_Pseudomonas_putida_DOC21_cluster1_edc_bit_score            0.0         0.0   \n",
       "P_Pseudomonas_stutzeri_Chol-1_edc_bit_score                105.1         0.0   \n",
       "Criteria_Bitscore                                          127.5         0.0   \n",
       "\n",
       "                                                   P_edc13535  P_edc13545  \\\n",
       "P_Comamonas_thiooxidans_CNB1_edc_bit_score              391.7       122.5   \n",
       "P_Novosphingobium_sp_strain_Chol11_edc_bit_score          0.0       111.0   \n",
       "P_Pseudomonas_putida_DOC21_cluster1_edc_bit_score       395.2       109.1   \n",
       "P_Pseudomonas_stutzeri_Chol-1_edc_bit_score             397.4       114.8   \n",
       "Criteria_Bitscore                                       397.4       122.5   \n",
       "\n",
       "                                                   P_edc13550  P_edc13555  \\\n",
       "P_Comamonas_thiooxidans_CNB1_edc_bit_score              188.1         0.0   \n",
       "P_Novosphingobium_sp_strain_Chol11_edc_bit_score          0.0        31.5   \n",
       "P_Pseudomonas_putida_DOC21_cluster1_edc_bit_score       136.3         0.0   \n",
       "P_Pseudomonas_stutzeri_Chol-1_edc_bit_score             135.0         0.0   \n",
       "Criteria_Bitscore                                       188.1        31.5   \n",
       "\n",
       "                                                   P_edc13560  P_edc13565  \\\n",
       "P_Comamonas_thiooxidans_CNB1_edc_bit_score              211.8       235.9   \n",
       "P_Novosphingobium_sp_strain_Chol11_edc_bit_score        200.1       249.1   \n",
       "P_Pseudomonas_putida_DOC21_cluster1_edc_bit_score       112.1       212.6   \n",
       "P_Pseudomonas_stutzeri_Chol-1_edc_bit_score             203.1       253.4   \n",
       "Criteria_Bitscore                                       211.8       253.4   \n",
       "\n",
       "                                                   P_edc13570_I_  P_edc13575  \\\n",
       "P_Comamonas_thiooxidans_CNB1_edc_bit_score                 217.4        22.6   \n",
       "P_Novosphingobium_sp_strain_Chol11_edc_bit_score           222.9        45.0   \n",
       "P_Pseudomonas_putida_DOC21_cluster1_edc_bit_score            0.0         0.0   \n",
       "P_Pseudomonas_stutzeri_Chol-1_edc_bit_score                221.4        35.9   \n",
       "Criteria_Bitscore                                          222.9        45.0   \n",
       "\n",
       "                                                   P_edc13580_I_  P_edc13585  \\\n",
       "P_Comamonas_thiooxidans_CNB1_edc_bit_score                1496.4       254.9   \n",
       "P_Novosphingobium_sp_strain_Chol11_edc_bit_score             0.0       228.2   \n",
       "P_Pseudomonas_putida_DOC21_cluster1_edc_bit_score            0.0       243.1   \n",
       "P_Pseudomonas_stutzeri_Chol-1_edc_bit_score                  0.0       241.4   \n",
       "Criteria_Bitscore                                         1496.4       254.9   \n",
       "\n",
       "                                                   P_edc13590  P_edc13595  \n",
       "P_Comamonas_thiooxidans_CNB1_edc_bit_score                0.0        91.7  \n",
       "P_Novosphingobium_sp_strain_Chol11_edc_bit_score          0.0        19.0  \n",
       "P_Pseudomonas_putida_DOC21_cluster1_edc_bit_score         0.0        99.7  \n",
       "P_Pseudomonas_stutzeri_Chol-1_edc_bit_score              51.3        99.5  \n",
       "Criteria_Bitscore                                        51.3        99.7  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "# HMM actino negative control bit-score threshold\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "\n",
    "#------------------------------------------------------------------------\n",
    "# create a function can get the datafarme of each control's best bit-score, e-vale and coverage dataframes\n",
    "# hmm domtblout name should be ...edc_I_... NOT ..._I_edc_....\n",
    "def negative_control_df(control_dir, control_names, BitScore_df):\n",
    "    # Create an empty dictionary to store the best bit-scores; best e-vale; best coverage for each hmmsearch\n",
    "    best_bit_scores = {}\n",
    "\n",
    "    # Loop over the hmmsearches and parse the corresponding \"domtblout\" file\n",
    "    for control_name in control_names:\n",
    "        # Load the \"domtblout\" file into a pandas DataFrame\n",
    "        file_path = os.path.join(control_dir, control_name + \".domtblout\")\n",
    "        try:\n",
    "            df = pd.read_csv(file_path, comment=\"#\", sep='\\s+', header=None)\n",
    "        except pd.errors.EmptyDataError:\n",
    "            df = pd.DataFrame()\n",
    "\n",
    "        if df.empty:\n",
    "            pattern = r'P_.*_edc'\n",
    "            query_name = re.sub(pattern, 'P_edc', control_name)         \n",
    "            best_bit_scores[query_name] = 0\n",
    "        else:        \n",
    "            # Assign column names to the DataFrame\n",
    "            df.columns = [\"target_name\", \"accession\", \"tlen\", \"query_name\", \"accession2\", \"qlen\", \"E-value\", \"score\", \"bias\",\n",
    "                          \"num_domains_index\", \"num_domains_total\", \"c-Evalue\", \"i-Evalue\", \"score2\", \"bias2\", \"hmm_from\", \"hmm_to\", \"ali_from\", \"ali_to\",\n",
    "                          \"env_from\", \"env_to\", \"acc\", \"description\"]\n",
    "            # Calculate the coverage for each hit\n",
    "            df[\"coverage\"] = (df[\"ali_to\"] - df[\"ali_from\"] + 1) / df[\"tlen\"]\n",
    "\n",
    "            # Filter the DataFrame by E-value and coverage, and sort by bit-score\n",
    "            significant_hits = df[(df[\"E-value\"] < 0.001) & (df[\"coverage\"] > 0.50)].sort_values(by=\"score\", ascending=False)\n",
    "\n",
    "            # replace strains name to aed..\n",
    "            pattern = r'P_.*_edc'\n",
    "            query_name = re.sub(pattern, 'P_edc', control_name)  \n",
    "\n",
    "            # Extract the best bit-score and store it in the dictionary\n",
    "            if not significant_hits.empty:\n",
    "                best_bit_score = significant_hits.iloc[0][\"score\"]\n",
    "                best_bit_scores[query_name] = best_bit_score\n",
    "            else:\n",
    "                best_bit_scores[query_name] = 0\n",
    "\n",
    "    # create the index names for three df\n",
    "    # Get the strain name\n",
    "    pattern = r'P_.*_edc'\n",
    "    StrainName = re.findall(pattern, control_names[0])\n",
    "\n",
    "    # create index of bit score\n",
    "    bitscore_name = StrainName[0] + '_bit_score'\n",
    "    BitScore_Name = [bitscore_name]\n",
    "\n",
    "    # create a dataframe of bit score\n",
    "    BitScore_df = pd.DataFrame(best_bit_scores, index=BitScore_Name)\n",
    "\n",
    "    return BitScore_df\n",
    "\n",
    "\n",
    "#---------------------------\n",
    "# P_Comamonas_thiooxidans_CNB1\n",
    "# Define the directory that contains the \"domtblout\" files, 這個genome protein fasta要把discription的文字刪除，因有空格會影響read\n",
    "control_dir = \"../data/raw/Proteo_HMM_NegativeControl/P_Comamonas_thiooxidans_CNB1/\"\n",
    "\n",
    "# Define the names of the control_names (without the file extension); delet 13540\n",
    "control_names = ['P_Comamonas_thiooxidans_CNB1_edc13525_I_', 'P_Comamonas_thiooxidans_CNB1_edc13530', 'P_Comamonas_thiooxidans_CNB1_edc13535', 'P_Comamonas_thiooxidans_CNB1_edc13545', 'P_Comamonas_thiooxidans_CNB1_edc13550', 'P_Comamonas_thiooxidans_CNB1_edc13555', 'P_Comamonas_thiooxidans_CNB1_edc13560', 'P_Comamonas_thiooxidans_CNB1_edc13565', 'P_Comamonas_thiooxidans_CNB1_edc13570_I_', 'P_Comamonas_thiooxidans_CNB1_edc13575', 'P_Comamonas_thiooxidans_CNB1_edc13580_I_', 'P_Comamonas_thiooxidans_CNB1_edc13585', 'P_Comamonas_thiooxidans_CNB1_edc13590', 'P_Comamonas_thiooxidans_CNB1_edc13595']\n",
    "\n",
    "# Use positive_control_df function to get the df (control_dir, control_names, BitScore_df)\n",
    "CNB1_BitScore = pd.DataFrame()\n",
    "CNB1_BitScore = negative_control_df(control_dir, control_names, CNB1_BitScore)\n",
    "\n",
    "\n",
    "#---------------------------\n",
    "# P_Novosphingobium_sp_strain_Chol11\n",
    "# Define the directory that contains the \"domtblout\" files, 這個genome protein fasta要把discription的文字刪除，因有空格會影響read\n",
    "control_dir = \"../data/raw/Proteo_HMM_NegativeControl/P_Novosphingobium_sp_strain_Chol11/\"\n",
    "\n",
    "# Define the names of the control_names (without the file extension); delet 13540\n",
    "control_names = ['P_Novosphingobium_sp_strain_Chol11_edc13525_I_', 'P_Novosphingobium_sp_strain_Chol11_edc13530', 'P_Novosphingobium_sp_strain_Chol11_edc13535', 'P_Novosphingobium_sp_strain_Chol11_edc13545', 'P_Novosphingobium_sp_strain_Chol11_edc13550', 'P_Novosphingobium_sp_strain_Chol11_edc13555', 'P_Novosphingobium_sp_strain_Chol11_edc13560', 'P_Novosphingobium_sp_strain_Chol11_edc13565', 'P_Novosphingobium_sp_strain_Chol11_edc13570_I_', 'P_Novosphingobium_sp_strain_Chol11_edc13575', 'P_Novosphingobium_sp_strain_Chol11_edc13580_I_', 'P_Novosphingobium_sp_strain_Chol11_edc13585', 'P_Novosphingobium_sp_strain_Chol11_edc13590', 'P_Novosphingobium_sp_strain_Chol11_edc13595']\n",
    "\n",
    "# Use positive_control_df function to get the df (control_dir, control_names, BitScore_df)\n",
    "Chol11_BitScore = pd.DataFrame()\n",
    "Chol11_BitScore = negative_control_df(control_dir, control_names, Chol11_BitScore)\n",
    "\n",
    "\n",
    "#---------------------------\n",
    "# P_Pseudomonas_putida_DOC21_cluster\n",
    "# Define the directory that contains the \"domtblout\" files, 這個genome protein fasta要把discription的文字刪除，因有空格會影響read\n",
    "control_dir = \"../data/raw/Proteo_HMM_NegativeControl/P_Pseudomonas_putida_DOC21_cluster/\"\n",
    "\n",
    "# Define the names of the control_names (without the file extension); delet 13540\n",
    "control_names = ['P_Pseudomonas_putida_DOC21_cluster1_edc13525_I_', 'P_Pseudomonas_putida_DOC21_cluster1_edc13530', 'P_Pseudomonas_putida_DOC21_cluster1_edc13535', 'P_Pseudomonas_putida_DOC21_cluster1_edc13545', 'P_Pseudomonas_putida_DOC21_cluster1_edc13550', 'P_Pseudomonas_putida_DOC21_cluster1_edc13555', 'P_Pseudomonas_putida_DOC21_cluster1_edc13560', 'P_Pseudomonas_putida_DOC21_cluster1_edc13565', 'P_Pseudomonas_putida_DOC21_cluster1_edc13570_I_', 'P_Pseudomonas_putida_DOC21_cluster1_edc13575', 'P_Pseudomonas_putida_DOC21_cluster1_edc13580_I_', 'P_Pseudomonas_putida_DOC21_cluster1_edc13585', 'P_Pseudomonas_putida_DOC21_cluster1_edc13590', 'P_Pseudomonas_putida_DOC21_cluster1_edc13595']\n",
    "\n",
    "# Use positive_control_df function to get the df (control_dir, control_names, BitScore_df)\n",
    "DOC21_BitScore = pd.DataFrame()\n",
    "DOC21_BitScore = negative_control_df(control_dir, control_names, DOC21_BitScore)\n",
    "\n",
    "\n",
    "#---------------------------\n",
    "# P_Pseudomonas_stutzeri_Chol-1\n",
    "# Define the directory that contains the \"domtblout\" files, 這個genome protein fasta要把discription的文字刪除，因有空格會影響read\n",
    "control_dir = \"../data/raw/Proteo_HMM_NegativeControl/P_Pseudomonas_stutzeri_Chol-1/\"\n",
    "\n",
    "# Define the names of the control_names (without the file extension); delet 13540\n",
    "control_names = ['P_Pseudomonas_stutzeri_Chol-1_edc13525_I_', 'P_Pseudomonas_stutzeri_Chol-1_edc13530', 'P_Pseudomonas_stutzeri_Chol-1_edc13535', 'P_Pseudomonas_stutzeri_Chol-1_edc13545', 'P_Pseudomonas_stutzeri_Chol-1_edc13550', 'P_Pseudomonas_stutzeri_Chol-1_edc13555', 'P_Pseudomonas_stutzeri_Chol-1_edc13560', 'P_Pseudomonas_stutzeri_Chol-1_edc13565', 'P_Pseudomonas_stutzeri_Chol-1_edc13570_I_', 'P_Pseudomonas_stutzeri_Chol-1_edc13575', 'P_Pseudomonas_stutzeri_Chol-1_edc13580_I_', 'P_Pseudomonas_stutzeri_Chol-1_edc13585', 'P_Pseudomonas_stutzeri_Chol-1_edc13590', 'P_Pseudomonas_stutzeri_Chol-1_edc13595']\n",
    "\n",
    "# Use positive_control_df function to get the df (control_dir, control_names, BitScore_df)\n",
    "Chol01_BitScore = pd.DataFrame()\n",
    "Chol01_BitScore = negative_control_df(control_dir, control_names, Chol01_BitScore)\n",
    "\n",
    "\n",
    "#-------------------------------------------------------------------------------------------------------\n",
    "# merge four negative best bit score df\n",
    "Proteo_Negative_df = pd.concat([CNB1_BitScore, Chol11_BitScore, DOC21_BitScore, Chol01_BitScore], axis=0)\n",
    "\n",
    "# Got the highest bit score of the df\n",
    "Proteo_BitScore_Criteria = Proteo_Negative_df.max()\n",
    "Proteo_BitScore_Criteria = Proteo_BitScore_Criteria.iloc[0:]\n",
    "Proteo_BitScore_Criteria = Proteo_BitScore_Criteria.to_frame()\n",
    "Proteo_BitScore_Criteria.columns = ['Criteria_Bitscore']\n",
    "\n",
    "# transpose the DataFrame\n",
    "Proteo_BitScore_Criteria_T = Proteo_BitScore_Criteria.T\n",
    "\n",
    "# merge all nad Min table\n",
    "Proteo_Negative_df = pd.concat([Proteo_Negative_df, Proteo_BitScore_Criteria_T], axis=0)\n",
    "Proteo_Negative_df.to_csv('../data/processed/Final/ForReader/ControlData/Proteo_edc/Proteo_NegativeBitscore.csv')\n",
    "\n",
    "# done\n",
    "display(Proteo_BitScore_Criteria)\n",
    "display(Proteo_Negative_df)\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b6c3224-a219-431f-aac2-97e0f7bc59fa",
   "metadata": {},
   "source": [
    "### 2. 使用上述的 bit score 來篩選 MAGS 的 outpout data (不含有 hits 的篩選)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "170185a7-526c-4044-8948-38f05c2ba69f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n",
      "unique query name:  ['Q_P_13595' 'Q_P_13590' 'Q_P_13585' 'Q_P_I_edcC_13580' 'Q_P_13575'\n",
      " 'Q_P_I_edcB_13570' 'Q_P_13565' 'Q_P_13560' 'Q_P_13555' 'Q_P_13550'\n",
      " 'Q_P_13545' 'Q_P_13535' 'Q_P_13530' 'Q_P_I_edcA_13525']  /  14\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_name</th>\n",
       "      <th>accession</th>\n",
       "      <th>tlen</th>\n",
       "      <th>query_name</th>\n",
       "      <th>accession2</th>\n",
       "      <th>qlen</th>\n",
       "      <th>E-value</th>\n",
       "      <th>score</th>\n",
       "      <th>bias</th>\n",
       "      <th>num_domains_index</th>\n",
       "      <th>...</th>\n",
       "      <th>bias2</th>\n",
       "      <th>hmm_from</th>\n",
       "      <th>hmm_to</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>ali_to</th>\n",
       "      <th>env_from</th>\n",
       "      <th>env_to</th>\n",
       "      <th>acc</th>\n",
       "      <th>description</th>\n",
       "      <th>coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3300024060.a:Ga0209987_10000324_8</td>\n",
       "      <td>-</td>\n",
       "      <td>161</td>\n",
       "      <td>Q_P_13595</td>\n",
       "      <td>-</td>\n",
       "      <td>145</td>\n",
       "      <td>1.400000e-51</td>\n",
       "      <td>186.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>145</td>\n",
       "      <td>15</td>\n",
       "      <td>160</td>\n",
       "      <td>15</td>\n",
       "      <td>160</td>\n",
       "      <td>0.99</td>\n",
       "      <td>-</td>\n",
       "      <td>0.906832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3300024431.a:Ga0209988_10001403_11</td>\n",
       "      <td>-</td>\n",
       "      <td>161</td>\n",
       "      <td>Q_P_13595</td>\n",
       "      <td>-</td>\n",
       "      <td>145</td>\n",
       "      <td>1.400000e-51</td>\n",
       "      <td>186.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>145</td>\n",
       "      <td>15</td>\n",
       "      <td>160</td>\n",
       "      <td>15</td>\n",
       "      <td>160</td>\n",
       "      <td>0.99</td>\n",
       "      <td>-</td>\n",
       "      <td>0.906832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3300012090.a:Ga0153956_1000496_23</td>\n",
       "      <td>-</td>\n",
       "      <td>161</td>\n",
       "      <td>Q_P_13595</td>\n",
       "      <td>-</td>\n",
       "      <td>145</td>\n",
       "      <td>2.200000e-51</td>\n",
       "      <td>185.8</td>\n",
       "      <td>0.2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2</td>\n",
       "      <td>3</td>\n",
       "      <td>145</td>\n",
       "      <td>16</td>\n",
       "      <td>160</td>\n",
       "      <td>14</td>\n",
       "      <td>160</td>\n",
       "      <td>0.97</td>\n",
       "      <td>-</td>\n",
       "      <td>0.900621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3300026193.a:Ga0208129_1000342_13</td>\n",
       "      <td>-</td>\n",
       "      <td>162</td>\n",
       "      <td>Q_P_13595</td>\n",
       "      <td>-</td>\n",
       "      <td>145</td>\n",
       "      <td>2.800000e-51</td>\n",
       "      <td>185.5</td>\n",
       "      <td>0.1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1</td>\n",
       "      <td>1</td>\n",
       "      <td>145</td>\n",
       "      <td>16</td>\n",
       "      <td>161</td>\n",
       "      <td>16</td>\n",
       "      <td>161</td>\n",
       "      <td>0.99</td>\n",
       "      <td>-</td>\n",
       "      <td>0.901235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3300026202.a:Ga0207984_1001029_7</td>\n",
       "      <td>-</td>\n",
       "      <td>162</td>\n",
       "      <td>Q_P_13595</td>\n",
       "      <td>-</td>\n",
       "      <td>145</td>\n",
       "      <td>2.800000e-51</td>\n",
       "      <td>185.5</td>\n",
       "      <td>0.1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1</td>\n",
       "      <td>1</td>\n",
       "      <td>145</td>\n",
       "      <td>16</td>\n",
       "      <td>161</td>\n",
       "      <td>16</td>\n",
       "      <td>161</td>\n",
       "      <td>0.99</td>\n",
       "      <td>-</td>\n",
       "      <td>0.901235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16530</th>\n",
       "      <td>3300014205.a:Ga0172380_10055494_2</td>\n",
       "      <td>-</td>\n",
       "      <td>412</td>\n",
       "      <td>Q_P_I_edcA_13525</td>\n",
       "      <td>-</td>\n",
       "      <td>393</td>\n",
       "      <td>4.700000e-33</td>\n",
       "      <td>127.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12</td>\n",
       "      <td>369</td>\n",
       "      <td>21</td>\n",
       "      <td>383</td>\n",
       "      <td>16</td>\n",
       "      <td>409</td>\n",
       "      <td>0.83</td>\n",
       "      <td>-</td>\n",
       "      <td>0.881068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16531</th>\n",
       "      <td>3300026118.a:Ga0207675_100018219_6</td>\n",
       "      <td>-</td>\n",
       "      <td>408</td>\n",
       "      <td>Q_P_I_edcA_13525</td>\n",
       "      <td>-</td>\n",
       "      <td>393</td>\n",
       "      <td>4.800000e-33</td>\n",
       "      <td>127.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>69</td>\n",
       "      <td>390</td>\n",
       "      <td>83</td>\n",
       "      <td>400</td>\n",
       "      <td>9</td>\n",
       "      <td>403</td>\n",
       "      <td>0.87</td>\n",
       "      <td>-</td>\n",
       "      <td>0.779412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16532</th>\n",
       "      <td>3300026980.a:Ga0207829_101107_6</td>\n",
       "      <td>-</td>\n",
       "      <td>408</td>\n",
       "      <td>Q_P_I_edcA_13525</td>\n",
       "      <td>-</td>\n",
       "      <td>393</td>\n",
       "      <td>4.800000e-33</td>\n",
       "      <td>127.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>77</td>\n",
       "      <td>368</td>\n",
       "      <td>98</td>\n",
       "      <td>380</td>\n",
       "      <td>89</td>\n",
       "      <td>391</td>\n",
       "      <td>0.92</td>\n",
       "      <td>-</td>\n",
       "      <td>0.693627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16551</th>\n",
       "      <td>3300015206.a:Ga0167644_1000100_80</td>\n",
       "      <td>-</td>\n",
       "      <td>409</td>\n",
       "      <td>Q_P_I_edcA_13525</td>\n",
       "      <td>-</td>\n",
       "      <td>393</td>\n",
       "      <td>4.800000e-33</td>\n",
       "      <td>127.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>70</td>\n",
       "      <td>351</td>\n",
       "      <td>86</td>\n",
       "      <td>368</td>\n",
       "      <td>74</td>\n",
       "      <td>382</td>\n",
       "      <td>0.91</td>\n",
       "      <td>-</td>\n",
       "      <td>0.691932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16552</th>\n",
       "      <td>3300027623.a:Ga0209828_1000783_10</td>\n",
       "      <td>-</td>\n",
       "      <td>412</td>\n",
       "      <td>Q_P_I_edcA_13525</td>\n",
       "      <td>-</td>\n",
       "      <td>393</td>\n",
       "      <td>4.800000e-33</td>\n",
       "      <td>127.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15</td>\n",
       "      <td>366</td>\n",
       "      <td>19</td>\n",
       "      <td>387</td>\n",
       "      <td>13</td>\n",
       "      <td>411</td>\n",
       "      <td>0.82</td>\n",
       "      <td>-</td>\n",
       "      <td>0.895631</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>83953 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              target_name accession tlen        query_name  \\\n",
       "0       3300024060.a:Ga0209987_10000324_8         -  161         Q_P_13595   \n",
       "1      3300024431.a:Ga0209988_10001403_11         -  161         Q_P_13595   \n",
       "2       3300012090.a:Ga0153956_1000496_23         -  161         Q_P_13595   \n",
       "3       3300026193.a:Ga0208129_1000342_13         -  162         Q_P_13595   \n",
       "4        3300026202.a:Ga0207984_1001029_7         -  162         Q_P_13595   \n",
       "...                                   ...       ...  ...               ...   \n",
       "16530   3300014205.a:Ga0172380_10055494_2         -  412  Q_P_I_edcA_13525   \n",
       "16531  3300026118.a:Ga0207675_100018219_6         -  408  Q_P_I_edcA_13525   \n",
       "16532     3300026980.a:Ga0207829_101107_6         -  408  Q_P_I_edcA_13525   \n",
       "16551   3300015206.a:Ga0167644_1000100_80         -  409  Q_P_I_edcA_13525   \n",
       "16552   3300027623.a:Ga0209828_1000783_10         -  412  Q_P_I_edcA_13525   \n",
       "\n",
       "      accession2 qlen       E-value  score  bias num_domains_index  ... bias2  \\\n",
       "0              -  145  1.400000e-51  186.5   0.0                 1  ...   0.0   \n",
       "1              -  145  1.400000e-51  186.5   0.0                 1  ...   0.0   \n",
       "2              -  145  2.200000e-51  185.8   0.2                 1  ...   0.2   \n",
       "3              -  145  2.800000e-51  185.5   0.1                 1  ...   0.1   \n",
       "4              -  145  2.800000e-51  185.5   0.1                 1  ...   0.1   \n",
       "...          ...  ...           ...    ...   ...               ...  ...   ...   \n",
       "16530          -  393  4.700000e-33  127.6   0.0                 1  ...   0.0   \n",
       "16531          -  393  4.800000e-33  127.6   0.0                 1  ...   0.0   \n",
       "16532          -  393  4.800000e-33  127.6   0.0                 1  ...   0.0   \n",
       "16551          -  393  4.800000e-33  127.6   0.0                 1  ...   0.0   \n",
       "16552          -  393  4.800000e-33  127.6   0.0                 1  ...   0.0   \n",
       "\n",
       "       hmm_from  hmm_to  ali_from  ali_to env_from env_to   acc description  \\\n",
       "0             1     145        15     160       15    160  0.99           -   \n",
       "1             1     145        15     160       15    160  0.99           -   \n",
       "2             3     145        16     160       14    160  0.97           -   \n",
       "3             1     145        16     161       16    161  0.99           -   \n",
       "4             1     145        16     161       16    161  0.99           -   \n",
       "...         ...     ...       ...     ...      ...    ...   ...         ...   \n",
       "16530        12     369        21     383       16    409  0.83           -   \n",
       "16531        69     390        83     400        9    403  0.87           -   \n",
       "16532        77     368        98     380       89    391  0.92           -   \n",
       "16551        70     351        86     368       74    382  0.91           -   \n",
       "16552        15     366        19     387       13    411  0.82           -   \n",
       "\n",
       "       coverage  \n",
       "0      0.906832  \n",
       "1      0.906832  \n",
       "2      0.900621  \n",
       "3      0.901235  \n",
       "4      0.901235  \n",
       "...         ...  \n",
       "16530  0.881068  \n",
       "16531  0.779412  \n",
       "16532  0.693627  \n",
       "16551  0.691932  \n",
       "16552  0.895631  \n",
       "\n",
       "[83953 rows x 24 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Proteo_HMM_MAGs\n",
    "# 需執行前一個cell程式\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "# Aed Cluster to MAGs\n",
    "# Define the directory that contains the \"domtblout\" files.需要刪除discription\n",
    "domtblout_dir = \"../data/raw/Proteo_HMM_MAGs_domtblout/\"\n",
    "\n",
    "# Create an empty dictionary to store the target name for each hmmsearch\n",
    "MAGs_Hits = {}\n",
    "MAGs_Hits_name = []\n",
    "\n",
    "# covert criteria dataframe to serires\n",
    "Proteo_BitScore_Criteria_S = Proteo_BitScore_Criteria['Criteria_Bitscore']\n",
    "Proteo_BitScore_Criteria_S = Proteo_BitScore_Criteria_S.astype(float)  \n",
    "\n",
    "# create a dataframe for a all hits \n",
    "columns = [\"target_name\", \"accession\", \"tlen\", \"query_name\", \"accession2\", \"qlen\", \"E-value\", \"score\", \"bias\",\n",
    "           \"num_domains_index\", \"num_domains_total\", \"c-Evalue\", \"i-Evalue\", \"score2\", \"bias2\", \"hmm_from\",\n",
    "           \"hmm_to\", \"ali_from\", \"ali_to\", \"env_from\", \"env_to\", \"acc\", \"description\"]\n",
    "\n",
    "All_edc_Hits_df = pd.DataFrame(columns=columns)\n",
    "\n",
    "# Loop over the HMM DOMTBLOUT files and filter the results based on bit score, e-value and coverage\n",
    "# hmm name and bit score are in the Actino_BitScore_Criteria series\n",
    "for hmmsearch, threshold in Proteo_BitScore_Criteria_S.items():\n",
    "    # Load the \"domtblout\" file into a pandas DataFrame\n",
    "    file_path = os.path.join(domtblout_dir, hmmsearch + \".domtblout\")\n",
    "    df = pd.read_csv(file_path, comment=\"#\", sep='\\s+', header=None)    \n",
    "    # Assign column names to the DataFrame\n",
    "    df.columns = [\"target_name\", \"accession\", \"tlen\", \"query_name\", \"accession2\", \"qlen\", \"E-value\", \"score\", \"bias\",\n",
    "                  \"num_domains_index\", \"num_domains_total\", \"c-Evalue\", \"i-Evalue\", \"score2\", \"bias2\", \"hmm_from\", \"hmm_to\", \"ali_from\", \"ali_to\",\n",
    "                  \"env_from\", \"env_to\", \"acc\", \"description\"]\n",
    "\n",
    "    # Calculate the coverage for each hit\n",
    "    df[\"coverage\"] = (df[\"ali_to\"] - df[\"ali_from\"] + 1) / df[\"tlen\"]\n",
    "    \n",
    "    # Filter the DataFrame by E-value, coverage, and bit-score\n",
    "    significant_hits = df[(df[\"E-value\"] < 0.001) & (df[\"coverage\"] > 0.50) & (df[\"score\"] > threshold)]\n",
    "\n",
    "    # Extract Target nmae and store it in the dictionary\n",
    "    if not significant_hits.empty:\n",
    "        MAGs_Hits_name = significant_hits[\"target_name\"].tolist()\n",
    "        MAGs_Hits[hmmsearch] = MAGs_Hits_name\n",
    "    else:\n",
    "        MAGs_Hits_name = None\n",
    "        MAGs_Hits[hmmsearch] = MAGs_Hits_name\n",
    "    \n",
    "    # add hits table to a df\n",
    "    All_edc_Hits_df = pd.concat([significant_hits, All_edc_Hits_df], axis=0)\n",
    "\n",
    "# done\n",
    "print('done')\n",
    "print('unique query name: ', All_edc_Hits_df['query_name'].unique(), ' / ', len(All_edc_Hits_df['query_name'].unique()))\n",
    "All_edc_Hits_df.to_csv('../data/processed/All_edc_Hits_df_bitscore.csv')  \n",
    "All_edc_Hits_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bab7ff9a-e28a-40bd-8d17-018cdb680eb9",
   "metadata": {},
   "source": [
    "#### 2.1 獲取positive的target name並連結MAGsID，整理出一個類似heatmap的table，再以寬鬆hits數量進行篩選(>8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "083c8f41-0738-4bf7-b188-c56e3bba95ed",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Any Null in TargetToMAGsID:  False\n",
      "done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>query_name</th>\n",
       "      <th>genome_id</th>\n",
       "      <th>num_hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3300026101_4</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3300020369_2</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3300024344_6</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3300027685_11</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3300024431_16</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592</th>\n",
       "      <td>3300010158_16</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>3300020432_23</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>594</th>\n",
       "      <td>3300024514_1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595</th>\n",
       "      <td>3300020477_54</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>596</th>\n",
       "      <td>3300024339_10</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>597 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "query_name      genome_id  num_hits\n",
       "0            3300026101_4        14\n",
       "1            3300020369_2        14\n",
       "2            3300024344_6        13\n",
       "3           3300027685_11        13\n",
       "4           3300024431_16        13\n",
       "..                    ...       ...\n",
       "592         3300010158_16         8\n",
       "593         3300020432_23         8\n",
       "594          3300024514_1         8\n",
       "595         3300020477_54         8\n",
       "596         3300024339_10         8\n",
       "\n",
       "[597 rows x 2 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Got target name and query name\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "All_edc_Hits_df = pd.read_csv('../data/processed/All_edc_Hits_df_bitscore.csv') \n",
    "All_edc_Hits_TargetAndQuery = All_edc_Hits_df[['target_name', 'query_name']]\n",
    "\n",
    "# Load a Dataframe with the lookup values for merge protein id to MAGs and merge them\n",
    "TarToMAGs_edc = pd.read_csv('../data/interim/edc_All_TarToMAGsID.csv')\n",
    "\n",
    "# use merge() function to join the MAGsID data\n",
    "edc_hits_TargetAndMAGsID = pd.merge(All_edc_Hits_TargetAndQuery, TarToMAGs_edc, on='target_name', how='left')                                                                               \n",
    "\n",
    "# check the null value\n",
    "print('Any Null in TargetToMAGsID: ', edc_hits_TargetAndMAGsID['MAGsID'].isnull().any())\n",
    "\n",
    "# create the crosstab table (like heatmap)\n",
    "edc_hits_heatmap = pd.crosstab(edc_hits_TargetAndMAGsID['query_name'], edc_hits_TargetAndMAGsID['MAGsID'], dropna=False)\n",
    "edc_hits_heatmap = edc_hits_heatmap.transpose()\n",
    "edc_hits_heatmap\n",
    "# edc_hits_heatmap.to_csv('../data/processed/edc_hits_heatmap_bitscore.csv')\n",
    "\n",
    "# Count the non-zero values in hmm profiles hit row to calculate the number of different HMM profiles that have hits in a given MAG. \n",
    "def count_nonzero(row):\n",
    "    return len(row[row != 0])\n",
    "\n",
    "num_hits = edc_hits_heatmap.apply(count_nonzero, axis=1)\n",
    "edc_hits_heatmap['num_hits'] = num_hits\n",
    "\n",
    "# sort them by hits numer\n",
    "edc_hits_heatmap = edc_hits_heatmap.sort_values(by=\"num_hits\", ascending=False)\n",
    "\n",
    "# extract >8 hmm profiles hits and the necessary hits (aedA、aedB、aedJ)\n",
    "edc_hits_FinalFilter =  edc_hits_heatmap[(edc_hits_heatmap['num_hits'] >= 8)]\n",
    "# >7 = 1020, >8 = 597, >9 = 294, >10 = 111 先用8看看 大於一半的query gene\n",
    "\n",
    "# Reset index and move index column to first position\n",
    "edc_hits_FinalFilter.index.name = None\n",
    "edc_hits_FinalFilter = edc_hits_FinalFilter.reset_index()\n",
    "edc_hits_FinalFilter.insert(0, 'index', edc_hits_FinalFilter.pop('index'))\n",
    "\n",
    "# rename MAGsID\n",
    "edc_hits_FinalFilter = edc_hits_FinalFilter.rename(columns={'index': 'genome_id'})\n",
    "\n",
    "# extract the MAGsID and num_hits column\n",
    "edc_Positive_MAGsID = edc_hits_FinalFilter[['genome_id', 'num_hits']]\n",
    "edc_Positive_MAGsID\n",
    "\n",
    "# done\n",
    "print('done')\n",
    "edc_Positive_MAGsID"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7618e20a-774e-4aeb-ae07-8da563f02f0f",
   "metadata": {},
   "source": [
    "##### 2.1.1 連結loose positive MAGs到 metagenomes 總資料中並進行處理成最株能給R繪製 stacket barchart 的資料格式"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fc3cd789-002d-415f-9377-8e9905761ad6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['p__Proteobacteria' 'p__Myxococcota' 'p__Actinobacteriota' 'p__UBP10'\n",
      " 'p__Chloroflexota' 'p__Methylomirabilota' 'p__Entotheonellota'\n",
      " 'p__Desulfobacterota' 'p__SAR324' 'p__Spirochaetota']\n",
      "Any Null in edc_Positive_metagenomes_df:\n",
      " genome_id             False\n",
      "metagenome_id         False\n",
      "num_hits              False\n",
      "Phylum                False\n",
      "Class                  True\n",
      "ecosystem             False\n",
      "ecosystem_category    False\n",
      "longitude              True\n",
      "latitude               True\n",
      "Homologous_cluster    False\n",
      "dtype: bool\n",
      "doen\n",
      "Number of p__UBP10: 48\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>metagenome_id</th>\n",
       "      <th>num_hits</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>ecosystem</th>\n",
       "      <th>ecosystem_category</th>\n",
       "      <th>longitude</th>\n",
       "      <th>latitude</th>\n",
       "      <th>Homologous_cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3300026101_4</td>\n",
       "      <td>3300026101</td>\n",
       "      <td>14</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>-55.3027</td>\n",
       "      <td>9.7046</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3300020369_2</td>\n",
       "      <td>3300020369</td>\n",
       "      <td>14</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>-3.2659</td>\n",
       "      <td>-20.3356</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3300024344_6</td>\n",
       "      <td>3300024344</td>\n",
       "      <td>13</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__UBA9160</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>25.4868</td>\n",
       "      <td>36.5264</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3300027685_11</td>\n",
       "      <td>3300027685</td>\n",
       "      <td>13</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__UBA9160</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>-44.9905</td>\n",
       "      <td>-37.9728</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3300024431_16</td>\n",
       "      <td>3300024431</td>\n",
       "      <td>13</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__UBA9160</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>-177.6630</td>\n",
       "      <td>-32.8507</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592</th>\n",
       "      <td>3300010158_16</td>\n",
       "      <td>3300010158</td>\n",
       "      <td>8</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>-72.5000</td>\n",
       "      <td>46.8319</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>3300020432_23</td>\n",
       "      <td>3300020432</td>\n",
       "      <td>8</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>-88.4916</td>\n",
       "      <td>25.7109</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>594</th>\n",
       "      <td>3300024514_1</td>\n",
       "      <td>3300024514</td>\n",
       "      <td>8</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Alphaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>-81.6053</td>\n",
       "      <td>31.4271</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595</th>\n",
       "      <td>3300020477_54</td>\n",
       "      <td>3300020477</td>\n",
       "      <td>8</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>-142.5770</td>\n",
       "      <td>-8.9103</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>596</th>\n",
       "      <td>3300024339_10</td>\n",
       "      <td>3300024339</td>\n",
       "      <td>8</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Alphaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>-123.5010</td>\n",
       "      <td>48.5847</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>597 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         genome_id  metagenome_id  num_hits               Phylum  \\\n",
       "0     3300026101_4     3300026101        14    p__Proteobacteria   \n",
       "1     3300020369_2     3300020369        14    p__Proteobacteria   \n",
       "2     3300024344_6     3300024344        13       p__Myxococcota   \n",
       "3    3300027685_11     3300027685        13       p__Myxococcota   \n",
       "4    3300024431_16     3300024431        13       p__Myxococcota   \n",
       "..             ...            ...       ...                  ...   \n",
       "592  3300010158_16     3300010158         8  p__Actinobacteriota   \n",
       "593  3300020432_23     3300020432         8  p__Actinobacteriota   \n",
       "594   3300024514_1     3300024514         8    p__Proteobacteria   \n",
       "595  3300020477_54     3300020477         8  p__Actinobacteriota   \n",
       "596  3300024339_10     3300024339         8    p__Proteobacteria   \n",
       "\n",
       "                      Class ecosystem ecosystem_category  longitude  latitude  \\\n",
       "0    c__Gammaproteobacteria   Aquatic             Marine   -55.3027    9.7046   \n",
       "1    c__Gammaproteobacteria   Aquatic             Marine    -3.2659  -20.3356   \n",
       "2                c__UBA9160   Aquatic             Marine    25.4868   36.5264   \n",
       "3                c__UBA9160   Aquatic             Marine   -44.9905  -37.9728   \n",
       "4                c__UBA9160   Aquatic             Marine  -177.6630  -32.8507   \n",
       "..                      ...       ...                ...        ...       ...   \n",
       "592       c__Acidimicrobiia   Aquatic         Freshwater   -72.5000   46.8319   \n",
       "593       c__Acidimicrobiia   Aquatic             Marine   -88.4916   25.7109   \n",
       "594  c__Alphaproteobacteria   Aquatic         Freshwater   -81.6053   31.4271   \n",
       "595       c__Acidimicrobiia   Aquatic             Marine  -142.5770   -8.9103   \n",
       "596  c__Alphaproteobacteria   Aquatic             Marine  -123.5010   48.5847   \n",
       "\n",
       "     Homologous_cluster  \n",
       "0    Proteo_edc_cluster  \n",
       "1    Proteo_edc_cluster  \n",
       "2    Proteo_edc_cluster  \n",
       "3    Proteo_edc_cluster  \n",
       "4    Proteo_edc_cluster  \n",
       "..                  ...  \n",
       "592  Proteo_edc_cluster  \n",
       "593  Proteo_edc_cluster  \n",
       "594  Proteo_edc_cluster  \n",
       "595  Proteo_edc_cluster  \n",
       "596  Proteo_edc_cluster  \n",
       "\n",
       "[597 rows x 10 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 需先執行上一個cell\n",
    "\n",
    "# open metagenome csv files\n",
    "metagenmoes_df = pd.read_csv('../data/external/genome_metadata_editForAnalysis_NotReference.csv')\n",
    "\n",
    "# merge positive MAGs with metagenome\n",
    "edc_Positive_metagenomes_df = pd.merge(edc_Positive_MAGsID, metagenmoes_df, on='genome_id', how='left')\n",
    "\n",
    "# print(edc_Positive_metagenomes_df['ecosystem_category'].unique())\n",
    "# print(edc_Positive_metagenomes_df['ecosystem_type'].unique())\n",
    "# display(edc_Positive_metagenomes_df[edc_Positive_metagenomes_df['ecosystem_type'] == 'Bacteria']) #Digestive system, Anaerobic\n",
    "\n",
    "# extract certain column\n",
    "edc_Positive_metagenomes_df = edc_Positive_metagenomes_df[['genome_id', 'metagenome_id','taxonomy', 'ecosystem', 'ecosystem_category','num_hits', 'longitude', 'latitude']]\n",
    "\n",
    "# extract phylum and class from taxonomy column and expand to new column\n",
    "edc_Positive_metagenomes_df['Phylum'] = edc_Positive_metagenomes_df['taxonomy'].str.extract('(p__\\w+)', expand=True)\n",
    "edc_Positive_metagenomes_df['Class'] = edc_Positive_metagenomes_df['taxonomy'].str.extract('(c__\\w+)', expand=True)\n",
    "\n",
    "# # print non-duplicated values in ecosystem_type than check the lable\n",
    "# print(edc_Positive_metagenomes_df['ecosystem_type'].unique())\n",
    "\n",
    "# # check the None value in ecosystem column Create a Boolean mask to identify NaN values\n",
    "# mask = edc_addEco_df.isna()\n",
    "# edc_addEco_df_nan_rows = edc_addEco_df[mask.any(axis=1)]\n",
    "# edc_addEco_df_nan_rows\n",
    "\n",
    "# remove, add and rearrange column\n",
    "edc_Positive_metagenomes_df.drop('taxonomy', axis=1, inplace=True)\n",
    "edc_Positive_metagenomes_df = edc_Positive_metagenomes_df.reindex(columns=['genome_id', 'metagenome_id', 'num_hits', 'Phylum', 'Class', 'ecosystem', 'ecosystem_category', 'longitude', 'latitude'])\n",
    "edc_Positive_metagenomes_df['Homologous_cluster']='Proteo_edc_cluster'\n",
    "\n",
    "#check phylum data and null value\n",
    "print(edc_Positive_metagenomes_df['Phylum'].unique())\n",
    "print('Any Null in edc_Positive_metagenomes_df:\\n', edc_Positive_metagenomes_df.isnull().any())\n",
    "\n",
    "# write file\n",
    "edc_Positive_metagenomes_df.to_csv('../data/processed/Final/Proteo/edc_PositiveHits_ForR_loose.csv')\n",
    "print('doen')\n",
    "print('Number of p__UBP10:', edc_Positive_metagenomes_df[edc_Positive_metagenomes_df['Phylum'] == 'p__UBP10'].shape[0])\n",
    "edc_Positive_metagenomes_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c41d41a2-274c-4913-a422-22a4d2dcc43f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### 2.2 獲取positive的target name並連結MAGsID，整理出一個類似heatmap的table，再以嚴格hits數量進行篩選(>8, aedA, aedB, aedJ)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6abcda42-2615-49f5-8510-13958f49e06b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Any Null in TargetToMAGsID:  False\n",
      "done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>query_name</th>\n",
       "      <th>genome_id</th>\n",
       "      <th>num_hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3300020369_2</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3300026101_4</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3300024344_6</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3300027685_11</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3300025886_13</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3300017444_33</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3300006913_5</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3300020460_9</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3300020463_34</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3300021343_7</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>3300014833_14</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>3300020347_2</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>3300025547_3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>3300025216_2</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>3300025709_5</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>3300025213_3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>3300025259_2</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>3300025236_4</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>3300027812_7</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>3300002092_1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>3300012149_4</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>3300027826_14</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>3300025044_4</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>3300025079_3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "query_name      genome_id  num_hits\n",
       "0            3300020369_2        13\n",
       "1            3300026101_4        13\n",
       "2            3300024344_6        12\n",
       "3           3300027685_11        12\n",
       "4           3300025886_13        11\n",
       "5           3300017444_33        11\n",
       "6            3300006913_5        11\n",
       "7            3300020460_9        10\n",
       "8           3300020463_34        10\n",
       "9            3300021343_7        10\n",
       "10          3300014833_14         9\n",
       "11           3300020347_2         9\n",
       "12           3300025547_3         8\n",
       "13           3300025216_2         8\n",
       "14           3300025709_5         8\n",
       "15           3300025213_3         8\n",
       "16           3300025259_2         8\n",
       "17           3300025236_4         8\n",
       "18           3300027812_7         8\n",
       "19           3300002092_1         8\n",
       "20           3300012149_4         8\n",
       "21          3300027826_14         8\n",
       "22           3300025044_4         8\n",
       "23           3300025079_3         8"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "\n",
    "# read table\n",
    "All_edc_Hits_df = pd.read_csv('../data/processed/All_edc_Hits_df_bitscore.csv') \n",
    "All_edc_Hits_TargetAndQuery = All_edc_Hits_df[['target_name', 'query_name']]\n",
    "\n",
    "# link target name to metagenome dataframe\n",
    "\n",
    "# Load a Dataframe with the lookup values for merge protein id to MAGs and merge them\n",
    "TarToMAGs_edc = pd.read_csv('../data/interim/edc_All_TarToMAGsID.csv')\n",
    "\n",
    "# use merge() function to join the MAGsID data\n",
    "edc_hits_TargetAndMAGsID = pd.merge(All_edc_Hits_TargetAndQuery, TarToMAGs_edc, on='target_name', how='left')                                                                               \n",
    "\n",
    "# check the null value\n",
    "print('Any Null in TargetToMAGsID: ', edc_hits_TargetAndMAGsID['MAGsID'].isnull().any())\n",
    "\n",
    "# create the crosstab table (like heatmap)\n",
    "edc_hits_heatmap = pd.crosstab(edc_hits_TargetAndMAGsID['query_name'], edc_hits_TargetAndMAGsID['MAGsID'], dropna=False)\n",
    "edc_hits_heatmap = edc_hits_heatmap.transpose()\n",
    "edc_hits_heatmap\n",
    "# edc_hits_heatmap.to_csv('../data/processed/edc_hits_heatmap_bitscore.csv')\n",
    "\n",
    "# Count the non-zero values in hmm profiles hit row to calculate the number of different HMM profiles that have hits in a given MAG. \n",
    "def count_nonzero(row):\n",
    "    return len(row[row != 0])\n",
    "\n",
    "num_hits = edc_hits_heatmap.apply(count_nonzero, axis=1)\n",
    "edc_hits_heatmap['num_hits'] = num_hits\n",
    "\n",
    "# sort them by hits numer\n",
    "edc_hits_heatmap = edc_hits_heatmap.sort_values(by=\"num_hits\", ascending=False)\n",
    "\n",
    "# extract >8 hmm profiles hits and the necessary hits (edcA、edcB、edcC)\n",
    "edc_hits_FinalFilter =  edc_hits_heatmap[(edc_hits_heatmap['num_hits'] >= 8) & (edc_hits_heatmap['Q_P_I_edcC_13580'] != 0) & (edc_hits_heatmap['Q_P_I_edcA_13525'] != 0) & (edc_hits_heatmap['Q_P_I_edcB_13570'] != 0)]\n",
    "# 34個hits\n",
    "\n",
    "# Reset index and move index column to first position\n",
    "edc_hits_FinalFilter.index.name = None\n",
    "edc_hits_FinalFilter = edc_hits_FinalFilter.reset_index()\n",
    "edc_hits_FinalFilter.insert(0, 'index', edc_hits_FinalFilter.pop('index'))\n",
    "\n",
    "# rename MAGsID\n",
    "edc_hits_FinalFilter = edc_hits_FinalFilter.rename(columns={'index': 'genome_id'})\n",
    "\n",
    "# extract the MAGsID and num_hits column\n",
    "edc_Positive_MAGsID = edc_hits_FinalFilter[['genome_id', 'num_hits']]\n",
    "edc_Positive_MAGsID\n",
    "\n",
    "# done\n",
    "print('done')\n",
    "edc_Positive_MAGsID"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5aaa3c35-e3d4-4478-9b20-490fb4787f7c",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### 2.2.1 連結critical_positive MAGs到 metagenomes 總資料中並進行處理成最株能給R繪製 stacket barchart 的資料格式"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1981e429-bb10-425c-8250-2d9c5786165a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['p__Proteobacteria' 'p__Myxococcota']\n",
      "Any Null in edc_Positive_metagenomes_df:\n",
      " genome_id             False\n",
      "metagenome_id         False\n",
      "num_hits              False\n",
      "Phylum                False\n",
      "Class                 False\n",
      "ecosystem             False\n",
      "ecosystem_category    False\n",
      "longitude             False\n",
      "latitude              False\n",
      "Homologous_cluster    False\n",
      "dtype: bool\n",
      "doen\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>metagenome_id</th>\n",
       "      <th>num_hits</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>ecosystem</th>\n",
       "      <th>ecosystem_category</th>\n",
       "      <th>longitude</th>\n",
       "      <th>latitude</th>\n",
       "      <th>Homologous_cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3300020369_2</td>\n",
       "      <td>3300020369</td>\n",
       "      <td>13</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>-3.2659</td>\n",
       "      <td>-20.3356</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3300026101_4</td>\n",
       "      <td>3300026101</td>\n",
       "      <td>13</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>-55.3027</td>\n",
       "      <td>9.7046</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3300024344_6</td>\n",
       "      <td>3300024344</td>\n",
       "      <td>12</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__UBA9160</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>25.4868</td>\n",
       "      <td>36.5264</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3300027685_11</td>\n",
       "      <td>3300027685</td>\n",
       "      <td>12</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__UBA9160</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>-44.9905</td>\n",
       "      <td>-37.9728</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3300025886_13</td>\n",
       "      <td>3300025886</td>\n",
       "      <td>11</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__UBA9160</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>7.9000</td>\n",
       "      <td>54.1842</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3300017444_33</td>\n",
       "      <td>3300017444</td>\n",
       "      <td>11</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__UBA9160</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Thermal springs</td>\n",
       "      <td>-125.5130</td>\n",
       "      <td>60.1987</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3300006913_5</td>\n",
       "      <td>3300006913</td>\n",
       "      <td>11</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Annelida</td>\n",
       "      <td>12.3946</td>\n",
       "      <td>51.3216</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3300020460_9</td>\n",
       "      <td>3300020460</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>37.2517</td>\n",
       "      <td>23.4217</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3300020463_34</td>\n",
       "      <td>3300020463</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Alphaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>-87.0917</td>\n",
       "      <td>-32.7806</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3300021343_7</td>\n",
       "      <td>3300021343</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>-81.2145</td>\n",
       "      <td>29.6703</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>3300014833_14</td>\n",
       "      <td>3300014833</td>\n",
       "      <td>9</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Engineered</td>\n",
       "      <td>Wastewater</td>\n",
       "      <td>121.5000</td>\n",
       "      <td>31.3000</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>3300020347_2</td>\n",
       "      <td>3300020347</td>\n",
       "      <td>9</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>17.7000</td>\n",
       "      <td>32.1000</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>3300025547_3</td>\n",
       "      <td>3300025547</td>\n",
       "      <td>8</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>-123.5000</td>\n",
       "      <td>48.6000</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>3300025216_2</td>\n",
       "      <td>3300025216</td>\n",
       "      <td>8</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>-133.3150</td>\n",
       "      <td>18.0640</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>3300025709_5</td>\n",
       "      <td>3300025709</td>\n",
       "      <td>8</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>-123.5000</td>\n",
       "      <td>48.6000</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>3300025213_3</td>\n",
       "      <td>3300025213</td>\n",
       "      <td>8</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>63.2500</td>\n",
       "      <td>-27.9800</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>3300025259_2</td>\n",
       "      <td>3300025259</td>\n",
       "      <td>8</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>-17.5322</td>\n",
       "      <td>37.1330</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>3300025236_4</td>\n",
       "      <td>3300025236</td>\n",
       "      <td>8</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>-23.6838</td>\n",
       "      <td>29.1475</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>3300027812_7</td>\n",
       "      <td>3300027812</td>\n",
       "      <td>8</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Alphaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>-128.0900</td>\n",
       "      <td>51.6200</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>3300002092_1</td>\n",
       "      <td>3300002092</td>\n",
       "      <td>8</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Alphaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>13.0225</td>\n",
       "      <td>53.1527</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>3300012149_4</td>\n",
       "      <td>3300012149</td>\n",
       "      <td>8</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Alphaproteobacteria</td>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>-74.5207</td>\n",
       "      <td>39.8720</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>3300027826_14</td>\n",
       "      <td>3300027826</td>\n",
       "      <td>8</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Alphaproteobacteria</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>-76.3402</td>\n",
       "      <td>40.7999</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>3300025044_4</td>\n",
       "      <td>3300025044</td>\n",
       "      <td>8</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>-145.0000</td>\n",
       "      <td>49.9998</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>3300025079_3</td>\n",
       "      <td>3300025079</td>\n",
       "      <td>8</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>-144.9960</td>\n",
       "      <td>50.0005</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        genome_id  metagenome_id  num_hits             Phylum  \\\n",
       "0    3300020369_2     3300020369        13  p__Proteobacteria   \n",
       "1    3300026101_4     3300026101        13  p__Proteobacteria   \n",
       "2    3300024344_6     3300024344        12     p__Myxococcota   \n",
       "3   3300027685_11     3300027685        12     p__Myxococcota   \n",
       "4   3300025886_13     3300025886        11     p__Myxococcota   \n",
       "5   3300017444_33     3300017444        11     p__Myxococcota   \n",
       "6    3300006913_5     3300006913        11  p__Proteobacteria   \n",
       "7    3300020460_9     3300020460        10  p__Proteobacteria   \n",
       "8   3300020463_34     3300020463        10  p__Proteobacteria   \n",
       "9    3300021343_7     3300021343        10  p__Proteobacteria   \n",
       "10  3300014833_14     3300014833         9  p__Proteobacteria   \n",
       "11   3300020347_2     3300020347         9  p__Proteobacteria   \n",
       "12   3300025547_3     3300025547         8  p__Proteobacteria   \n",
       "13   3300025216_2     3300025216         8  p__Proteobacteria   \n",
       "14   3300025709_5     3300025709         8  p__Proteobacteria   \n",
       "15   3300025213_3     3300025213         8  p__Proteobacteria   \n",
       "16   3300025259_2     3300025259         8  p__Proteobacteria   \n",
       "17   3300025236_4     3300025236         8  p__Proteobacteria   \n",
       "18   3300027812_7     3300027812         8  p__Proteobacteria   \n",
       "19   3300002092_1     3300002092         8  p__Proteobacteria   \n",
       "20   3300012149_4     3300012149         8  p__Proteobacteria   \n",
       "21  3300027826_14     3300027826         8  p__Proteobacteria   \n",
       "22   3300025044_4     3300025044         8  p__Proteobacteria   \n",
       "23   3300025079_3     3300025079         8  p__Proteobacteria   \n",
       "\n",
       "                     Class        ecosystem ecosystem_category  longitude  \\\n",
       "0   c__Gammaproteobacteria          Aquatic             Marine    -3.2659   \n",
       "1   c__Gammaproteobacteria          Aquatic             Marine   -55.3027   \n",
       "2               c__UBA9160          Aquatic             Marine    25.4868   \n",
       "3               c__UBA9160          Aquatic             Marine   -44.9905   \n",
       "4               c__UBA9160          Aquatic             Marine     7.9000   \n",
       "5               c__UBA9160          Aquatic    Thermal springs  -125.5130   \n",
       "6   c__Gammaproteobacteria  Host-associated           Annelida    12.3946   \n",
       "7   c__Gammaproteobacteria          Aquatic             Marine    37.2517   \n",
       "8   c__Alphaproteobacteria          Aquatic             Marine   -87.0917   \n",
       "9   c__Gammaproteobacteria          Aquatic             Marine   -81.2145   \n",
       "10  c__Gammaproteobacteria       Engineered         Wastewater   121.5000   \n",
       "11  c__Gammaproteobacteria          Aquatic             Marine    17.7000   \n",
       "12  c__Gammaproteobacteria          Aquatic             Marine  -123.5000   \n",
       "13  c__Gammaproteobacteria          Aquatic             Marine  -133.3150   \n",
       "14  c__Gammaproteobacteria          Aquatic             Marine  -123.5000   \n",
       "15  c__Gammaproteobacteria          Aquatic             Marine    63.2500   \n",
       "16  c__Gammaproteobacteria          Aquatic             Marine   -17.5322   \n",
       "17  c__Gammaproteobacteria          Aquatic             Marine   -23.6838   \n",
       "18  c__Alphaproteobacteria          Aquatic         Freshwater  -128.0900   \n",
       "19  c__Alphaproteobacteria          Aquatic         Freshwater    13.0225   \n",
       "20  c__Alphaproteobacteria  Host-associated              Fungi   -74.5207   \n",
       "21  c__Alphaproteobacteria      Terrestrial               Soil   -76.3402   \n",
       "22  c__Gammaproteobacteria          Aquatic             Marine  -145.0000   \n",
       "23  c__Gammaproteobacteria          Aquatic             Marine  -144.9960   \n",
       "\n",
       "    latitude  Homologous_cluster  \n",
       "0   -20.3356  Proteo_edc_cluster  \n",
       "1     9.7046  Proteo_edc_cluster  \n",
       "2    36.5264  Proteo_edc_cluster  \n",
       "3   -37.9728  Proteo_edc_cluster  \n",
       "4    54.1842  Proteo_edc_cluster  \n",
       "5    60.1987  Proteo_edc_cluster  \n",
       "6    51.3216  Proteo_edc_cluster  \n",
       "7    23.4217  Proteo_edc_cluster  \n",
       "8   -32.7806  Proteo_edc_cluster  \n",
       "9    29.6703  Proteo_edc_cluster  \n",
       "10   31.3000  Proteo_edc_cluster  \n",
       "11   32.1000  Proteo_edc_cluster  \n",
       "12   48.6000  Proteo_edc_cluster  \n",
       "13   18.0640  Proteo_edc_cluster  \n",
       "14   48.6000  Proteo_edc_cluster  \n",
       "15  -27.9800  Proteo_edc_cluster  \n",
       "16   37.1330  Proteo_edc_cluster  \n",
       "17   29.1475  Proteo_edc_cluster  \n",
       "18   51.6200  Proteo_edc_cluster  \n",
       "19   53.1527  Proteo_edc_cluster  \n",
       "20   39.8720  Proteo_edc_cluster  \n",
       "21   40.7999  Proteo_edc_cluster  \n",
       "22   49.9998  Proteo_edc_cluster  \n",
       "23   50.0005  Proteo_edc_cluster  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 需先執行上一個cell\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "# open metagenome csv files\n",
    "metagenmoes_df = pd.read_csv('../data/external/genome_metadata_editForAnalysis_NotReference.csv')\n",
    "\n",
    "# merge positive MAGs with metagenome\n",
    "edc_Positive_metagenomes_df = pd.merge(edc_Positive_MAGsID, metagenmoes_df, on='genome_id', how='left')\n",
    "\n",
    "# print(edc_Positive_metagenomes_df['ecosystem_category'].unique())\n",
    "# print(edc_Positive_metagenomes_df['ecosystem_type'].unique())\n",
    "# display(edc_Positive_metagenomes_df[edc_Positive_metagenomes_df['ecosystem_type'] == 'Bacteria']) #Digestive system, Anaerobic\n",
    "\n",
    "# extract certain column\n",
    "edc_Positive_metagenomes_df = edc_Positive_metagenomes_df[['genome_id', 'metagenome_id', 'taxonomy', 'ecosystem', 'ecosystem_category', 'num_hits', 'longitude', 'latitude']]\n",
    "\n",
    "# extract phylum and class from taxonomy column and expand to new column\n",
    "edc_Positive_metagenomes_df['Phylum'] = edc_Positive_metagenomes_df['taxonomy'].str.extract('(p__\\w+)', expand=True)\n",
    "edc_Positive_metagenomes_df['Class'] = edc_Positive_metagenomes_df['taxonomy'].str.extract('(c__\\w+)', expand=True)\n",
    "\n",
    "# # print non-duplicated values in ecosystem_type than check the lable\n",
    "# print(edc_Positive_metagenomes_df['ecosystem_type'].unique())\n",
    "\n",
    "# # check the None value in ecosystem column Create a Boolean mask to identify NaN values\n",
    "# mask = edc_addEco_df.isna()\n",
    "# edc_addEco_df_nan_rows = edc_addEco_df[mask.any(axis=1)]\n",
    "# edc_addEco_df_nan_rows\n",
    "\n",
    "# remove, add and rearrange column\n",
    "edc_Positive_metagenomes_df.drop('taxonomy', axis=1, inplace=True)\n",
    "edc_Positive_metagenomes_df = edc_Positive_metagenomes_df.reindex(columns=['genome_id', 'metagenome_id', 'num_hits', 'Phylum', 'Class', 'ecosystem', 'ecosystem_category', 'longitude', 'latitude'])\n",
    "edc_Positive_metagenomes_df['Homologous_cluster']='Proteo_edc_cluster'\n",
    "\n",
    "# check phylum data\n",
    "print(edc_Positive_metagenomes_df['Phylum'].unique())\n",
    "print('Any Null in edc_Positive_metagenomes_df:\\n', edc_Positive_metagenomes_df.isnull().any())\n",
    "\n",
    "# write file\n",
    "edc_Positive_metagenomes_df.to_csv('../data/processed/Final/Proteo/edc_PositiveHits_ForR_critical.csv')\n",
    "print('doen')\n",
    "edc_Positive_metagenomes_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25cb0cf7-37c5-4b75-b26e-2b5fa7ef5a8d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# 不執行"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14071a88-c791-4073-9119-7f29a4596724",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 2.3 獲取positive的target name並連結MAGsID，整理出一個類似heatmap的table，再以本先討論的hits數量進行篩選來看差異(>10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7cec1b1a-5d7d-49cb-ae13-c6609e345e84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Any Null:  False\n",
      "done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>query_name</th>\n",
       "      <th>genome_id</th>\n",
       "      <th>num_hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3300020369_2</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3300026101_4</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3300027685_11</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3300024344_6</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3300024431_16</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>3300027838_39</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>3300026091_10</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>3300006083_2</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>3300007519_68</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>3300020195_29</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "query_name      genome_id  num_hits\n",
       "0            3300020369_2        14\n",
       "1            3300026101_4        14\n",
       "2           3300027685_11        13\n",
       "3            3300024344_6        13\n",
       "4           3300024431_16        13\n",
       "..                    ...       ...\n",
       "106         3300027838_39        10\n",
       "107         3300026091_10        10\n",
       "108          3300006083_2        10\n",
       "109         3300007519_68        10\n",
       "110         3300020195_29        10\n",
       "\n",
       "[111 rows x 2 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Got target name and query name\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "All_edc_Hits_df = pd.read_csv('../data/processed/All_edc_Hits_df_bitscore.csv') \n",
    "All_edc_Hits_TargetAndQuery = All_edc_Hits_df[['target_name', 'query_name']]\n",
    "\n",
    "# link target name to metagenome dataframe\n",
    "\n",
    "# Load a Dataframe with the lookup values for merge protein id to MAGs and merge them\n",
    "TarToMAGs_edc = pd.read_csv('../data/interim/edc_All_TarToMAGsID.csv')\n",
    "\n",
    "# use merge() function to join the MAGsID data\n",
    "edc_hits_TargetAndMAGsID = pd.merge(All_edc_Hits_TargetAndQuery, TarToMAGs_edc, on='target_name', how='left')                                                                               \n",
    "\n",
    "# check the null value\n",
    "print('Any Null: ', edc_hits_TargetAndMAGsID['MAGsID'].isnull().any())\n",
    "\n",
    "# create the crosstab table (like heatmap)\n",
    "edc_hits_heatmap = pd.crosstab(edc_hits_TargetAndMAGsID['query_name'], edc_hits_TargetAndMAGsID['MAGsID'], dropna=False)\n",
    "edc_hits_heatmap = edc_hits_heatmap.transpose()\n",
    "edc_hits_heatmap\n",
    "# edc_hits_heatmap.to_csv('../data/processed/edc_hits_heatmap_bitscore.csv')\n",
    "\n",
    "# Count the non-zero values in hmm profiles hit row to calculate the number of different HMM profiles that have hits in a given MAG. \n",
    "def count_nonzero(row):\n",
    "    return len(row[row != 0])\n",
    "\n",
    "num_hits = edc_hits_heatmap.apply(count_nonzero, axis=1)\n",
    "edc_hits_heatmap['num_hits'] = num_hits\n",
    "\n",
    "# sort them by hits numer\n",
    "edc_hits_heatmap = edc_hits_heatmap.sort_values(by=\"num_hits\", ascending=False)\n",
    "\n",
    "# extract >8 hmm profiles hits and the necessary hits (edcA、edcB、edcC)\n",
    "edc_hits_FinalFilter =  edc_hits_heatmap[(edc_hits_heatmap['num_hits'] >= 10)]\n",
    "# 111個hits\n",
    "\n",
    "# Reset index and move index column to first position\n",
    "edc_hits_FinalFilter.index.name = None\n",
    "edc_hits_FinalFilter = edc_hits_FinalFilter.reset_index()\n",
    "edc_hits_FinalFilter.insert(0, 'index', edc_hits_FinalFilter.pop('index'))\n",
    "\n",
    "# rename MAGsID\n",
    "edc_hits_FinalFilter = edc_hits_FinalFilter.rename(columns={'index': 'genome_id'})\n",
    "\n",
    "# extract the MAGsID and num_hits column\n",
    "edc_Positive_MAGsID = edc_hits_FinalFilter[['genome_id', 'num_hits']]\n",
    "edc_Positive_MAGsID\n",
    "\n",
    "# done\n",
    "print('done')\n",
    "edc_Positive_MAGsID"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d78a89d6-1e09-428a-8296-73b8a3cd8c49",
   "metadata": {},
   "source": [
    "##### 2.3.1 連結 10 Hits positive (>10) MAGs到 metagenomes 總資料中並進行處理成最株能給R繪製 stacket barchart 的資料格式"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ea7d7cae-c247-4c62-8d6b-d9e9d5f0263b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Marine' 'Fungi' 'Thermal spring' 'Soil' 'Freshwater' 'Annelida' 'Plant'\n",
      " 'Cnidaria' 'Wastewater' 'Lab enrichment' 'Sediment']\n",
      "doen\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>num_hits</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>ecosystem</th>\n",
       "      <th>ecosystem_type</th>\n",
       "      <th>Homologous_cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3300020369_2</td>\n",
       "      <td>14</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3300026101_4</td>\n",
       "      <td>14</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3300027685_11</td>\n",
       "      <td>13</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__UBA9160</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3300024344_6</td>\n",
       "      <td>13</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__UBA9160</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3300024431_16</td>\n",
       "      <td>13</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__UBA9160</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>3300027838_39</td>\n",
       "      <td>10</td>\n",
       "      <td>p__UBP10</td>\n",
       "      <td>c__GR</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>3300026091_10</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>3300006083_2</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>3300007519_68</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>3300020195_29</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         genome_id  num_hits               Phylum                   Class  \\\n",
       "0     3300020369_2        14    p__Proteobacteria  c__Gammaproteobacteria   \n",
       "1     3300026101_4        14    p__Proteobacteria  c__Gammaproteobacteria   \n",
       "2    3300027685_11        13       p__Myxococcota              c__UBA9160   \n",
       "3     3300024344_6        13       p__Myxococcota              c__UBA9160   \n",
       "4    3300024431_16        13       p__Myxococcota              c__UBA9160   \n",
       "..             ...       ...                  ...                     ...   \n",
       "106  3300027838_39        10             p__UBP10                   c__GR   \n",
       "107  3300026091_10        10  p__Actinobacteriota       c__Acidimicrobiia   \n",
       "108   3300006083_2        10  p__Actinobacteriota       c__Acidimicrobiia   \n",
       "109  3300007519_68        10  p__Actinobacteriota       c__Acidimicrobiia   \n",
       "110  3300020195_29        10    p__Proteobacteria  c__Gammaproteobacteria   \n",
       "\n",
       "    ecosystem ecosystem_type  Homologous_cluster  \n",
       "0     Aquatic         Marine  Proteo_edc_cluster  \n",
       "1     Aquatic         Marine  Proteo_edc_cluster  \n",
       "2     Aquatic         Marine  Proteo_edc_cluster  \n",
       "3     Aquatic         Marine  Proteo_edc_cluster  \n",
       "4     Aquatic         Marine  Proteo_edc_cluster  \n",
       "..        ...            ...                 ...  \n",
       "106   Aquatic         Marine  Proteo_edc_cluster  \n",
       "107   Aquatic         Marine  Proteo_edc_cluster  \n",
       "108   Aquatic         Marine  Proteo_edc_cluster  \n",
       "109   Aquatic     Freshwater  Proteo_edc_cluster  \n",
       "110   Aquatic     Freshwater  Proteo_edc_cluster  \n",
       "\n",
       "[111 rows x 7 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 需先執行上一個cell\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "# open metagenome csv files\n",
    "metagenmoes_df = pd.read_csv('../data/external/Paper_genome_metadata.csv')\n",
    "metagenmoes_df = metagenmoes_df.rename(columns={'ecosystem': 'taxonomy'})\n",
    "\n",
    "# merge positive MAGs with metagenome\n",
    "edc_Positive_metagenomes_df = pd.merge(edc_Positive_MAGsID, metagenmoes_df, on='genome_id', how='left')\n",
    "\n",
    "# print(edc_Positive_metagenomes_df['ecosystem_category'].unique())\n",
    "# print(edc_Positive_metagenomes_df['ecosystem_type'].unique())\n",
    "# display(edc_Positive_metagenomes_df[edc_Positive_metagenomes_df['ecosystem_type'] == 'Bacteria']) #Digestive system, Anaerobic\n",
    "\n",
    "# extract certain column\n",
    "edc_Positive_metagenomes_df = edc_Positive_metagenomes_df[['genome_id', 'taxonomy', 'ecosystem_type', 'num_hits']]\n",
    "\n",
    "# extract phylum and class from taxonomy column and expand to new column\n",
    "edc_Positive_metagenomes_df['Phylum'] = edc_Positive_metagenomes_df['taxonomy'].str.extract('(p__\\w+)', expand=True)\n",
    "edc_Positive_metagenomes_df['Class'] = edc_Positive_metagenomes_df['taxonomy'].str.extract('(c__\\w+)', expand=True)\n",
    "\n",
    "# # print non-duplicated values in ecosystem_type than check the lable\n",
    "# print(edc_Positive_metagenomes_df['ecosystem_type'].unique())\n",
    "\n",
    "# Load a EcosystemToEco_type dataframe for lookup\n",
    "ForEcoLookup = pd.read_csv(\"../data/interim/EcosystemToEco_type.csv\")\n",
    "\n",
    "# merge them with ecosystem_type\n",
    "edc_addEco_df = pd.merge(edc_Positive_metagenomes_df, ForEcoLookup, on='ecosystem_type', how='left')\n",
    "\n",
    "# # check the None value in ecosystem column Create a Boolean mask to identify NaN values\n",
    "# mask = edc_addEco_df.isna()\n",
    "# edc_addEco_df_nan_rows = edc_addEco_df[mask.any(axis=1)]\n",
    "# edc_addEco_df_nan_rows\n",
    "\n",
    "# !!!! change the ecosystem_type value name\n",
    "edc_addEco_df['ecosystem_type'] = edc_addEco_df['ecosystem_type'].replace({'Mycelium': 'Fungi', 'Rhizosphere': 'Plant', 'Industrial wastewater': 'Wastewater', \n",
    "                                                                          \"Nutrient removal\": \"Wastewater\", \"Rhizoplane\": \"Plant\", \"Activated Sludge\": \"Wastewater\",\"Defined media\": \"Lab enrichment\"})\n",
    "print(edc_addEco_df['ecosystem_type'].unique())\n",
    "\n",
    "# remove, add and rearrange column\n",
    "edc_addEco_df.drop('taxonomy', axis=1, inplace=True)\n",
    "edc_addEco_df= edc_addEco_df.reindex(columns=['genome_id', 'num_hits', 'Phylum', 'Class', 'ecosystem', 'ecosystem_type'])\n",
    "edc_addEco_df['Homologous_cluster']='Proteo_edc_cluster'\n",
    "\n",
    "# check phylum data\n",
    "# print(edc_addEco_df['Phylum'].unique())\n",
    "\n",
    "# write file\n",
    "edc_addEco_df.to_csv('../data/processed/Final/Proteo/edc_PositiveHits_ForR_10Hits.csv')\n",
    "print('doen')\n",
    "edc_addEco_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec5aad36-8850-446f-88c8-1611eb1ca310",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 第二部分: 舊版參考前人標準 (e-value coverage)，不含有bit score\n",
    "DOI: https://doi.org/10.1128/mBio.00166-16"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e893b51c-f4fd-4473-91dd-250be147f237",
   "metadata": {},
   "source": [
    "### 1. 用一致的e-value與coverage進行篩選"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8fd15e5f-f477-4efa-868b-f787c358a7d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_name</th>\n",
       "      <th>accession</th>\n",
       "      <th>tlen</th>\n",
       "      <th>query_name</th>\n",
       "      <th>accession2</th>\n",
       "      <th>qlen</th>\n",
       "      <th>E-value</th>\n",
       "      <th>score</th>\n",
       "      <th>bias</th>\n",
       "      <th>num_domains_index</th>\n",
       "      <th>...</th>\n",
       "      <th>bias2</th>\n",
       "      <th>hmm_from</th>\n",
       "      <th>hmm_to</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>ali_to</th>\n",
       "      <th>env_from</th>\n",
       "      <th>env_to</th>\n",
       "      <th>acc</th>\n",
       "      <th>description</th>\n",
       "      <th>coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3300020463.a:Ga0211676_10006390_8</td>\n",
       "      <td>-</td>\n",
       "      <td>1162</td>\n",
       "      <td>Q_P_I_edcC_13580</td>\n",
       "      <td>-</td>\n",
       "      <td>1145</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1994.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1143</td>\n",
       "      <td>12</td>\n",
       "      <td>1142</td>\n",
       "      <td>12</td>\n",
       "      <td>1144</td>\n",
       "      <td>0.99</td>\n",
       "      <td>-</td>\n",
       "      <td>0.973322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3300020369.a:Ga0211709_10000577_18</td>\n",
       "      <td>-</td>\n",
       "      <td>1185</td>\n",
       "      <td>Q_P_I_edcC_13580</td>\n",
       "      <td>-</td>\n",
       "      <td>1145</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1991.4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1144</td>\n",
       "      <td>20</td>\n",
       "      <td>1167</td>\n",
       "      <td>20</td>\n",
       "      <td>1168</td>\n",
       "      <td>0.98</td>\n",
       "      <td>-</td>\n",
       "      <td>0.968776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3300014833.a:Ga0119870_1000327_7</td>\n",
       "      <td>-</td>\n",
       "      <td>1168</td>\n",
       "      <td>Q_P_I_edcC_13580</td>\n",
       "      <td>-</td>\n",
       "      <td>1145</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1871.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1144</td>\n",
       "      <td>6</td>\n",
       "      <td>1154</td>\n",
       "      <td>6</td>\n",
       "      <td>1155</td>\n",
       "      <td>0.98</td>\n",
       "      <td>-</td>\n",
       "      <td>0.983733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3300024431.a:Ga0209988_10007392_4</td>\n",
       "      <td>-</td>\n",
       "      <td>1169</td>\n",
       "      <td>Q_P_I_edcC_13580</td>\n",
       "      <td>-</td>\n",
       "      <td>1145</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1735.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1143</td>\n",
       "      <td>14</td>\n",
       "      <td>1162</td>\n",
       "      <td>14</td>\n",
       "      <td>1164</td>\n",
       "      <td>0.98</td>\n",
       "      <td>-</td>\n",
       "      <td>0.982891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3300017444.a:Ga0185300_10000551_81</td>\n",
       "      <td>-</td>\n",
       "      <td>1185</td>\n",
       "      <td>Q_P_I_edcC_13580</td>\n",
       "      <td>-</td>\n",
       "      <td>1145</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1721.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1144</td>\n",
       "      <td>37</td>\n",
       "      <td>1182</td>\n",
       "      <td>37</td>\n",
       "      <td>1183</td>\n",
       "      <td>0.98</td>\n",
       "      <td>-</td>\n",
       "      <td>0.967089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>3300020403.a:Ga0211532_10000028_18</td>\n",
       "      <td>-</td>\n",
       "      <td>149</td>\n",
       "      <td>Q_P_13530</td>\n",
       "      <td>-</td>\n",
       "      <td>144</td>\n",
       "      <td>1.000000e-25</td>\n",
       "      <td>102.6</td>\n",
       "      <td>0.4</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>6</td>\n",
       "      <td>134</td>\n",
       "      <td>16</td>\n",
       "      <td>135</td>\n",
       "      <td>13</td>\n",
       "      <td>147</td>\n",
       "      <td>0.87</td>\n",
       "      <td>-</td>\n",
       "      <td>0.805369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372</th>\n",
       "      <td>3300020417.a:Ga0211528_10001553_12</td>\n",
       "      <td>-</td>\n",
       "      <td>149</td>\n",
       "      <td>Q_P_13530</td>\n",
       "      <td>-</td>\n",
       "      <td>144</td>\n",
       "      <td>1.000000e-25</td>\n",
       "      <td>102.6</td>\n",
       "      <td>0.4</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>6</td>\n",
       "      <td>134</td>\n",
       "      <td>16</td>\n",
       "      <td>135</td>\n",
       "      <td>13</td>\n",
       "      <td>147</td>\n",
       "      <td>0.87</td>\n",
       "      <td>-</td>\n",
       "      <td>0.805369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>3300020418.a:Ga0211557_10001031_4</td>\n",
       "      <td>-</td>\n",
       "      <td>149</td>\n",
       "      <td>Q_P_13530</td>\n",
       "      <td>-</td>\n",
       "      <td>144</td>\n",
       "      <td>1.000000e-25</td>\n",
       "      <td>102.6</td>\n",
       "      <td>0.4</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>6</td>\n",
       "      <td>134</td>\n",
       "      <td>16</td>\n",
       "      <td>135</td>\n",
       "      <td>13</td>\n",
       "      <td>147</td>\n",
       "      <td>0.87</td>\n",
       "      <td>-</td>\n",
       "      <td>0.805369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>3300020453.a:Ga0211550_10000005_13</td>\n",
       "      <td>-</td>\n",
       "      <td>149</td>\n",
       "      <td>Q_P_13530</td>\n",
       "      <td>-</td>\n",
       "      <td>144</td>\n",
       "      <td>1.000000e-25</td>\n",
       "      <td>102.6</td>\n",
       "      <td>0.4</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>6</td>\n",
       "      <td>134</td>\n",
       "      <td>16</td>\n",
       "      <td>135</td>\n",
       "      <td>13</td>\n",
       "      <td>147</td>\n",
       "      <td>0.87</td>\n",
       "      <td>-</td>\n",
       "      <td>0.805369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>3300020460.a:Ga0211486_10004041_3</td>\n",
       "      <td>-</td>\n",
       "      <td>149</td>\n",
       "      <td>Q_P_13530</td>\n",
       "      <td>-</td>\n",
       "      <td>144</td>\n",
       "      <td>1.000000e-25</td>\n",
       "      <td>102.6</td>\n",
       "      <td>0.4</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>6</td>\n",
       "      <td>134</td>\n",
       "      <td>16</td>\n",
       "      <td>135</td>\n",
       "      <td>13</td>\n",
       "      <td>147</td>\n",
       "      <td>0.87</td>\n",
       "      <td>-</td>\n",
       "      <td>0.805369</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>599784 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            target_name accession  tlen        query_name  \\\n",
       "0     3300020463.a:Ga0211676_10006390_8         -  1162  Q_P_I_edcC_13580   \n",
       "1    3300020369.a:Ga0211709_10000577_18         -  1185  Q_P_I_edcC_13580   \n",
       "2      3300014833.a:Ga0119870_1000327_7         -  1168  Q_P_I_edcC_13580   \n",
       "3     3300024431.a:Ga0209988_10007392_4         -  1169  Q_P_I_edcC_13580   \n",
       "4    3300017444.a:Ga0185300_10000551_81         -  1185  Q_P_I_edcC_13580   \n",
       "..                                  ...       ...   ...               ...   \n",
       "371  3300020403.a:Ga0211532_10000028_18         -   149         Q_P_13530   \n",
       "372  3300020417.a:Ga0211528_10001553_12         -   149         Q_P_13530   \n",
       "373   3300020418.a:Ga0211557_10001031_4         -   149         Q_P_13530   \n",
       "374  3300020453.a:Ga0211550_10000005_13         -   149         Q_P_13530   \n",
       "375   3300020460.a:Ga0211486_10004041_3         -   149         Q_P_13530   \n",
       "\n",
       "    accession2  qlen       E-value   score  bias num_domains_index  ... bias2  \\\n",
       "0            -  1145  0.000000e+00  1994.3   0.0                 1  ...   0.0   \n",
       "1            -  1145  0.000000e+00  1991.4   0.0                 1  ...   0.0   \n",
       "2            -  1145  0.000000e+00  1871.7   0.0                 1  ...   0.0   \n",
       "3            -  1145  0.000000e+00  1735.0   0.0                 1  ...   0.0   \n",
       "4            -  1145  0.000000e+00  1721.2   0.0                 1  ...   0.0   \n",
       "..         ...   ...           ...     ...   ...               ...  ...   ...   \n",
       "371          -   144  1.000000e-25   102.6   0.4                 1  ...   0.4   \n",
       "372          -   144  1.000000e-25   102.6   0.4                 1  ...   0.4   \n",
       "373          -   144  1.000000e-25   102.6   0.4                 1  ...   0.4   \n",
       "374          -   144  1.000000e-25   102.6   0.4                 1  ...   0.4   \n",
       "375          -   144  1.000000e-25   102.6   0.4                 1  ...   0.4   \n",
       "\n",
       "     hmm_from  hmm_to  ali_from  ali_to env_from env_to   acc description  \\\n",
       "0           1    1143        12    1142       12   1144  0.99           -   \n",
       "1           1    1144        20    1167       20   1168  0.98           -   \n",
       "2           1    1144         6    1154        6   1155  0.98           -   \n",
       "3           1    1143        14    1162       14   1164  0.98           -   \n",
       "4           1    1144        37    1182       37   1183  0.98           -   \n",
       "..        ...     ...       ...     ...      ...    ...   ...         ...   \n",
       "371         6     134        16     135       13    147  0.87           -   \n",
       "372         6     134        16     135       13    147  0.87           -   \n",
       "373         6     134        16     135       13    147  0.87           -   \n",
       "374         6     134        16     135       13    147  0.87           -   \n",
       "375         6     134        16     135       13    147  0.87           -   \n",
       "\n",
       "     coverage  \n",
       "0    0.973322  \n",
       "1    0.968776  \n",
       "2    0.983733  \n",
       "3    0.982891  \n",
       "4    0.967089  \n",
       "..        ...  \n",
       "371  0.805369  \n",
       "372  0.805369  \n",
       "373  0.805369  \n",
       "374  0.805369  \n",
       "375  0.805369  \n",
       "\n",
       "[599784 rows x 24 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "# Proteo MAGs reference filtering\n",
    "# Define the directory that contains the \"domtblout\" files, 這個genome protein fasta要把discription的文字刪除，因有空格會影響read\n",
    "domtblout_dir = \"../data/raw/Proteo_HMM_MAGs_domtblout/\"\n",
    "\n",
    "# Define the names of the hmmsearches (without the file extension) delet aedC,Q,R.\n",
    "hmmsearches = ['MAGs_P_13530', 'MAGs_P_13535', 'MAGs_P_13540', 'MAGs_P_13545', 'MAGs_P_13550', 'MAGs_P_13555', 'MAGs_P_13560', 'MAGs_P_13565', 'MAGs_P_13575',\n",
    "               'MAGs_P_13585', 'MAGs_P_13590', 'MAGs_P_13595', 'MAGs_P_I_edcA_13525', 'MAGs_P_I_edcB_13570', 'MAGs_P_I_edcC_13580']\n",
    "\n",
    "# create a dataframe for a all hits \n",
    "columns = [\"target_name\", \"accession\", \"tlen\", \"query_name\", \"accession2\", \"qlen\", \"E-value\", \"score\", \"bias\",\n",
    "           \"num_domains_index\", \"num_domains_total\", \"c-Evalue\", \"i-Evalue\", \"score2\", \"bias2\", \"hmm_from\",\n",
    "           \"hmm_to\", \"ali_from\", \"ali_to\", \"env_from\", \"env_to\", \"acc\", \"description\"]\n",
    "\n",
    "All_edc_Hits_df = pd.DataFrame(columns=columns)\n",
    "\n",
    "MAGs_Hits = {}\n",
    "MAGs_Hits_name = []\n",
    "\n",
    "\n",
    "# Loop over the hmmsearches and parse the corresponding \"domtblout\" file\n",
    "for hmmsearch in hmmsearches:\n",
    "    # Load the \"domtblout\" file into a pandas DataFrame\n",
    "    file_path = os.path.join(domtblout_dir, hmmsearch + \".domtblout\")\n",
    "    df = pd.read_csv(file_path, comment=\"#\", sep='\\s+', header=None)    \n",
    "    # Assign column names to the DataFrame\n",
    "    df.columns = [\"target_name\", \"accession\", \"tlen\", \"query_name\", \"accession2\", \"qlen\", \"E-value\", \"score\", \"bias\",\n",
    "                  \"num_domains_index\", \"num_domains_total\", \"c-Evalue\", \"i-Evalue\", \"score2\", \"bias2\", \"hmm_from\", \"hmm_to\", \"ali_from\", \"ali_to\",\n",
    "                  \"env_from\", \"env_to\", \"acc\", \"description\"]\n",
    "    # Calculate the coverage for each hit\n",
    "    df[\"coverage\"] = (df[\"ali_to\"] - df[\"ali_from\"] + 1) / df[\"tlen\"]\n",
    "\n",
    "    # Filter the DataFrame by E-value and coverage, and sort by bit-score\n",
    "    significant_hits_Ref = df[(df[\"E-value\"] <= 1e-25) & (df[\"coverage\"] > 0.30)]\n",
    "\n",
    "    # Extract Target nmae and store it in the dictionary\n",
    "    if not significant_hits_Ref.empty:\n",
    "        MAGs_Hits_name = significant_hits_Ref[\"target_name\"].tolist()\n",
    "        MAGs_Hits[hmmsearch] = MAGs_Hits_name\n",
    "    else:\n",
    "        MAGs_Hits_name = None\n",
    "        MAGs_Hits[hmmsearch] = MAGs_Hits_name\n",
    "    \n",
    "    # add hits table to a df\n",
    "    All_edc_Hits_df = pd.concat([significant_hits_Ref, All_edc_Hits_df], axis=0)\n",
    "\n",
    "All_edc_Hits_df.to_csv('../data/processed/All_edc_Hits_df_refrence.csv')    \n",
    "\n",
    "# done\n",
    "print('done')\n",
    "All_edc_Hits_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a3f5b25-5ab0-4948-8cd3-89303bcc402f",
   "metadata": {},
   "source": [
    "### 2. 獲取 positive的 target name 並連結 MAGs 並整理出一個類似heatmap的table，再以> 10 hits數量進行篩選，並有三個必須基因的限制(edcA, edcB, edcC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a14be602-ba8a-4a30-b43e-b2c607ae68e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Any Null:  False\n",
      "done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>query_name</th>\n",
       "      <th>genome_id</th>\n",
       "      <th>num_hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3300020369_2</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3300012940_39</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3300020193_17</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3300024344_6</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3300027951_10</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>3300025617_10</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>3300027851_25</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288</th>\n",
       "      <td>3300025590_3</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>3300025594_3</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>290</th>\n",
       "      <td>3300020266_5</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>291 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "query_name      genome_id  num_hits\n",
       "0            3300020369_2        15\n",
       "1           3300012940_39        15\n",
       "2           3300020193_17        15\n",
       "3            3300024344_6        15\n",
       "4           3300027951_10        14\n",
       "..                    ...       ...\n",
       "286         3300025617_10        10\n",
       "287         3300027851_25        10\n",
       "288          3300025590_3        10\n",
       "289          3300025594_3        10\n",
       "290          3300020266_5        10\n",
       "\n",
       "[291 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "\n",
    "# read table\n",
    "All_edc_Hits_df = pd.read_csv('../data/processed/All_edc_Hits_df_refrence.csv') \n",
    "All_edc_Hits_TargetAndQuery = All_edc_Hits_df[['target_name', 'query_name']]\n",
    "\n",
    "# link target name to metagenome dataframe\n",
    "\n",
    "# Load a Dataframe with the lookup values for merge protein id to MAGs and merge them\n",
    "TarToMAGs_edc = pd.read_csv('../data/interim/edc_All_TarToMAGsID.csv')\n",
    "\n",
    "# use merge() function to join the MAGsID data\n",
    "edc_hits_TargetAndMAGsID = pd.merge(All_edc_Hits_TargetAndQuery, TarToMAGs_edc, on='target_name', how='left')                                                                               \n",
    "\n",
    "# check the null value\n",
    "print('Any Null: ', edc_hits_TargetAndMAGsID['MAGsID'].isnull().any())\n",
    "\n",
    "# create the crosstab table (like heatmap)\n",
    "edc_hits_heatmap = pd.crosstab(edc_hits_TargetAndMAGsID['query_name'], edc_hits_TargetAndMAGsID['MAGsID'], dropna=False)\n",
    "edc_hits_heatmap = edc_hits_heatmap.transpose()\n",
    "edc_hits_heatmap\n",
    "# edc_hits_heatmap.to_csv('../data/processed/edc_hits_heatmap_bitscore.csv')\n",
    "\n",
    "# Count the non-zero values in hmm profiles hit row to calculate the number of different HMM profiles that have hits in a given MAG. \n",
    "def count_nonzero(row):\n",
    "    return len(row[row != 0])\n",
    "\n",
    "num_hits = edc_hits_heatmap.apply(count_nonzero, axis=1)\n",
    "edc_hits_heatmap['num_hits'] = num_hits\n",
    "\n",
    "# sort them by hits numer\n",
    "edc_hits_heatmap = edc_hits_heatmap.sort_values(by=\"num_hits\", ascending=False)\n",
    "\n",
    "# extract >8 hmm profiles hits and the necessary hits (edcA、edcB、edcC)\n",
    "edc_hits_FinalFilter =  edc_hits_heatmap[(edc_hits_heatmap['num_hits'] >= 10) & (edc_hits_heatmap['Q_P_I_edcC_13580'] != 0) & (edc_hits_heatmap['Q_P_I_edcA_13525'] != 0) & (edc_hits_heatmap['Q_P_I_edcB_13570'] != 0)]\n",
    "# \n",
    "\n",
    "# Reset index and move index column to first position\n",
    "edc_hits_FinalFilter.index.name = None\n",
    "edc_hits_FinalFilter = edc_hits_FinalFilter.reset_index()\n",
    "edc_hits_FinalFilter.insert(0, 'index', edc_hits_FinalFilter.pop('index'))\n",
    "\n",
    "# rename MAGsID\n",
    "edc_hits_FinalFilter = edc_hits_FinalFilter.rename(columns={'index': 'genome_id'})\n",
    "\n",
    "# extract the MAGsID and num_hits column\n",
    "edc_Positive_MAGsID = edc_hits_FinalFilter[['genome_id', 'num_hits']]\n",
    "edc_Positive_MAGsID\n",
    "\n",
    "# done\n",
    "print('done')\n",
    "edc_Positive_MAGsID"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff9706be",
   "metadata": {},
   "source": [
    "### 3. 連結 Reference_positive MAGs到 metagenomes 總資料中並進行處理成最株能給R繪製 stacket barchart 的資料格式"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "94966641",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Marine' 'Soil' 'Freshwater' 'Porifera' 'Thermal spring' 'Wastewater'\n",
      " 'Plant' 'Annelida' 'Built environment' 'Deep subsurface' 'Fungi'\n",
      " 'Non-marine Saline and Alkaline' 'Lab enrichment' 'Bioreactor'\n",
      " 'Bioremediation' 'Algae' 'Sediment' 'Arthropoda' 'Mammal' 'Solid waste'\n",
      " 'Insecta']\n",
      "doen\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>num_hits</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>ecosystem</th>\n",
       "      <th>ecosystem_type</th>\n",
       "      <th>Homologous_cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3300020369_2</td>\n",
       "      <td>15</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3300012940_39</td>\n",
       "      <td>15</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3300020193_17</td>\n",
       "      <td>15</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__UBA9160</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3300024344_6</td>\n",
       "      <td>15</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__UBA9160</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3300027951_10</td>\n",
       "      <td>14</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Porifera</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>3300025617_10</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>3300027851_25</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288</th>\n",
       "      <td>3300025590_3</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Actinobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>3300025594_3</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Actinobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>290</th>\n",
       "      <td>3300020266_5</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Alphaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Proteo_edc_cluster</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>291 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         genome_id  num_hits               Phylum                   Class  \\\n",
       "0     3300020369_2        15    p__Proteobacteria  c__Gammaproteobacteria   \n",
       "1    3300012940_39        15    p__Proteobacteria  c__Gammaproteobacteria   \n",
       "2    3300020193_17        15       p__Myxococcota              c__UBA9160   \n",
       "3     3300024344_6        15       p__Myxococcota              c__UBA9160   \n",
       "4    3300027951_10        14  p__Actinobacteriota       c__Acidimicrobiia   \n",
       "..             ...       ...                  ...                     ...   \n",
       "286  3300025617_10        10    p__Proteobacteria  c__Gammaproteobacteria   \n",
       "287  3300027851_25        10    p__Proteobacteria  c__Gammaproteobacteria   \n",
       "288   3300025590_3        10  p__Actinobacteriota       c__Actinobacteria   \n",
       "289   3300025594_3        10  p__Actinobacteriota       c__Actinobacteria   \n",
       "290   3300020266_5        10    p__Proteobacteria  c__Alphaproteobacteria   \n",
       "\n",
       "           ecosystem ecosystem_type  Homologous_cluster  \n",
       "0            Aquatic         Marine  Proteo_edc_cluster  \n",
       "1        Terrestrial           Soil  Proteo_edc_cluster  \n",
       "2            Aquatic     Freshwater  Proteo_edc_cluster  \n",
       "3            Aquatic         Marine  Proteo_edc_cluster  \n",
       "4    Host-associated       Porifera  Proteo_edc_cluster  \n",
       "..               ...            ...                 ...  \n",
       "286          Aquatic         Marine  Proteo_edc_cluster  \n",
       "287          Aquatic     Freshwater  Proteo_edc_cluster  \n",
       "288          Aquatic         Marine  Proteo_edc_cluster  \n",
       "289          Aquatic         Marine  Proteo_edc_cluster  \n",
       "290          Aquatic         Marine  Proteo_edc_cluster  \n",
       "\n",
       "[291 rows x 7 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 需先執行上一個cell\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "# open metagenome csv files\n",
    "metagenmoes_df = pd.read_csv('../data/external/Paper_genome_metadata.csv')\n",
    "metagenmoes_df = metagenmoes_df.rename(columns={'ecosystem': 'taxonomy'})\n",
    "\n",
    "# merge positive MAGs with metagenome\n",
    "edc_Positive_metagenomes_df = pd.merge(edc_Positive_MAGsID, metagenmoes_df, on='genome_id', how='left')\n",
    "\n",
    "# print(edc_Positive_metagenomes_df['ecosystem_category'].unique())\n",
    "# print(edc_Positive_metagenomes_df['ecosystem_type'].unique())\n",
    "# display(edc_Positive_metagenomes_df[edc_Positive_metagenomes_df['ecosystem_type'] == 'Bacteria']) #Anaerobic\n",
    "\n",
    "# extract certain column\n",
    "edc_Positive_metagenomes_df = edc_Positive_metagenomes_df[['genome_id', 'taxonomy', 'ecosystem_type', 'num_hits']]\n",
    "\n",
    "# extract phylum and class from taxonomy column and expand to new column\n",
    "edc_Positive_metagenomes_df['Phylum'] = edc_Positive_metagenomes_df['taxonomy'].str.extract('(p__\\w+)', expand=True)\n",
    "edc_Positive_metagenomes_df['Class'] = edc_Positive_metagenomes_df['taxonomy'].str.extract('(c__\\w+)', expand=True)\n",
    "\n",
    "# # print non-duplicated values in ecosystem_type than check the lable\n",
    "# print(edc_Positive_metagenomes_df['ecosystem_type'].unique())\n",
    "\n",
    "# Load a EcosystemToEco_type dataframe for lookup\n",
    "ForEcoLookup = pd.read_csv(\"../data/interim/EcosystemToEco_type.csv\")\n",
    "\n",
    "# merge them with ecosystem_type\n",
    "edc_addEco_df = pd.merge(edc_Positive_metagenomes_df, ForEcoLookup, on='ecosystem_type', how='left')\n",
    "\n",
    "# # check the None value in ecosystem column Create a Boolean mask to identify NaN values\n",
    "# mask = edc_addEco_df.isna()\n",
    "# edc_addEco_df_nan_rows = edc_addEco_df[mask.any(axis=1)]\n",
    "# edc_addEco_df_nan_rows\n",
    "\n",
    "# !!!! change the ecosystem_type value name\n",
    "edc_addEco_df['ecosystem_type'] = edc_addEco_df['ecosystem_type'].replace({\"Activated Sludge\": \"Wastewater\", \"Nutrient removal\": \"Wastewater\", \"Rhizoplane\": \"Plant\", 'Mycelium': 'Fungi', 'Peat moss': \"Plant\",\n",
    "                                                                           'Rhizosphere': 'Plant', \"Defined media\": \"Lab enrichment\", 'Continuous culture': 'Bioreactor',\n",
    "                                                                           'Geologic': 'Soil', 'Tetrachloroethylene and derivatives': 'Bioremediation', 'Red algae': 'Algae',\n",
    "                                                                           'Green algae': 'Algae', 'Ant dump': 'Arthropoda', 'Simulated communities (microbial mixture)': 'Lab enrichment',\n",
    "                                                                           'Anaerobic digestor': 'Wastewater', 'Composting': 'Solid waste', 'Phylloplane': \"Plant\", 'City': 'Built environment'})\n",
    "print(edc_addEco_df['ecosystem_type'].unique())\n",
    "\n",
    "# remove, add and rearrange column\n",
    "edc_addEco_df.drop('taxonomy', axis=1, inplace=True)\n",
    "edc_addEco_df= edc_addEco_df.reindex(columns=['genome_id', 'num_hits', 'Phylum', 'Class', 'ecosystem', 'ecosystem_type'])\n",
    "edc_addEco_df['Homologous_cluster']='Proteo_edc_cluster'\n",
    "\n",
    "# check phylum data\n",
    "# print(edc_addEco_df['Phylum'].unique())\n",
    "\n",
    "# write file\n",
    "edc_addEco_df.to_csv('../data/processed/Final/Proteo/edc_PositiveHits_ForR_Reference.csv')\n",
    "print('doen')\n",
    "edc_addEco_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9525ca0c-afd6-48cf-86e6-54f0df237701",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# 補充部分"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "089d7841-d648-4f5c-98ef-23c66458178c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # \n",
    "# TarToMAGs_edc2560 = pd.read_csv('../data/interim/edc_135_2560_TargetMAGsID.csv')\n",
    "# TarToMAGs_edc6595 = pd.read_csv('../data/interim/edc_135_6595_TargetMAGsID.csv')\n",
    "# TarToMAGs_edc_All = pd.concat([TarToMAGs_edc2560, TarToMAGs_edc6595], axis=0)\n",
    "# TarToMAGs_edc_All = TarToMAGs_edc_All.drop_duplicates(subset='target_name')\n",
    "\n",
    "# TarToMAGs_edc_All.to_csv('../data/interim/edc_All_TarToMAGsID.csv')\n",
    "# # TarToMAGs_aed_All.to_csv('../data/interim/aed_TargetToMAGsID_all.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f9d8c55a-e186-48c6-90e6-73e098912d3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>query_name</th>\n",
       "      <th>genome_id</th>\n",
       "      <th>num_hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [genome_id, num_hits]\n",
       "Index: []"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edc_Positive_MAGsID[edc_Positive_MAGsID['genome_id'] == '3300026863_7']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e8481270",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>ecosystem_type</th>\n",
       "      <th>num_hits</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>3300026195_9</td>\n",
       "      <td>d__Bacteria;p__Actinobacteriota;c__Actinobacte...</td>\n",
       "      <td>Anaerobic</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Actinobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>3300026194_9</td>\n",
       "      <td>d__Bacteria;p__Actinobacteriota;c__Actinobacte...</td>\n",
       "      <td>Anaerobic</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Actinobacteria</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        genome_id                                           taxonomy  \\\n",
       "280  3300026195_9  d__Bacteria;p__Actinobacteriota;c__Actinobacte...   \n",
       "281  3300026194_9  d__Bacteria;p__Actinobacteriota;c__Actinobacte...   \n",
       "\n",
       "    ecosystem_type  num_hits               Phylum              Class  \n",
       "280      Anaerobic        10  p__Actinobacteriota  c__Actinobacteria  \n",
       "281      Anaerobic        10  p__Actinobacteriota  c__Actinobacteria  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edc_Positive_metagenomes_df[edc_Positive_metagenomes_df['ecosystem_type'] == 'Anaerobic']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "d76b71f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Aquatic' 'Terrestrial' 'Host-associated' 'Wastewater' 'Plant'\n",
      " 'Digestive system' 'Built environment' 'Fungi' 'Lab enrichment'\n",
      " 'Bioreactor' 'Anaerobic' 'Bioremediation' 'Algae' 'Arthropoda' 'Modeled'\n",
      " 'Solid waste' 'Unclassified']\n",
      "['Marine' 'Soil' 'Freshwater' 'Porifera' 'Thermal spring'\n",
      " 'Activated Sludge' 'Nutrient removal' 'Rhizoplane' 'Annelida' 'City'\n",
      " 'Deep subsurface' 'Mycelium' 'Peat moss' 'Non-marine Saline and Alkaline'\n",
      " 'Rhizosphere' 'Defined media' 'Continuous culture' 'Bioreactor'\n",
      " 'Geologic' 'Tetrachloroethylene and derivatives' 'Red algae'\n",
      " 'Green algae' 'Sediment' 'Ant dump'\n",
      " 'Simulated communities (microbial mixture)' 'Mammal' 'Anaerobic digestor'\n",
      " 'Composting' 'Arthropoda' 'Insecta' 'Phylloplane']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>num_hits</th>\n",
       "      <th>metagenome_id</th>\n",
       "      <th>genome_length</th>\n",
       "      <th>num_contigs</th>\n",
       "      <th>n50</th>\n",
       "      <th>num_16s</th>\n",
       "      <th>num_5s</th>\n",
       "      <th>num_23s</th>\n",
       "      <th>num_trna</th>\n",
       "      <th>...</th>\n",
       "      <th>contamination</th>\n",
       "      <th>quality_score</th>\n",
       "      <th>mimag_quality</th>\n",
       "      <th>otu_id</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>ecosystem_category</th>\n",
       "      <th>ecosystem_type</th>\n",
       "      <th>habitat</th>\n",
       "      <th>longitude</th>\n",
       "      <th>latitude</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [genome_id, num_hits, metagenome_id, genome_length, num_contigs, n50, num_16s, num_5s, num_23s, num_trna, completeness, contamination, quality_score, mimag_quality, otu_id, taxonomy, ecosystem_category, ecosystem_type, habitat, longitude, latitude]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 21 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>num_hits</th>\n",
       "      <th>metagenome_id</th>\n",
       "      <th>genome_length</th>\n",
       "      <th>num_contigs</th>\n",
       "      <th>n50</th>\n",
       "      <th>num_16s</th>\n",
       "      <th>num_5s</th>\n",
       "      <th>num_23s</th>\n",
       "      <th>num_trna</th>\n",
       "      <th>...</th>\n",
       "      <th>contamination</th>\n",
       "      <th>quality_score</th>\n",
       "      <th>mimag_quality</th>\n",
       "      <th>otu_id</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>ecosystem_category</th>\n",
       "      <th>ecosystem_type</th>\n",
       "      <th>habitat</th>\n",
       "      <th>longitude</th>\n",
       "      <th>latitude</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [genome_id, num_hits, metagenome_id, genome_length, num_contigs, n50, num_16s, num_5s, num_23s, num_trna, completeness, contamination, quality_score, mimag_quality, otu_id, taxonomy, ecosystem_category, ecosystem_type, habitat, longitude, latitude]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 21 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# open metagenome csv files\n",
    "metagenmoes_df = pd.read_csv('../data/external/Paper_genome_metadata.csv')\n",
    "metagenmoes_df = metagenmoes_df.rename(columns={'ecosystem': 'taxonomy'})\n",
    "\n",
    "# merge positive MAGs with metagenome\n",
    "edc_Positive_metagenomes_df = pd.merge(edc_Positive_MAGsID, metagenmoes_df, on='genome_id', how='left')\n",
    "\n",
    "print(edc_Positive_metagenomes_df['ecosystem_category'].unique())\n",
    "print(edc_Positive_metagenomes_df['ecosystem_type'].unique())\n",
    "display(edc_Positive_metagenomes_df[edc_Positive_metagenomes_df['ecosystem_type'] == 'Unclassified']) #Anaerobic\n",
    "\n",
    "edc_Positive_metagenomes_df[edc_Positive_metagenomes_df['ecosystem_type'] == 'Anaerobic']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f16c9250",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>query_name</th>\n",
       "      <th>genome_id</th>\n",
       "      <th>num_hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>272</th>\n",
       "      <td>3300026863_7</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "query_name     genome_id  num_hits\n",
       "272         3300026863_7         9"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edc_Positive_MAGsID[edc_Positive_MAGsID['genome_id'] == '3300026863_7']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5199f50d-ac76-4c00-aba3-6b34305f9927",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
