{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c6743c28-5992-4381-908c-ec41e4e22b3d",
   "metadata": {},
   "source": [
    "# Note4_統整資料並製作出最終表來給R作圖"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f310335d-6ea7-487d-af62-2a9c8eb1f45d",
   "metadata": {
    "tags": []
   },
   "source": [
    "# 1. 整理 aed 與 edc 的 loose positive MAGs 資料"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9c6273e7-82f7-449b-88f1-78f1ccc82361",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B50_final Row (1): 609\n",
      "B50_final Row (2): 609\n",
      "Any Null in B50_HMMFinal_df:\n",
      " genome_id             False\n",
      "num_hits              False\n",
      "taxonomy              False\n",
      "ecosystem             False\n",
      "ecosystem_category    False\n",
      "Homologous_cluster    False\n",
      "longitude              True\n",
      "latitude               True\n",
      "dtype: bool\n",
      "['Actinobacteria' 'Alphaproteobacteria' 'Chloroflexi'\n",
      " 'Gammaproteobacteria' 'Myxococcia' 'Thermodesulfobacterota'\n",
      " 'Unclassified Bacteria' 'Other Bacterial Phylum']\n",
      "['Aquatic' 'Engineered' 'Host-associated' 'Terrestrial']\n",
      "['Freshwater' 'Marine' 'Non-marine Saline and Alkaline' 'Sediment'\n",
      " 'Thermal springs' 'Bioreactor' 'Built environment' 'Lab enrichment'\n",
      " 'Lab synthesis' 'Modeled' 'Solid waste' 'Wastewater' 'Annelida' 'Fungi'\n",
      " 'Invertebrates' 'Mammals' 'Microbial' 'Plants' 'Porifera'\n",
      " 'Deep subsurface' 'Peat' 'Soil']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ecosystem</th>\n",
       "      <th>ecosystem_category</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>NumberOfMAGs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Chloroflexi</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Myxococcia</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Chloroflexi</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Other Bacterial Phylum</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Unclassified Bacteria</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      ecosystem ecosystem_category                taxonomy  NumberOfMAGs\n",
       "0       Aquatic         Freshwater          Actinobacteria            52\n",
       "1       Aquatic         Freshwater     Alphaproteobacteria            17\n",
       "2       Aquatic         Freshwater             Chloroflexi             8\n",
       "3       Aquatic         Freshwater     Gammaproteobacteria            12\n",
       "4       Aquatic         Freshwater              Myxococcia             1\n",
       "..          ...                ...                     ...           ...\n",
       "57  Terrestrial               Soil     Alphaproteobacteria            10\n",
       "58  Terrestrial               Soil             Chloroflexi             2\n",
       "59  Terrestrial               Soil     Gammaproteobacteria             5\n",
       "60  Terrestrial               Soil  Other Bacterial Phylum             1\n",
       "61  Terrestrial               Soil   Unclassified Bacteria            19\n",
       "\n",
       "[62 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "# Read table\n",
    "aed_final_df = pd.read_csv('../data/processed/Final/Actino/aed_PositiveHits_ForR_loose.csv')\n",
    "edc_final_df = pd.read_csv('../data/processed/Final/Proteo/edc_PositiveHits_ForR_loose.csv')\n",
    "\n",
    "# concatenate aed and edc final data\n",
    "B50_HMMFinal_df = pd.concat([aed_final_df, edc_final_df])\n",
    "\n",
    "# remove duplicated and update the 'Both_clsuter' in homologous_cluster column\n",
    "# 找重複 (有2)\n",
    "B50_duplicated = B50_HMMFinal_df['genome_id'].value_counts()\n",
    "\n",
    "# set new column in DataFrame with value counts\n",
    "B50_HMMFinal_df['num_duplicated'] = B50_HMMFinal_df['genome_id'].map(B50_duplicated)\n",
    "\n",
    "# if number is 2 (duplicated) change the homnologous column to 'Both_cluster'\n",
    "B50_HMMFinal_df.loc[B50_HMMFinal_df['num_duplicated'] == 2 , 'Homologous_cluster'] = 'Both_cluster'\n",
    "\n",
    "# remove duplicated row\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.drop_duplicates(subset=['genome_id'])\n",
    "print('B50_final Row (1):', len(B50_HMMFinal_df.index))\n",
    "\n",
    "# remove, add and rearrange column\n",
    "B50_HMMFinal_df.drop('num_duplicated', axis=1, inplace=True)\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.iloc[:, 1:]\n",
    "\n",
    "# open file for vlookup from phylum to final taxonomy ID\n",
    "PhylumToTaxonomy = pd.read_csv('../data/interim/PhylumToTaxonomy.csv')\n",
    "ClassToTaxnomy = pd.read_csv('../data/interim/ClassToTaxonomy.csv')\n",
    "\n",
    "# merge with phylum column\n",
    "B50_HMMFinal_df1 = pd.merge(B50_HMMFinal_df, PhylumToTaxonomy, on='Phylum', how='left')\n",
    "\n",
    "# merge with Class\n",
    "B50_HMMFinal_df2 = pd.merge(B50_HMMFinal_df, ClassToTaxnomy, on='Class', how='left')\n",
    "\n",
    "# concatenate the two dataframes\n",
    "B50_HMMFinal_df = pd.concat([B50_HMMFinal_df1, B50_HMMFinal_df2], axis=0, ignore_index=True)\n",
    "\n",
    "# drop rows with NaN values in taxonomy column\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.dropna(subset=['taxonomy'])\n",
    "print('B50_final Row (2):', len(B50_HMMFinal_df.index))\n",
    "\n",
    "# remove and rerrange column\n",
    "B50_HMMFinal_df.drop(columns=['Phylum', 'Class'], inplace=True)\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.reindex(columns=['genome_id', 'num_hits', 'taxonomy', 'ecosystem', 'ecosystem_category', 'Homologous_cluster', 'longitude', 'latitude'])\n",
    "B50_HMMFinal_df.to_csv('../data/processed/Final/Combined/Combined_PositiveHits_loose.csv')\n",
    "print('Any Null in B50_HMMFinal_df:\\n', B50_HMMFinal_df.isnull().any())\n",
    "\n",
    "# calculate hits number\n",
    "B50_HMMFinal_number = B50_HMMFinal_df.groupby(['ecosystem', 'ecosystem_category', 'taxonomy']).size().reset_index(name='NumberOfMAGs')\n",
    "B50_HMMFinal_number.to_csv('../data/processed/Final/R/B50_HMMFinal_loose_ForBarChart.csv', index=False)\n",
    "\n",
    "print(B50_HMMFinal_number['taxonomy'].unique())\n",
    "print(B50_HMMFinal_number['ecosystem'].unique())\n",
    "print(B50_HMMFinal_number['ecosystem_category'].unique())\n",
    "display(B50_HMMFinal_number)\n",
    "\n",
    "# mmerge reference datasheet to get a full table for readers and Output\n",
    "\n",
    "# read files\n",
    "genome_metadata = pd.read_csv('../data/external/Reference_Data/genome_metadata.csv')\n",
    "genmoe_metadata_edit = pd.read_csv('../data/external/Reference_Data/genome_metadata_edit.csv')\n",
    "genome_metadata_editForAnalysis_NotReference = pd.read_csv('../data/external/genome_metadata_editForAnalysis_NotReference.csv')\n",
    "\n",
    "# Got the specific columns in B50_HMMFinal_df\n",
    "B50_HMMFinal_df_HH = B50_HMMFinal_df[['genome_id', 'num_hits', 'Homologous_cluster']]\n",
    "\n",
    "# merge table\n",
    "M_genome_metadata = pd.merge(B50_HMMFinal_df_HH, genome_metadata, on='genome_id', how='left')\n",
    "M_genmoe_metadata_edit = pd.merge(B50_HMMFinal_df_HH, genmoe_metadata_edit, on='genome_id', how='left')\n",
    "M_genome_metadata_editForAnalysis_NotReference = pd.merge(B50_HMMFinal_df_HH, genome_metadata_editForAnalysis_NotReference, on='genome_id', how='left')\n",
    "\n",
    "# save files\n",
    "M_genome_metadata.to_csv('../data/processed/Final/ForReader/PositiveHits_loose_genome_metadata.csv', index=False)\n",
    "M_genmoe_metadata_edit.to_csv('../data/processed/Final/ForReader/PositiveHits_loose_genmoe_metadata_edit.csv', index=False)\n",
    "M_genome_metadata_editForAnalysis_NotReference.to_csv('../data/processed/Final/ForReader/PositiveHits_loose_genome_metadata_editForAnalysis_NotReference.csv', index=False)\n",
    "\n",
    "# done\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf571f63-59c9-4b88-bdff-6d183c2669c0",
   "metadata": {
    "tags": []
   },
   "source": [
    "# 2. loose hits distribution of world Map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f5332860-13a0-40e0-b3d9-eb2a168a43e3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B50_final Row (1): 609\n",
      "B50_final Row (2): 609\n",
      "There are 57 row(s) with null values.\n",
      "metagenomic_id with missing coordinate: \n",
      " <bound method Series.unique of 17      3300026210\n",
      "46      3300012014\n",
      "57      3300026262\n",
      "58      3300026200\n",
      "59      3300026199\n",
      "61      3300026202\n",
      "63      3300026206\n",
      "64      3300026261\n",
      "67      3300026256\n",
      "68      3300026193\n",
      "69      3300026209\n",
      "89      3300026212\n",
      "92      3300012005\n",
      "120     3300026267\n",
      "128     3300026265\n",
      "131     3300026108\n",
      "141     3300026262\n",
      "142     3300026192\n",
      "144     3300026262\n",
      "149     3300026092\n",
      "161     3300026091\n",
      "215     3300027855\n",
      "235     3300026213\n",
      "237     3300026211\n",
      "238     3300026208\n",
      "341     3300026079\n",
      "345     3300026087\n",
      "346     3300026188\n",
      "349     3300026092\n",
      "355     3300026108\n",
      "358     3300026093\n",
      "360     3300026119\n",
      "365     3300006902\n",
      "381     3300026200\n",
      "424     3300026279\n",
      "425     3300026269\n",
      "426     3300026268\n",
      "427     3300026266\n",
      "429     3300026265\n",
      "431     3300026264\n",
      "432     3300026263\n",
      "433     3300026260\n",
      "434     3300026260\n",
      "435     3300026260\n",
      "436     3300026258\n",
      "437     3300026257\n",
      "438     3300026256\n",
      "441     3300026254\n",
      "443     3300026213\n",
      "451     3300026292\n",
      "476     3300026321\n",
      "898     3300027790\n",
      "910     3300027820\n",
      "962     3300026093\n",
      "1039    3300026264\n",
      "1080    3300011906\n",
      "1088    3300026079\n",
      "Name: metagenome_id, dtype: int64>\n",
      "Any Null in edc_Positive_metagenomes_df:\n",
      " Ecosystem             False\n",
      "Ecosystem_Category    False\n",
      "Latitude              False\n",
      "Longitude             False\n",
      "count                 False\n",
      "dtype: bool\n",
      "done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ecosystem</th>\n",
       "      <th>Ecosystem_Category</th>\n",
       "      <th>Latitude</th>\n",
       "      <th>Longitude</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>-78.0836</td>\n",
       "      <td>164.284</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>-78.0544</td>\n",
       "      <td>164.021</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>-78.0482</td>\n",
       "      <td>164.113</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>-78.0267</td>\n",
       "      <td>163.820</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>-77.7140</td>\n",
       "      <td>162.445</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>68.6139</td>\n",
       "      <td>-149.314</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>71.2838</td>\n",
       "      <td>-156.598</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>71.2906</td>\n",
       "      <td>-156.789</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>71.2990</td>\n",
       "      <td>-156.610</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>71.2999</td>\n",
       "      <td>-156.610</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>230 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Ecosystem Ecosystem_Category  Latitude  Longitude  count\n",
       "0        Aquatic         Freshwater  -78.0836    164.284      1\n",
       "1        Aquatic         Freshwater  -78.0544    164.021      1\n",
       "2        Aquatic         Freshwater  -78.0482    164.113      1\n",
       "3        Aquatic         Freshwater  -78.0267    163.820      1\n",
       "4        Aquatic         Freshwater  -77.7140    162.445      8\n",
       "..           ...                ...       ...        ...    ...\n",
       "225  Terrestrial               Soil   68.6139   -149.314      1\n",
       "226  Terrestrial               Soil   71.2838   -156.598      2\n",
       "227  Terrestrial               Soil   71.2906   -156.789      1\n",
       "228  Terrestrial               Soil   71.2990   -156.610      2\n",
       "229  Terrestrial               Soil   71.2999   -156.610      8\n",
       "\n",
       "[230 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "# Read table\n",
    "aed_final_df = pd.read_csv('../data/processed/Final/Actino/aed_PositiveHits_ForR_loose.csv')\n",
    "edc_final_df = pd.read_csv('../data/processed/Final/Proteo/edc_PositiveHits_ForR_loose.csv')\n",
    "\n",
    "# concatenate aed and edc final data\n",
    "B50_HMMFinal_df = pd.concat([aed_final_df, edc_final_df])\n",
    "\n",
    "# remove duplicated and update the 'Both_clsuter' in homologous_cluster column\n",
    "# 找重複 (有2)\n",
    "B50_duplicated = B50_HMMFinal_df['genome_id'].value_counts()\n",
    "\n",
    "# set new column in DataFrame with value counts\n",
    "B50_HMMFinal_df['num_duplicated'] = B50_HMMFinal_df['genome_id'].map(B50_duplicated)\n",
    "\n",
    "# if number is 2 (duplicated) change the homnologous column to 'Both_cluster'\n",
    "B50_HMMFinal_df.loc[B50_HMMFinal_df['num_duplicated'] == 2 , 'Homologous_cluster'] = 'Both_cluster'\n",
    "\n",
    "# remove duplicated row\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.drop_duplicates(subset=['genome_id'])\n",
    "print('B50_final Row (1):', len(B50_HMMFinal_df.index))\n",
    "\n",
    "# remove, add and rearrange column\n",
    "B50_HMMFinal_df.drop('num_duplicated', axis=1, inplace=True)\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.iloc[:, 1:]\n",
    "\n",
    "# open file for vlookup from phylum to final taxonomy ID\n",
    "PhylumToTaxonomy = pd.read_csv('../data/interim/PhylumToTaxonomy.csv')\n",
    "ClassToTaxnomy = pd.read_csv('../data/interim/ClassToTaxonomy.csv')\n",
    "\n",
    "# merge with phylum column\n",
    "B50_HMMFinal_df1 = pd.merge(B50_HMMFinal_df, PhylumToTaxonomy, on='Phylum', how='left')\n",
    "\n",
    "# merge with Class\n",
    "B50_HMMFinal_df2 = pd.merge(B50_HMMFinal_df, ClassToTaxnomy, on='Class', how='left')\n",
    "\n",
    "# concatenate the two dataframes\n",
    "B50_HMMFinal_df = pd.concat([B50_HMMFinal_df1, B50_HMMFinal_df2], axis=0, ignore_index=True)\n",
    "\n",
    "# drop rows with NaN values in taxonomy column\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.dropna(subset=['taxonomy'])\n",
    "print('B50_final Row (2):', len(B50_HMMFinal_df.index))\n",
    "\n",
    "# remove and rerrange column\n",
    "B50_HMMFinal_df.drop(columns=['Phylum', 'Class', 'num_hits', 'taxonomy', 'Homologous_cluster'], inplace=True)\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.reindex(columns=['genome_id', 'metagenome_id', 'ecosystem', 'ecosystem_category', 'longitude', 'latitude'])\n",
    "\n",
    "# check missing row of coordinates\n",
    "missing_rows = B50_HMMFinal_df[B50_HMMFinal_df[['longitude', 'latitude']].isnull().any(axis=1)]\n",
    "print(f'There are {missing_rows.shape[0]} row(s) with null values.')\n",
    "\n",
    "# Got the metagenomic_id with missing coordinate\n",
    "print('metagenomic_id with missing coordinate: \\n',missing_rows['metagenome_id'].unique )\n",
    "\n",
    "# remove unnecessary columns\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.drop(['genome_id', 'metagenome_id'], axis=1)\n",
    "\n",
    "# rename columns\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.rename(columns={'ecosystem': 'Ecosystem', 'ecosystem_category': 'Ecosystem_Category', 'longitude': 'Longitude', 'latitude': 'Latitude'})\n",
    "\n",
    "# reset index and write files\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.reset_index(drop=True)\n",
    "\n",
    "# assuming your data is in a DataFrame called df\n",
    "B50_HMMFinal_df_grouped = B50_HMMFinal_df.groupby(['Ecosystem', 'Ecosystem_Category', 'Latitude', 'Longitude']).size().reset_index(name='count')\n",
    "B50_HMMFinal_df_grouped.to_csv('../data/processed/Final/R/B50_HMMFinal_loose_ForMaps.csv', index=False)\n",
    "\n",
    "print('Any Null in edc_Positive_metagenomes_df:\\n', B50_HMMFinal_df_grouped.isnull().any())\n",
    "print('done')\n",
    "B50_HMMFinal_df_grouped\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca4a10c1-31fb-413e-bc85-defb530c1dc0",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## 不使用了:整理 aed 與 edc 的 critical positive MAGs 資料"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9fcd6ca0-5b25-4490-802f-ddab2785c398",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B50_final Row (1): 35\n",
      "B50_final Row (2): 35\n",
      "Any Null in B50_HMMFinal_df:\n",
      " genome_id             False\n",
      "num_hits              False\n",
      "taxonomy              False\n",
      "ecosystem             False\n",
      "ecosystem_category    False\n",
      "Homologous_cluster    False\n",
      "longitude             False\n",
      "latitude              False\n",
      "dtype: bool\n",
      "['p_Actinobacteriota_c_Acidimicrobiia'\n",
      " 'p_Proteobacteria_c_Alphaproteobacteria' 'p_Myxococcota'\n",
      " 'p_Proteobacteria_c_Gammaproteobacteria']\n",
      "['Aquatic' 'Engineered' 'Host-associated' 'Terrestrial']\n",
      "['Freshwater' 'Marine' 'Thermal springs' 'Wastewater' 'Annelida' 'Fungi'\n",
      " 'Plants' 'Soil']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ecosystem</th>\n",
       "      <th>ecosystem_category</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>NumberOfMAGs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>p_Actinobacteriota_c_Acidimicrobiia</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>p_Myxococcota</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Thermal springs</td>\n",
       "      <td>p_Myxococcota</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Engineered</td>\n",
       "      <td>Wastewater</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Annelida</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Plants</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          ecosystem ecosystem_category  \\\n",
       "0           Aquatic         Freshwater   \n",
       "1           Aquatic         Freshwater   \n",
       "2           Aquatic             Marine   \n",
       "3           Aquatic             Marine   \n",
       "4           Aquatic             Marine   \n",
       "5           Aquatic    Thermal springs   \n",
       "6        Engineered         Wastewater   \n",
       "7   Host-associated           Annelida   \n",
       "8   Host-associated              Fungi   \n",
       "9   Host-associated             Plants   \n",
       "10      Terrestrial               Soil   \n",
       "\n",
       "                                  taxonomy  NumberOfMAGs  \n",
       "0      p_Actinobacteriota_c_Acidimicrobiia             1  \n",
       "1   p_Proteobacteria_c_Alphaproteobacteria             3  \n",
       "2                            p_Myxococcota             3  \n",
       "3   p_Proteobacteria_c_Alphaproteobacteria             1  \n",
       "4   p_Proteobacteria_c_Gammaproteobacteria            16  \n",
       "5                            p_Myxococcota             1  \n",
       "6   p_Proteobacteria_c_Gammaproteobacteria             1  \n",
       "7   p_Proteobacteria_c_Gammaproteobacteria             1  \n",
       "8   p_Proteobacteria_c_Alphaproteobacteria             2  \n",
       "9   p_Proteobacteria_c_Alphaproteobacteria             2  \n",
       "10  p_Proteobacteria_c_Alphaproteobacteria             4  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "# Read table\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "aed_final_df = pd.read_csv('../data/processed/Final/Actino/aed_PositiveHits_ForR_critical.csv')\n",
    "edc_final_df = pd.read_csv('../data/processed/Final/Proteo/edc_PositiveHits_ForR_critical.csv')\n",
    "\n",
    "# concatenate aed and edc final data\n",
    "B50_HMMFinal_df = pd.concat([aed_final_df, edc_final_df])\n",
    "\n",
    "# remove duplicated and update the 'Both_clsuter' in homologous_cluster column\n",
    "# 找重複 (有2)\n",
    "B50_duplicated = B50_HMMFinal_df['genome_id'].value_counts()\n",
    "\n",
    "# set new column in DataFrame with value counts\n",
    "B50_HMMFinal_df['num_duplicated'] = B50_HMMFinal_df['genome_id'].map(B50_duplicated)\n",
    "\n",
    "# if number is 2 (duplicated) change the homnologous column to 'Both_cluster'\n",
    "B50_HMMFinal_df.loc[B50_HMMFinal_df['num_duplicated'] == 2 , 'Homologous_cluster'] = 'Both_cluster'\n",
    "\n",
    "# remove duplicated row\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.drop_duplicates(subset=['genome_id'])\n",
    "print('B50_final Row (1):', len(B50_HMMFinal_df.index))\n",
    "\n",
    "# remove, add and rearrange column\n",
    "B50_HMMFinal_df.drop('num_duplicated', axis=1, inplace=True)\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.iloc[:, 1:]\n",
    "\n",
    "# open file for vlookup from phylum to final taxonomy ID\n",
    "PhylumToTaxonomy = pd.read_csv('../data/interim/PhylumToTaxonomy.csv')\n",
    "ClassToTaxnomy = pd.read_csv('../data/interim/ClassToTaxonomy.csv')\n",
    "\n",
    "# merge with phylum column\n",
    "B50_HMMFinal_df1 = pd.merge(B50_HMMFinal_df, PhylumToTaxonomy, on='Phylum', how='left')\n",
    "\n",
    "\n",
    "# merge with Class\n",
    "B50_HMMFinal_df2 = pd.merge(B50_HMMFinal_df, ClassToTaxnomy, on='Class', how='left')\n",
    "\n",
    "# concatenate the two dataframes\n",
    "B50_HMMFinal_df = pd.concat([B50_HMMFinal_df1, B50_HMMFinal_df2], axis=0, ignore_index=True)\n",
    "\n",
    "# drop rows with NaN values in taxonomy column\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.dropna(subset=['taxonomy'])\n",
    "print('B50_final Row (2):', len(B50_HMMFinal_df.index))\n",
    "\n",
    "# remove and rerrange column\n",
    "B50_HMMFinal_df.drop(columns=['Phylum', 'Class'], inplace=True)\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.reindex(columns=['genome_id', 'num_hits', 'taxonomy', 'ecosystem', 'ecosystem_category', 'Homologous_cluster', 'longitude', 'latitude'])\n",
    "B50_HMMFinal_df.to_csv('../data/processed/Final/Combined/Combined_PositiveHits_critical.csv')\n",
    "print('Any Null in B50_HMMFinal_df:\\n', B50_HMMFinal_df.isnull().any())\n",
    "\n",
    "# calculate hits number\n",
    "B50_HMMFinal_number = B50_HMMFinal_df.groupby(['ecosystem', 'ecosystem_category', 'taxonomy']).size().reset_index(name='NumberOfMAGs')\n",
    "B50_HMMFinal_number.to_csv('../data/processed/Final/R/B50_HMMFinal_critical_ForBarChart.csv', index=False)\n",
    "\n",
    "print(B50_HMMFinal_number['taxonomy'].unique())\n",
    "print(B50_HMMFinal_number['ecosystem'].unique())\n",
    "print(B50_HMMFinal_number['ecosystem_category'].unique())\n",
    "display(B50_HMMFinal_number)\n",
    "\n",
    "# mmerge reference datasheet to get a full table for readers and Output\n",
    "\n",
    "# read files\n",
    "genome_metadata = pd.read_csv('../data/external/Reference_Data/genome_metadata.csv')\n",
    "genmoe_metadata_edit = pd.read_csv('../data/external/Reference_Data/genome_metadata_edit.csv')\n",
    "genome_metadata_editForAnalysis_NotReference = pd.read_csv('../data/external/genome_metadata_editForAnalysis_NotReference.csv')\n",
    "\n",
    "# Got the specific columns in B50_HMMFinal_df\n",
    "B50_HMMFinal_df_HH = B50_HMMFinal_df[['genome_id', 'num_hits', 'Homologous_cluster']]\n",
    "\n",
    "# merge table\n",
    "M_genome_metadata = pd.merge(B50_HMMFinal_df_HH, genome_metadata, on='genome_id', how='left')\n",
    "M_genmoe_metadata_edit = pd.merge(B50_HMMFinal_df_HH, genmoe_metadata_edit, on='genome_id', how='left')\n",
    "M_genome_metadata_editForAnalysis_NotReference = pd.merge(B50_HMMFinal_df_HH, genome_metadata_editForAnalysis_NotReference, on='genome_id', how='left')\n",
    "\n",
    "# save files\n",
    "M_genome_metadata.to_csv('../data/processed/Final/ForReader/PositiveHits_critical_genome_metadata.csv', index=False)\n",
    "M_genmoe_metadata_edit.to_csv('../data/processed/Final/ForReader/PositiveHits_critical_genmoe_metadata_edit.csv', index=False)\n",
    "M_genome_metadata_editForAnalysis_NotReference.to_csv('../data/processed/Final/ForReader/PositiveHits_critical_genome_metadata_editForAnalysis_NotReference.csv', index=False)\n",
    "\n",
    "# done\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f0fc303-1cb4-4bac-bc7e-f8d61a3725d7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# 不執行"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c5eddc2-6d9c-4ac1-ae85-e43a5c2dfc13",
   "metadata": {},
   "source": [
    "## 3. 整理 aed 與 edc 的 10 Hits positive MAGs 資料"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "371a3111-872d-4f8a-9622-491111486fc5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B50_final Row (1): 114\n",
      "B50_final Row (2): 114\n",
      "['p_Actinobacteriota_c_Acidimicrobiia' 'p_Myxococcota'\n",
      " 'p_Proteobacteria_c_Gammaproteobacteria'\n",
      " 'p_Proteobacteria_c_Alphaproteobacteria' 'p_Unclassified Bacteria'\n",
      " 'p_Actinobacteriota_c_Actinobacteria']\n",
      "['Aquatic' 'Engineered' 'Host-associated' 'Terrestrial']\n",
      "['Freshwater' 'Marine' 'Sediment' 'Thermal spring' 'Lab enrichment'\n",
      " 'Wastewater' 'Annelida' 'Cnidaria' 'Fungi' 'Plant' 'Soil']\n",
      "done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ecosystem</th>\n",
       "      <th>ecosystem_type</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>NumberOfMAGs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>p_Actinobacteriota_c_Acidimicrobiia</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>p_Myxococcota</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>p_Actinobacteriota_c_Acidimicrobiia</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>p_Myxococcota</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>p_Unclassified Bacteria</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>p_Actinobacteriota_c_Actinobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Thermal spring</td>\n",
       "      <td>p_Actinobacteriota_c_Acidimicrobiia</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Thermal spring</td>\n",
       "      <td>p_Myxococcota</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Thermal spring</td>\n",
       "      <td>p_Unclassified Bacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Engineered</td>\n",
       "      <td>Lab enrichment</td>\n",
       "      <td>p_Actinobacteriota_c_Actinobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Engineered</td>\n",
       "      <td>Wastewater</td>\n",
       "      <td>p_Actinobacteriota_c_Acidimicrobiia</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Engineered</td>\n",
       "      <td>Wastewater</td>\n",
       "      <td>p_Myxococcota</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Engineered</td>\n",
       "      <td>Wastewater</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Annelida</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Cnidaria</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Plant</td>\n",
       "      <td>p_Actinobacteriota_c_Actinobacteria</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>p_Actinobacteriota_c_Acidimicrobiia</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>p_Unclassified Bacteria</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          ecosystem  ecosystem_type                                taxonomy  \\\n",
       "0           Aquatic      Freshwater     p_Actinobacteriota_c_Acidimicrobiia   \n",
       "1           Aquatic      Freshwater                           p_Myxococcota   \n",
       "2           Aquatic      Freshwater  p_Proteobacteria_c_Gammaproteobacteria   \n",
       "3           Aquatic          Marine     p_Actinobacteriota_c_Acidimicrobiia   \n",
       "4           Aquatic          Marine                           p_Myxococcota   \n",
       "5           Aquatic          Marine  p_Proteobacteria_c_Alphaproteobacteria   \n",
       "6           Aquatic          Marine  p_Proteobacteria_c_Gammaproteobacteria   \n",
       "7           Aquatic          Marine                 p_Unclassified Bacteria   \n",
       "8           Aquatic        Sediment     p_Actinobacteriota_c_Actinobacteria   \n",
       "9           Aquatic  Thermal spring     p_Actinobacteriota_c_Acidimicrobiia   \n",
       "10          Aquatic  Thermal spring                           p_Myxococcota   \n",
       "11          Aquatic  Thermal spring                 p_Unclassified Bacteria   \n",
       "12       Engineered  Lab enrichment     p_Actinobacteriota_c_Actinobacteria   \n",
       "13       Engineered      Wastewater     p_Actinobacteriota_c_Acidimicrobiia   \n",
       "14       Engineered      Wastewater                           p_Myxococcota   \n",
       "15       Engineered      Wastewater  p_Proteobacteria_c_Gammaproteobacteria   \n",
       "16  Host-associated        Annelida  p_Proteobacteria_c_Gammaproteobacteria   \n",
       "17  Host-associated        Cnidaria  p_Proteobacteria_c_Gammaproteobacteria   \n",
       "18  Host-associated           Fungi  p_Proteobacteria_c_Gammaproteobacteria   \n",
       "19  Host-associated           Plant     p_Actinobacteriota_c_Actinobacteria   \n",
       "20      Terrestrial            Soil     p_Actinobacteriota_c_Acidimicrobiia   \n",
       "21      Terrestrial            Soil  p_Proteobacteria_c_Gammaproteobacteria   \n",
       "22      Terrestrial            Soil                 p_Unclassified Bacteria   \n",
       "\n",
       "    NumberOfMAGs  \n",
       "0             15  \n",
       "1              1  \n",
       "2              2  \n",
       "3             31  \n",
       "4             10  \n",
       "5              1  \n",
       "6              9  \n",
       "7              4  \n",
       "8              1  \n",
       "9              1  \n",
       "10             1  \n",
       "11             1  \n",
       "12             1  \n",
       "13             2  \n",
       "14             2  \n",
       "15             1  \n",
       "16             1  \n",
       "17             1  \n",
       "18             1  \n",
       "19             2  \n",
       "20            18  \n",
       "21             1  \n",
       "22             7  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read table\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "aed_final_df = pd.read_csv('../data/processed/Final/Actino/aed_PositiveHits_ForR_10Hits.csv')\n",
    "edc_final_df = pd.read_csv('../data/processed/Final/Proteo/edc_PositiveHits_ForR_10Hits.csv')\n",
    "\n",
    "# concatenate aed and edc final data\n",
    "B50_HMMFinal_df = pd.concat([aed_final_df, edc_final_df])\n",
    "\n",
    "# remove duplicated and update the 'Both_clsuter' in homologous_cluster column\n",
    "# 找重複 (有2)\n",
    "B50_duplicated = B50_HMMFinal_df['genome_id'].value_counts()\n",
    "\n",
    "# set new column in DataFrame with value counts\n",
    "B50_HMMFinal_df['num_duplicated'] = B50_HMMFinal_df['genome_id'].map(B50_duplicated)\n",
    "\n",
    "# if number is 2 (duplicated) change the homnologous column to 'Both_cluster'\n",
    "B50_HMMFinal_df.loc[B50_HMMFinal_df['num_duplicated'] == 2 , 'Homologous_cluster'] = 'Both_cluster'\n",
    "\n",
    "# remove duplicated row\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.drop_duplicates(subset=['genome_id'])\n",
    "print('B50_final Row (1):', len(B50_HMMFinal_df.index))\n",
    "\n",
    "# remove, add and rearrange column\n",
    "B50_HMMFinal_df.drop('num_duplicated', axis=1, inplace=True)\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.iloc[:, 1:]\n",
    "\n",
    "# open file for vlookup from phylum to final taxonomy ID\n",
    "PhylumToTaxonomy = pd.read_csv('../data/interim/PhylumToTaxonomy.csv')\n",
    "ClassToTaxnomy = pd.read_csv('../data/interim/ClassToTaxonomy.csv')\n",
    "\n",
    "# merge with phylum column\n",
    "B50_HMMFinal_df1 = pd.merge(B50_HMMFinal_df, PhylumToTaxonomy, on='Phylum', how='left')\n",
    "\n",
    "\n",
    "# merge with Class\n",
    "B50_HMMFinal_df2 = pd.merge(B50_HMMFinal_df, ClassToTaxnomy, on='Class', how='left')\n",
    "\n",
    "# concatenate the two dataframes\n",
    "B50_HMMFinal_df = pd.concat([B50_HMMFinal_df1, B50_HMMFinal_df2], axis=0, ignore_index=True)\n",
    "\n",
    "# drop rows with NaN values in taxonomy column\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.dropna(subset=['taxonomy'])\n",
    "print('B50_final Row (2):', len(B50_HMMFinal_df.index))\n",
    "B50_HMMFinal_df\n",
    "B50_HMMFinal_df.to_csv('../data/processed/Final/Combined/Combined_PositiveHits_10Hits.csv')\n",
    "\n",
    "# remove and rerrange column\n",
    "B50_HMMFinal_df.drop(columns=['Phylum', 'Class'], inplace=True)\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.reindex(columns=['genome_id', 'num_hits', 'taxonomy', 'ecosystem', 'ecosystem_type', 'Homologous_cluster'])\n",
    "\n",
    "# calculate hits number\n",
    "B50_HMMFinal_number = B50_HMMFinal_df.groupby(['ecosystem', 'ecosystem_type', 'taxonomy']).size().reset_index(name='NumberOfMAGs')\n",
    "B50_HMMFinal_number.to_csv('../data/processed/Final/R/B50_HMMFinal_10Hits_ForBarChart.csv', index=False)\n",
    "\n",
    "\n",
    "# done\n",
    "print(B50_HMMFinal_number['taxonomy'].unique())\n",
    "print(B50_HMMFinal_number['ecosystem'].unique())\n",
    "print(B50_HMMFinal_number['ecosystem_type'].unique())\n",
    "print('done')\n",
    "B50_HMMFinal_number"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3476f5e8-4814-4919-902b-0841baff6e21",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 4. 整理 aed 與 edc 的 reference positive MAGs 資料"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f5132847-abec-4647-bdac-5d2a5b618673",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B50_final Row (1): 315\n",
      "B50_final Row (2): 315\n",
      "['p_Actinobacteriota_c_Acidimicrobiia'\n",
      " 'p_Actinobacteriota_c_Actinobacteria' 'p_Myxococcota'\n",
      " 'p_Proteobacteria_c_Alphaproteobacteria'\n",
      " 'p_Proteobacteria_c_Gammaproteobacteria'\n",
      " 'p_Chloroflexota_c_Dehalococcoidia' 'p_Other bacterial phylum']\n",
      "['Aquatic' 'Engineered' 'Host-associated' 'Terrestrial']\n",
      "['Freshwater' 'Marine' 'Non-marine Saline and Alkaline' 'Sediment'\n",
      " 'Thermal spring' 'Bioreactor' 'Bioremediation' 'Built environment'\n",
      " 'Lab enrichment' 'Solid waste' 'Wastewater' 'Algae' 'Annelida'\n",
      " 'Arthropoda' 'Cnidaria' 'Fungi' 'Insecta' 'Mammal' 'Plant' 'Porifera'\n",
      " 'Deep subsurface' 'Soil']\n",
      "done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ecosystem</th>\n",
       "      <th>ecosystem_type</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>NumberOfMAGs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>p_Actinobacteriota_c_Acidimicrobiia</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>p_Actinobacteriota_c_Actinobacteria</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>p_Myxococcota</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>p_Actinobacteriota_c_Acidimicrobiia</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>p_Actinobacteriota_c_Actinobacteria</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>p_Chloroflexota_c_Dehalococcoidia</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>p_Myxococcota</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Non-marine Saline and Alkaline</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Non-marine Saline and Alkaline</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Thermal spring</td>\n",
       "      <td>p_Actinobacteriota_c_Actinobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Thermal spring</td>\n",
       "      <td>p_Chloroflexota_c_Dehalococcoidia</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Thermal spring</td>\n",
       "      <td>p_Myxococcota</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Engineered</td>\n",
       "      <td>Bioreactor</td>\n",
       "      <td>p_Actinobacteriota_c_Actinobacteria</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Engineered</td>\n",
       "      <td>Bioreactor</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Engineered</td>\n",
       "      <td>Bioremediation</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Engineered</td>\n",
       "      <td>Built environment</td>\n",
       "      <td>p_Actinobacteriota_c_Actinobacteria</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Engineered</td>\n",
       "      <td>Built environment</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Engineered</td>\n",
       "      <td>Built environment</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Engineered</td>\n",
       "      <td>Lab enrichment</td>\n",
       "      <td>p_Actinobacteriota_c_Actinobacteria</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Engineered</td>\n",
       "      <td>Lab enrichment</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Engineered</td>\n",
       "      <td>Lab enrichment</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Engineered</td>\n",
       "      <td>Solid waste</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Engineered</td>\n",
       "      <td>Wastewater</td>\n",
       "      <td>p_Actinobacteriota_c_Acidimicrobiia</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Engineered</td>\n",
       "      <td>Wastewater</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Engineered</td>\n",
       "      <td>Wastewater</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Algae</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Annelida</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Annelida</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>p_Actinobacteriota_c_Actinobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Cnidaria</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>p_Actinobacteriota_c_Actinobacteria</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Mammal</td>\n",
       "      <td>p_Actinobacteriota_c_Actinobacteria</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Plant</td>\n",
       "      <td>p_Actinobacteriota_c_Actinobacteria</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Plant</td>\n",
       "      <td>p_Other bacterial phylum</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Plant</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Plant</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Porifera</td>\n",
       "      <td>p_Actinobacteriota_c_Acidimicrobiia</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Porifera</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Deep subsurface</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Deep subsurface</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>p_Actinobacteriota_c_Acidimicrobiia</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>p_Actinobacteriota_c_Actinobacteria</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>p_Other bacterial phylum</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>p_Proteobacteria_c_Alphaproteobacteria</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>p_Proteobacteria_c_Gammaproteobacteria</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          ecosystem                  ecosystem_type  \\\n",
       "0           Aquatic                      Freshwater   \n",
       "1           Aquatic                      Freshwater   \n",
       "2           Aquatic                      Freshwater   \n",
       "3           Aquatic                      Freshwater   \n",
       "4           Aquatic                      Freshwater   \n",
       "5           Aquatic                          Marine   \n",
       "6           Aquatic                          Marine   \n",
       "7           Aquatic                          Marine   \n",
       "8           Aquatic                          Marine   \n",
       "9           Aquatic                          Marine   \n",
       "10          Aquatic                          Marine   \n",
       "11          Aquatic  Non-marine Saline and Alkaline   \n",
       "12          Aquatic  Non-marine Saline and Alkaline   \n",
       "13          Aquatic                        Sediment   \n",
       "14          Aquatic                  Thermal spring   \n",
       "15          Aquatic                  Thermal spring   \n",
       "16          Aquatic                  Thermal spring   \n",
       "17       Engineered                      Bioreactor   \n",
       "18       Engineered                      Bioreactor   \n",
       "19       Engineered                  Bioremediation   \n",
       "20       Engineered               Built environment   \n",
       "21       Engineered               Built environment   \n",
       "22       Engineered               Built environment   \n",
       "23       Engineered                  Lab enrichment   \n",
       "24       Engineered                  Lab enrichment   \n",
       "25       Engineered                  Lab enrichment   \n",
       "26       Engineered                     Solid waste   \n",
       "27       Engineered                      Wastewater   \n",
       "28       Engineered                      Wastewater   \n",
       "29       Engineered                      Wastewater   \n",
       "30  Host-associated                           Algae   \n",
       "31  Host-associated                        Annelida   \n",
       "32  Host-associated                        Annelida   \n",
       "33  Host-associated                      Arthropoda   \n",
       "34  Host-associated                      Arthropoda   \n",
       "35  Host-associated                        Cnidaria   \n",
       "36  Host-associated                           Fungi   \n",
       "37  Host-associated                           Fungi   \n",
       "38  Host-associated                           Fungi   \n",
       "39  Host-associated                         Insecta   \n",
       "40  Host-associated                          Mammal   \n",
       "41  Host-associated                           Plant   \n",
       "42  Host-associated                           Plant   \n",
       "43  Host-associated                           Plant   \n",
       "44  Host-associated                           Plant   \n",
       "45  Host-associated                        Porifera   \n",
       "46  Host-associated                        Porifera   \n",
       "47      Terrestrial                 Deep subsurface   \n",
       "48      Terrestrial                 Deep subsurface   \n",
       "49      Terrestrial                            Soil   \n",
       "50      Terrestrial                            Soil   \n",
       "51      Terrestrial                            Soil   \n",
       "52      Terrestrial                            Soil   \n",
       "53      Terrestrial                            Soil   \n",
       "\n",
       "                                  taxonomy  NumberOfMAGs  \n",
       "0      p_Actinobacteriota_c_Acidimicrobiia             5  \n",
       "1      p_Actinobacteriota_c_Actinobacteria             4  \n",
       "2                            p_Myxococcota             1  \n",
       "3   p_Proteobacteria_c_Alphaproteobacteria            17  \n",
       "4   p_Proteobacteria_c_Gammaproteobacteria             8  \n",
       "5      p_Actinobacteriota_c_Acidimicrobiia             8  \n",
       "6      p_Actinobacteriota_c_Actinobacteria            15  \n",
       "7        p_Chloroflexota_c_Dehalococcoidia             1  \n",
       "8                            p_Myxococcota             3  \n",
       "9   p_Proteobacteria_c_Alphaproteobacteria            28  \n",
       "10  p_Proteobacteria_c_Gammaproteobacteria            50  \n",
       "11  p_Proteobacteria_c_Alphaproteobacteria             2  \n",
       "12  p_Proteobacteria_c_Gammaproteobacteria             8  \n",
       "13  p_Proteobacteria_c_Alphaproteobacteria             1  \n",
       "14     p_Actinobacteriota_c_Actinobacteria             1  \n",
       "15       p_Chloroflexota_c_Dehalococcoidia             2  \n",
       "16                           p_Myxococcota             1  \n",
       "17     p_Actinobacteriota_c_Actinobacteria            10  \n",
       "18  p_Proteobacteria_c_Alphaproteobacteria             1  \n",
       "19  p_Proteobacteria_c_Gammaproteobacteria             1  \n",
       "20     p_Actinobacteriota_c_Actinobacteria             5  \n",
       "21  p_Proteobacteria_c_Alphaproteobacteria             3  \n",
       "22  p_Proteobacteria_c_Gammaproteobacteria            37  \n",
       "23     p_Actinobacteriota_c_Actinobacteria             4  \n",
       "24  p_Proteobacteria_c_Alphaproteobacteria             1  \n",
       "25  p_Proteobacteria_c_Gammaproteobacteria             6  \n",
       "26  p_Proteobacteria_c_Gammaproteobacteria             1  \n",
       "27     p_Actinobacteriota_c_Acidimicrobiia             1  \n",
       "28  p_Proteobacteria_c_Alphaproteobacteria             4  \n",
       "29  p_Proteobacteria_c_Gammaproteobacteria             2  \n",
       "30  p_Proteobacteria_c_Alphaproteobacteria             2  \n",
       "31  p_Proteobacteria_c_Alphaproteobacteria             1  \n",
       "32  p_Proteobacteria_c_Gammaproteobacteria             1  \n",
       "33     p_Actinobacteriota_c_Actinobacteria             1  \n",
       "34  p_Proteobacteria_c_Gammaproteobacteria             1  \n",
       "35  p_Proteobacteria_c_Gammaproteobacteria             1  \n",
       "36     p_Actinobacteriota_c_Actinobacteria             3  \n",
       "37  p_Proteobacteria_c_Alphaproteobacteria             3  \n",
       "38  p_Proteobacteria_c_Gammaproteobacteria             1  \n",
       "39  p_Proteobacteria_c_Alphaproteobacteria             2  \n",
       "40     p_Actinobacteriota_c_Actinobacteria             1  \n",
       "41     p_Actinobacteriota_c_Actinobacteria             3  \n",
       "42                p_Other bacterial phylum             1  \n",
       "43  p_Proteobacteria_c_Alphaproteobacteria            12  \n",
       "44  p_Proteobacteria_c_Gammaproteobacteria             3  \n",
       "45     p_Actinobacteriota_c_Acidimicrobiia             6  \n",
       "46  p_Proteobacteria_c_Alphaproteobacteria             2  \n",
       "47  p_Proteobacteria_c_Alphaproteobacteria             2  \n",
       "48  p_Proteobacteria_c_Gammaproteobacteria             4  \n",
       "49     p_Actinobacteriota_c_Acidimicrobiia             3  \n",
       "50     p_Actinobacteriota_c_Actinobacteria            11  \n",
       "51                p_Other bacterial phylum             1  \n",
       "52  p_Proteobacteria_c_Alphaproteobacteria            12  \n",
       "53  p_Proteobacteria_c_Gammaproteobacteria             7  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read table\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "aed_final_df = pd.read_csv('../data/processed/Final/Actino/aed_PositiveHits_ForR_reference.csv')\n",
    "edc_final_df = pd.read_csv('../data/processed/Final/Proteo/edc_PositiveHits_ForR_reference.csv')\n",
    "\n",
    "# concatenate aed and edc final data\n",
    "B50_HMMFinal_df = pd.concat([aed_final_df, edc_final_df])\n",
    "\n",
    "# remove duplicated and update the 'Both_clsuter' in homologous_cluster column\n",
    "# 找重複 (有2)\n",
    "B50_duplicated = B50_HMMFinal_df['genome_id'].value_counts()\n",
    "\n",
    "# set new column in DataFrame with value counts\n",
    "B50_HMMFinal_df['num_duplicated'] = B50_HMMFinal_df['genome_id'].map(B50_duplicated)\n",
    "\n",
    "# if number is 2 (duplicated) change the homnologous column to 'Both_cluster'\n",
    "B50_HMMFinal_df.loc[B50_HMMFinal_df['num_duplicated'] == 2 , 'Homologous_cluster'] = 'Both_cluster'\n",
    "\n",
    "# remove duplicated row\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.drop_duplicates(subset=['genome_id'])\n",
    "print('B50_final Row (1):', len(B50_HMMFinal_df.index))\n",
    "\n",
    "# remove, add and rearrange column\n",
    "B50_HMMFinal_df.drop('num_duplicated', axis=1, inplace=True)\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.iloc[:, 1:]\n",
    "\n",
    "# open file for vlookup from phylum to final taxonomy ID\n",
    "PhylumToTaxonomy = pd.read_csv('../data/interim/PhylumToTaxonomy.csv')\n",
    "ClassToTaxnomy = pd.read_csv('../data/interim/ClassToTaxonomy.csv')\n",
    "\n",
    "# merge with phylum column\n",
    "B50_HMMFinal_df1 = pd.merge(B50_HMMFinal_df, PhylumToTaxonomy, on='Phylum', how='left')\n",
    "\n",
    "\n",
    "# merge with Class\n",
    "B50_HMMFinal_df2 = pd.merge(B50_HMMFinal_df, ClassToTaxnomy, on='Class', how='left')\n",
    "\n",
    "# concatenate the two dataframes\n",
    "B50_HMMFinal_df = pd.concat([B50_HMMFinal_df1, B50_HMMFinal_df2], axis=0, ignore_index=True)\n",
    "\n",
    "# drop rows with NaN values in taxonomy column\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.dropna(subset=['taxonomy'])\n",
    "print('B50_final Row (2):', len(B50_HMMFinal_df.index))\n",
    "B50_HMMFinal_df\n",
    "B50_HMMFinal_df.to_csv('../data/processed/Final/Combined/Combined_PositiveHits_reference.csv')\n",
    "\n",
    "# remove and rerrange column\n",
    "B50_HMMFinal_df.drop(columns=['Phylum', 'Class'], inplace=True)\n",
    "B50_HMMFinal_df = B50_HMMFinal_df.reindex(columns=['genome_id', 'num_hits', 'taxonomy', 'ecosystem', 'ecosystem_type', 'Homologous_cluster'])\n",
    "\n",
    "# calculate hits number\n",
    "B50_HMMFinal_number = B50_HMMFinal_df.groupby(['ecosystem', 'ecosystem_type', 'taxonomy']).size().reset_index(name='NumberOfMAGs')\n",
    "B50_HMMFinal_number.to_csv('../data/processed/Final/R/B50_HMMFinal_reference_ForBarChart.csv', index=False)\n",
    "\n",
    "\n",
    "# done\n",
    "print(B50_HMMFinal_number['taxonomy'].unique())\n",
    "print(B50_HMMFinal_number['ecosystem'].unique())\n",
    "print(B50_HMMFinal_number['ecosystem_type'].unique())\n",
    "print('done')\n",
    "B50_HMMFinal_number"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59878e7e",
   "metadata": {},
   "source": [
    "# 補充"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "795c0889-45e3-4d6b-bdb2-d1f0d0823cd6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>foo</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bar</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>baz</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     A  B\n",
       "0  foo  1\n",
       "1  bar  2\n",
       "2  baz  3"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # create a sample dataframe\n",
    "# import pandas as pd\n",
    "# data = {'A': ['foo', 'bar', 'baz', 'foo', 'bar', 'baz'], 'B': [1, 2, 3, 4, 5, 6]}\n",
    "# df = pd.DataFrame(data)\n",
    "\n",
    "# # remove duplicated rows and update values in column 'B'\n",
    "# mask = df.duplicated(subset=['A'])\n",
    "# df.loc[mask, 'B'] = 999\n",
    "# df = df[~mask]\n",
    "# df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62654f62-0bdc-4404-97a4-f288c3c5036c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e82be1e9-b3d7-451f-8f90-49c48987774c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
