{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b84f7a73-7758-47c0-9a34-ec20dde4b02c",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Note2_關於actino 的MAGs分析"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b094ad9-7910-4c2e-af85-791ddf3be147",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 第一部分: 含有bitscore的篩選分析"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8eefda97-0103-44e1-a3ec-147d5c3009ad",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 1.1 使用 positive control 來了解 hmm profiles 的分析情形 (e-value、coverage、bit-score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2bd3b51c-3d15-4c8f-8e2f-f8e805417524",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A_aedA_I_RS26385</th>\n",
       "      <th>A_aedB_I_RS26395</th>\n",
       "      <th>A_aedD_RS26370</th>\n",
       "      <th>A_aedE_RS26375</th>\n",
       "      <th>A_aedF_I_RS26380</th>\n",
       "      <th>A_aedG_I_RS26390</th>\n",
       "      <th>A_aedH_I_RS26400</th>\n",
       "      <th>A_aedI_RS26405</th>\n",
       "      <th>A_aedJ_I_RS26410</th>\n",
       "      <th>A_aedK_I_RS26415</th>\n",
       "      <th>A_aedL_RS26420</th>\n",
       "      <th>A_aedM_RS26425</th>\n",
       "      <th>A_aedN_RS26430</th>\n",
       "      <th>A_aedO_RS26435</th>\n",
       "      <th>A_aedP_RS26440</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A_Dietzia_sp_B32_aed_bit_score</th>\n",
       "      <td>793.5</td>\n",
       "      <td>588.3</td>\n",
       "      <td>450.9</td>\n",
       "      <td>596.0</td>\n",
       "      <td>771.9</td>\n",
       "      <td>555.5</td>\n",
       "      <td>552.7</td>\n",
       "      <td>220.7</td>\n",
       "      <td>965.6</td>\n",
       "      <td>814.7</td>\n",
       "      <td>261.4</td>\n",
       "      <td>533.0</td>\n",
       "      <td>845.5</td>\n",
       "      <td>420.8</td>\n",
       "      <td>755.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_Mycobacteroides_chelonae_S00154_aed_bit_score</th>\n",
       "      <td>726.7</td>\n",
       "      <td>569.5</td>\n",
       "      <td>432.5</td>\n",
       "      <td>586.1</td>\n",
       "      <td>736.0</td>\n",
       "      <td>552.4</td>\n",
       "      <td>518.3</td>\n",
       "      <td>217.1</td>\n",
       "      <td>942.7</td>\n",
       "      <td>801.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>532.7</td>\n",
       "      <td>656.6</td>\n",
       "      <td>126.6</td>\n",
       "      <td>439.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_Rhodococcus_equi_DSSKP_R_001_aed_bit_score</th>\n",
       "      <td>809.4</td>\n",
       "      <td>609.6</td>\n",
       "      <td>464.9</td>\n",
       "      <td>613.3</td>\n",
       "      <td>783.8</td>\n",
       "      <td>575.4</td>\n",
       "      <td>560.9</td>\n",
       "      <td>230.4</td>\n",
       "      <td>1007.8</td>\n",
       "      <td>838.8</td>\n",
       "      <td>273.9</td>\n",
       "      <td>547.4</td>\n",
       "      <td>865.6</td>\n",
       "      <td>443.7</td>\n",
       "      <td>784.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_Rhodococcus_sp_B50_aed_bit_score</th>\n",
       "      <td>804.4</td>\n",
       "      <td>612.8</td>\n",
       "      <td>466.2</td>\n",
       "      <td>614.9</td>\n",
       "      <td>785.2</td>\n",
       "      <td>577.2</td>\n",
       "      <td>562.7</td>\n",
       "      <td>233.0</td>\n",
       "      <td>1018.5</td>\n",
       "      <td>840.5</td>\n",
       "      <td>272.7</td>\n",
       "      <td>549.3</td>\n",
       "      <td>868.3</td>\n",
       "      <td>445.9</td>\n",
       "      <td>777.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_Tomitella_gaofuii_HY172_aed_bit_score</th>\n",
       "      <td>783.3</td>\n",
       "      <td>594.9</td>\n",
       "      <td>283.2</td>\n",
       "      <td>580.8</td>\n",
       "      <td>763.5</td>\n",
       "      <td>537.9</td>\n",
       "      <td>545.0</td>\n",
       "      <td>218.9</td>\n",
       "      <td>933.9</td>\n",
       "      <td>795.1</td>\n",
       "      <td>257.3</td>\n",
       "      <td>432.4</td>\n",
       "      <td>827.1</td>\n",
       "      <td>412.0</td>\n",
       "      <td>729.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PositiveControl_Min_BitScore</th>\n",
       "      <td>726.7</td>\n",
       "      <td>569.5</td>\n",
       "      <td>283.2</td>\n",
       "      <td>580.8</td>\n",
       "      <td>736.0</td>\n",
       "      <td>537.9</td>\n",
       "      <td>518.3</td>\n",
       "      <td>217.1</td>\n",
       "      <td>933.9</td>\n",
       "      <td>795.1</td>\n",
       "      <td>257.3</td>\n",
       "      <td>432.4</td>\n",
       "      <td>656.6</td>\n",
       "      <td>126.6</td>\n",
       "      <td>439.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 A_aedA_I_RS26385  \\\n",
       "A_Dietzia_sp_B32_aed_bit_score                              793.5   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_bit_score             726.7   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_bit_score                809.4   \n",
       "A_Rhodococcus_sp_B50_aed_bit_score                          804.4   \n",
       "A_Tomitella_gaofuii_HY172_aed_bit_score                     783.3   \n",
       "PositiveControl_Min_BitScore                                726.7   \n",
       "\n",
       "                                                 A_aedB_I_RS26395  \\\n",
       "A_Dietzia_sp_B32_aed_bit_score                              588.3   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_bit_score             569.5   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_bit_score                609.6   \n",
       "A_Rhodococcus_sp_B50_aed_bit_score                          612.8   \n",
       "A_Tomitella_gaofuii_HY172_aed_bit_score                     594.9   \n",
       "PositiveControl_Min_BitScore                                569.5   \n",
       "\n",
       "                                                 A_aedD_RS26370  \\\n",
       "A_Dietzia_sp_B32_aed_bit_score                            450.9   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_bit_score           432.5   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_bit_score              464.9   \n",
       "A_Rhodococcus_sp_B50_aed_bit_score                        466.2   \n",
       "A_Tomitella_gaofuii_HY172_aed_bit_score                   283.2   \n",
       "PositiveControl_Min_BitScore                              283.2   \n",
       "\n",
       "                                                 A_aedE_RS26375  \\\n",
       "A_Dietzia_sp_B32_aed_bit_score                            596.0   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_bit_score           586.1   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_bit_score              613.3   \n",
       "A_Rhodococcus_sp_B50_aed_bit_score                        614.9   \n",
       "A_Tomitella_gaofuii_HY172_aed_bit_score                   580.8   \n",
       "PositiveControl_Min_BitScore                              580.8   \n",
       "\n",
       "                                                 A_aedF_I_RS26380  \\\n",
       "A_Dietzia_sp_B32_aed_bit_score                              771.9   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_bit_score             736.0   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_bit_score                783.8   \n",
       "A_Rhodococcus_sp_B50_aed_bit_score                          785.2   \n",
       "A_Tomitella_gaofuii_HY172_aed_bit_score                     763.5   \n",
       "PositiveControl_Min_BitScore                                736.0   \n",
       "\n",
       "                                                 A_aedG_I_RS26390  \\\n",
       "A_Dietzia_sp_B32_aed_bit_score                              555.5   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_bit_score             552.4   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_bit_score                575.4   \n",
       "A_Rhodococcus_sp_B50_aed_bit_score                          577.2   \n",
       "A_Tomitella_gaofuii_HY172_aed_bit_score                     537.9   \n",
       "PositiveControl_Min_BitScore                                537.9   \n",
       "\n",
       "                                                 A_aedH_I_RS26400  \\\n",
       "A_Dietzia_sp_B32_aed_bit_score                              552.7   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_bit_score             518.3   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_bit_score                560.9   \n",
       "A_Rhodococcus_sp_B50_aed_bit_score                          562.7   \n",
       "A_Tomitella_gaofuii_HY172_aed_bit_score                     545.0   \n",
       "PositiveControl_Min_BitScore                                518.3   \n",
       "\n",
       "                                                 A_aedI_RS26405  \\\n",
       "A_Dietzia_sp_B32_aed_bit_score                            220.7   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_bit_score           217.1   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_bit_score              230.4   \n",
       "A_Rhodococcus_sp_B50_aed_bit_score                        233.0   \n",
       "A_Tomitella_gaofuii_HY172_aed_bit_score                   218.9   \n",
       "PositiveControl_Min_BitScore                              217.1   \n",
       "\n",
       "                                                 A_aedJ_I_RS26410  \\\n",
       "A_Dietzia_sp_B32_aed_bit_score                              965.6   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_bit_score             942.7   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_bit_score               1007.8   \n",
       "A_Rhodococcus_sp_B50_aed_bit_score                         1018.5   \n",
       "A_Tomitella_gaofuii_HY172_aed_bit_score                     933.9   \n",
       "PositiveControl_Min_BitScore                                933.9   \n",
       "\n",
       "                                                 A_aedK_I_RS26415  \\\n",
       "A_Dietzia_sp_B32_aed_bit_score                              814.7   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_bit_score             801.3   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_bit_score                838.8   \n",
       "A_Rhodococcus_sp_B50_aed_bit_score                          840.5   \n",
       "A_Tomitella_gaofuii_HY172_aed_bit_score                     795.1   \n",
       "PositiveControl_Min_BitScore                                795.1   \n",
       "\n",
       "                                                 A_aedL_RS26420  \\\n",
       "A_Dietzia_sp_B32_aed_bit_score                            261.4   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_bit_score             NaN   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_bit_score              273.9   \n",
       "A_Rhodococcus_sp_B50_aed_bit_score                        272.7   \n",
       "A_Tomitella_gaofuii_HY172_aed_bit_score                   257.3   \n",
       "PositiveControl_Min_BitScore                              257.3   \n",
       "\n",
       "                                                 A_aedM_RS26425  \\\n",
       "A_Dietzia_sp_B32_aed_bit_score                            533.0   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_bit_score           532.7   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_bit_score              547.4   \n",
       "A_Rhodococcus_sp_B50_aed_bit_score                        549.3   \n",
       "A_Tomitella_gaofuii_HY172_aed_bit_score                   432.4   \n",
       "PositiveControl_Min_BitScore                              432.4   \n",
       "\n",
       "                                                 A_aedN_RS26430  \\\n",
       "A_Dietzia_sp_B32_aed_bit_score                            845.5   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_bit_score           656.6   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_bit_score              865.6   \n",
       "A_Rhodococcus_sp_B50_aed_bit_score                        868.3   \n",
       "A_Tomitella_gaofuii_HY172_aed_bit_score                   827.1   \n",
       "PositiveControl_Min_BitScore                              656.6   \n",
       "\n",
       "                                                 A_aedO_RS26435  \\\n",
       "A_Dietzia_sp_B32_aed_bit_score                            420.8   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_bit_score           126.6   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_bit_score              443.7   \n",
       "A_Rhodococcus_sp_B50_aed_bit_score                        445.9   \n",
       "A_Tomitella_gaofuii_HY172_aed_bit_score                   412.0   \n",
       "PositiveControl_Min_BitScore                              126.6   \n",
       "\n",
       "                                                 A_aedP_RS26440  \n",
       "A_Dietzia_sp_B32_aed_bit_score                            755.6  \n",
       "A_Mycobacteroides_chelonae_S00154_aed_bit_score           439.4  \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_bit_score              784.6  \n",
       "A_Rhodococcus_sp_B50_aed_bit_score                        777.8  \n",
       "A_Tomitella_gaofuii_HY172_aed_bit_score                   729.9  \n",
       "PositiveControl_Min_BitScore                              439.4  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A_aedA_I_RS26385</th>\n",
       "      <th>A_aedB_I_RS26395</th>\n",
       "      <th>A_aedD_RS26370</th>\n",
       "      <th>A_aedE_RS26375</th>\n",
       "      <th>A_aedF_I_RS26380</th>\n",
       "      <th>A_aedG_I_RS26390</th>\n",
       "      <th>A_aedH_I_RS26400</th>\n",
       "      <th>A_aedI_RS26405</th>\n",
       "      <th>A_aedJ_I_RS26410</th>\n",
       "      <th>A_aedK_I_RS26415</th>\n",
       "      <th>A_aedL_RS26420</th>\n",
       "      <th>A_aedM_RS26425</th>\n",
       "      <th>A_aedN_RS26430</th>\n",
       "      <th>A_aedO_RS26435</th>\n",
       "      <th>A_aedP_RS26440</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A_Dietzia_sp_B32_aed_Evalue</th>\n",
       "      <td>3.400000e-242</td>\n",
       "      <td>2.300000e-180</td>\n",
       "      <td>8.100000e-139</td>\n",
       "      <td>1.100000e-182</td>\n",
       "      <td>1.100000e-235</td>\n",
       "      <td>1.900000e-170</td>\n",
       "      <td>8.600000e-170</td>\n",
       "      <td>4.100000e-69</td>\n",
       "      <td>3.200000e-294</td>\n",
       "      <td>8.600000e-249</td>\n",
       "      <td>1.100000e-81</td>\n",
       "      <td>8.400000e-164</td>\n",
       "      <td>3.300000e-258</td>\n",
       "      <td>1.100000e-129</td>\n",
       "      <td>7.900000e-231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_Mycobacteroides_chelonae_S00154_aed_Evalue</th>\n",
       "      <td>5.800000e-222</td>\n",
       "      <td>1.000000e-174</td>\n",
       "      <td>3.000000e-133</td>\n",
       "      <td>1.100000e-179</td>\n",
       "      <td>7.400000e-225</td>\n",
       "      <td>1.500000e-169</td>\n",
       "      <td>2.400000e-159</td>\n",
       "      <td>4.500000e-68</td>\n",
       "      <td>2.400000e-287</td>\n",
       "      <td>8.800000e-245</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.000000e-164</td>\n",
       "      <td>6.600000e-201</td>\n",
       "      <td>7.600000e-40</td>\n",
       "      <td>7.900000e-135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_Rhodococcus_equi_DSSKP_R_001_aed_Evalue</th>\n",
       "      <td>5.700000e-247</td>\n",
       "      <td>8.400000e-187</td>\n",
       "      <td>5.100000e-143</td>\n",
       "      <td>7.200000e-188</td>\n",
       "      <td>2.800000e-239</td>\n",
       "      <td>1.900000e-176</td>\n",
       "      <td>3.200000e-172</td>\n",
       "      <td>4.400000e-72</td>\n",
       "      <td>5.700000e-307</td>\n",
       "      <td>4.600000e-256</td>\n",
       "      <td>1.800000e-85</td>\n",
       "      <td>3.700000e-168</td>\n",
       "      <td>2.900000e-264</td>\n",
       "      <td>1.200000e-136</td>\n",
       "      <td>1.300000e-239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_Rhodococcus_sp_B50_aed_Evalue</th>\n",
       "      <td>1.800000e-245</td>\n",
       "      <td>8.600000e-188</td>\n",
       "      <td>2.000000e-143</td>\n",
       "      <td>2.400000e-188</td>\n",
       "      <td>1.000000e-239</td>\n",
       "      <td>5.500000e-177</td>\n",
       "      <td>8.700000e-173</td>\n",
       "      <td>7.000000e-73</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.400000e-256</td>\n",
       "      <td>4.300000e-85</td>\n",
       "      <td>9.900000e-169</td>\n",
       "      <td>4.400000e-265</td>\n",
       "      <td>2.700000e-137</td>\n",
       "      <td>1.600000e-237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_Tomitella_gaofuii_HY172_aed_Evalue</th>\n",
       "      <td>4.400000e-239</td>\n",
       "      <td>2.200000e-182</td>\n",
       "      <td>1.400000e-87</td>\n",
       "      <td>5.000000e-178</td>\n",
       "      <td>3.700000e-233</td>\n",
       "      <td>4.400000e-165</td>\n",
       "      <td>2.000000e-167</td>\n",
       "      <td>1.400000e-68</td>\n",
       "      <td>1.300000e-284</td>\n",
       "      <td>7.500000e-243</td>\n",
       "      <td>2.100000e-80</td>\n",
       "      <td>4.000000e-133</td>\n",
       "      <td>1.200000e-252</td>\n",
       "      <td>5.400000e-127</td>\n",
       "      <td>4.900000e-223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PositiveControl_Max_Evalue</th>\n",
       "      <td>5.800000e-222</td>\n",
       "      <td>1.000000e-174</td>\n",
       "      <td>1.400000e-87</td>\n",
       "      <td>5.000000e-178</td>\n",
       "      <td>7.400000e-225</td>\n",
       "      <td>4.400000e-165</td>\n",
       "      <td>2.400000e-159</td>\n",
       "      <td>4.500000e-68</td>\n",
       "      <td>1.300000e-284</td>\n",
       "      <td>7.500000e-243</td>\n",
       "      <td>2.100000e-80</td>\n",
       "      <td>4.000000e-133</td>\n",
       "      <td>6.600000e-201</td>\n",
       "      <td>7.600000e-40</td>\n",
       "      <td>7.900000e-135</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              A_aedA_I_RS26385  \\\n",
       "A_Dietzia_sp_B32_aed_Evalue                      3.400000e-242   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_Evalue     5.800000e-222   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_Evalue        5.700000e-247   \n",
       "A_Rhodococcus_sp_B50_aed_Evalue                  1.800000e-245   \n",
       "A_Tomitella_gaofuii_HY172_aed_Evalue             4.400000e-239   \n",
       "PositiveControl_Max_Evalue                       5.800000e-222   \n",
       "\n",
       "                                              A_aedB_I_RS26395  \\\n",
       "A_Dietzia_sp_B32_aed_Evalue                      2.300000e-180   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_Evalue     1.000000e-174   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_Evalue        8.400000e-187   \n",
       "A_Rhodococcus_sp_B50_aed_Evalue                  8.600000e-188   \n",
       "A_Tomitella_gaofuii_HY172_aed_Evalue             2.200000e-182   \n",
       "PositiveControl_Max_Evalue                       1.000000e-174   \n",
       "\n",
       "                                              A_aedD_RS26370  A_aedE_RS26375  \\\n",
       "A_Dietzia_sp_B32_aed_Evalue                    8.100000e-139   1.100000e-182   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_Evalue   3.000000e-133   1.100000e-179   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_Evalue      5.100000e-143   7.200000e-188   \n",
       "A_Rhodococcus_sp_B50_aed_Evalue                2.000000e-143   2.400000e-188   \n",
       "A_Tomitella_gaofuii_HY172_aed_Evalue            1.400000e-87   5.000000e-178   \n",
       "PositiveControl_Max_Evalue                      1.400000e-87   5.000000e-178   \n",
       "\n",
       "                                              A_aedF_I_RS26380  \\\n",
       "A_Dietzia_sp_B32_aed_Evalue                      1.100000e-235   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_Evalue     7.400000e-225   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_Evalue        2.800000e-239   \n",
       "A_Rhodococcus_sp_B50_aed_Evalue                  1.000000e-239   \n",
       "A_Tomitella_gaofuii_HY172_aed_Evalue             3.700000e-233   \n",
       "PositiveControl_Max_Evalue                       7.400000e-225   \n",
       "\n",
       "                                              A_aedG_I_RS26390  \\\n",
       "A_Dietzia_sp_B32_aed_Evalue                      1.900000e-170   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_Evalue     1.500000e-169   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_Evalue        1.900000e-176   \n",
       "A_Rhodococcus_sp_B50_aed_Evalue                  5.500000e-177   \n",
       "A_Tomitella_gaofuii_HY172_aed_Evalue             4.400000e-165   \n",
       "PositiveControl_Max_Evalue                       4.400000e-165   \n",
       "\n",
       "                                              A_aedH_I_RS26400  \\\n",
       "A_Dietzia_sp_B32_aed_Evalue                      8.600000e-170   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_Evalue     2.400000e-159   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_Evalue        3.200000e-172   \n",
       "A_Rhodococcus_sp_B50_aed_Evalue                  8.700000e-173   \n",
       "A_Tomitella_gaofuii_HY172_aed_Evalue             2.000000e-167   \n",
       "PositiveControl_Max_Evalue                       2.400000e-159   \n",
       "\n",
       "                                              A_aedI_RS26405  \\\n",
       "A_Dietzia_sp_B32_aed_Evalue                     4.100000e-69   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_Evalue    4.500000e-68   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_Evalue       4.400000e-72   \n",
       "A_Rhodococcus_sp_B50_aed_Evalue                 7.000000e-73   \n",
       "A_Tomitella_gaofuii_HY172_aed_Evalue            1.400000e-68   \n",
       "PositiveControl_Max_Evalue                      4.500000e-68   \n",
       "\n",
       "                                              A_aedJ_I_RS26410  \\\n",
       "A_Dietzia_sp_B32_aed_Evalue                      3.200000e-294   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_Evalue     2.400000e-287   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_Evalue        5.700000e-307   \n",
       "A_Rhodococcus_sp_B50_aed_Evalue                   0.000000e+00   \n",
       "A_Tomitella_gaofuii_HY172_aed_Evalue             1.300000e-284   \n",
       "PositiveControl_Max_Evalue                       1.300000e-284   \n",
       "\n",
       "                                              A_aedK_I_RS26415  \\\n",
       "A_Dietzia_sp_B32_aed_Evalue                      8.600000e-249   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_Evalue     8.800000e-245   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_Evalue        4.600000e-256   \n",
       "A_Rhodococcus_sp_B50_aed_Evalue                  1.400000e-256   \n",
       "A_Tomitella_gaofuii_HY172_aed_Evalue             7.500000e-243   \n",
       "PositiveControl_Max_Evalue                       7.500000e-243   \n",
       "\n",
       "                                              A_aedL_RS26420  A_aedM_RS26425  \\\n",
       "A_Dietzia_sp_B32_aed_Evalue                     1.100000e-81   8.400000e-164   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_Evalue             NaN   9.000000e-164   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_Evalue       1.800000e-85   3.700000e-168   \n",
       "A_Rhodococcus_sp_B50_aed_Evalue                 4.300000e-85   9.900000e-169   \n",
       "A_Tomitella_gaofuii_HY172_aed_Evalue            2.100000e-80   4.000000e-133   \n",
       "PositiveControl_Max_Evalue                      2.100000e-80   4.000000e-133   \n",
       "\n",
       "                                              A_aedN_RS26430  A_aedO_RS26435  \\\n",
       "A_Dietzia_sp_B32_aed_Evalue                    3.300000e-258   1.100000e-129   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_Evalue   6.600000e-201    7.600000e-40   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_Evalue      2.900000e-264   1.200000e-136   \n",
       "A_Rhodococcus_sp_B50_aed_Evalue                4.400000e-265   2.700000e-137   \n",
       "A_Tomitella_gaofuii_HY172_aed_Evalue           1.200000e-252   5.400000e-127   \n",
       "PositiveControl_Max_Evalue                     6.600000e-201    7.600000e-40   \n",
       "\n",
       "                                              A_aedP_RS26440  \n",
       "A_Dietzia_sp_B32_aed_Evalue                    7.900000e-231  \n",
       "A_Mycobacteroides_chelonae_S00154_aed_Evalue   7.900000e-135  \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_Evalue      1.300000e-239  \n",
       "A_Rhodococcus_sp_B50_aed_Evalue                1.600000e-237  \n",
       "A_Tomitella_gaofuii_HY172_aed_Evalue           4.900000e-223  \n",
       "PositiveControl_Max_Evalue                     7.900000e-135  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A_aedA_I_RS26385</th>\n",
       "      <th>A_aedB_I_RS26395</th>\n",
       "      <th>A_aedD_RS26370</th>\n",
       "      <th>A_aedE_RS26375</th>\n",
       "      <th>A_aedF_I_RS26380</th>\n",
       "      <th>A_aedG_I_RS26390</th>\n",
       "      <th>A_aedH_I_RS26400</th>\n",
       "      <th>A_aedI_RS26405</th>\n",
       "      <th>A_aedJ_I_RS26410</th>\n",
       "      <th>A_aedK_I_RS26415</th>\n",
       "      <th>A_aedL_RS26420</th>\n",
       "      <th>A_aedM_RS26425</th>\n",
       "      <th>A_aedN_RS26430</th>\n",
       "      <th>A_aedO_RS26435</th>\n",
       "      <th>A_aedP_RS26440</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A_Dietzia_sp_B32_aed_coverage</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.990000</td>\n",
       "      <td>0.930502</td>\n",
       "      <td>0.954984</td>\n",
       "      <td>0.966321</td>\n",
       "      <td>0.910543</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.992308</td>\n",
       "      <td>0.942164</td>\n",
       "      <td>0.989770</td>\n",
       "      <td>0.904762</td>\n",
       "      <td>0.988679</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.663793</td>\n",
       "      <td>0.984615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_Mycobacteroides_chelonae_S00154_aed_coverage</th>\n",
       "      <td>0.994975</td>\n",
       "      <td>0.973770</td>\n",
       "      <td>0.995902</td>\n",
       "      <td>0.973856</td>\n",
       "      <td>0.997375</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.988462</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.992141</td>\n",
       "      <td>0.994859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.981273</td>\n",
       "      <td>0.992366</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.929095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_Rhodococcus_equi_DSSKP_R_001_aed_coverage</th>\n",
       "      <td>0.975430</td>\n",
       "      <td>0.993333</td>\n",
       "      <td>0.899628</td>\n",
       "      <td>0.993333</td>\n",
       "      <td>0.976440</td>\n",
       "      <td>0.916933</td>\n",
       "      <td>0.934783</td>\n",
       "      <td>0.992308</td>\n",
       "      <td>0.988235</td>\n",
       "      <td>0.989770</td>\n",
       "      <td>0.917241</td>\n",
       "      <td>0.988679</td>\n",
       "      <td>0.997468</td>\n",
       "      <td>0.689150</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_Rhodococcus_sp_B50_aed_coverage</th>\n",
       "      <td>0.977887</td>\n",
       "      <td>0.993333</td>\n",
       "      <td>0.899628</td>\n",
       "      <td>0.990033</td>\n",
       "      <td>0.976440</td>\n",
       "      <td>0.916933</td>\n",
       "      <td>0.934783</td>\n",
       "      <td>0.992308</td>\n",
       "      <td>0.992126</td>\n",
       "      <td>0.989770</td>\n",
       "      <td>0.796407</td>\n",
       "      <td>0.988679</td>\n",
       "      <td>0.997468</td>\n",
       "      <td>0.899614</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_Tomitella_gaofuii_HY172_aed_coverage</th>\n",
       "      <td>0.958937</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.914729</td>\n",
       "      <td>0.996656</td>\n",
       "      <td>0.976440</td>\n",
       "      <td>0.914013</td>\n",
       "      <td>0.724719</td>\n",
       "      <td>0.984733</td>\n",
       "      <td>0.970817</td>\n",
       "      <td>0.989770</td>\n",
       "      <td>0.815951</td>\n",
       "      <td>0.966292</td>\n",
       "      <td>0.997468</td>\n",
       "      <td>0.692308</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PositiveControl_Min_coverage</th>\n",
       "      <td>0.958937</td>\n",
       "      <td>0.973770</td>\n",
       "      <td>0.899628</td>\n",
       "      <td>0.954984</td>\n",
       "      <td>0.966321</td>\n",
       "      <td>0.910543</td>\n",
       "      <td>0.724719</td>\n",
       "      <td>0.984733</td>\n",
       "      <td>0.942164</td>\n",
       "      <td>0.989770</td>\n",
       "      <td>0.796407</td>\n",
       "      <td>0.966292</td>\n",
       "      <td>0.992366</td>\n",
       "      <td>0.663793</td>\n",
       "      <td>0.929095</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                A_aedA_I_RS26385  \\\n",
       "A_Dietzia_sp_B32_aed_coverage                           1.000000   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_coverage          0.994975   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_coverage             0.975430   \n",
       "A_Rhodococcus_sp_B50_aed_coverage                       0.977887   \n",
       "A_Tomitella_gaofuii_HY172_aed_coverage                  0.958937   \n",
       "PositiveControl_Min_coverage                            0.958937   \n",
       "\n",
       "                                                A_aedB_I_RS26395  \\\n",
       "A_Dietzia_sp_B32_aed_coverage                           0.990000   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_coverage          0.973770   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_coverage             0.993333   \n",
       "A_Rhodococcus_sp_B50_aed_coverage                       0.993333   \n",
       "A_Tomitella_gaofuii_HY172_aed_coverage                  1.000000   \n",
       "PositiveControl_Min_coverage                            0.973770   \n",
       "\n",
       "                                                A_aedD_RS26370  \\\n",
       "A_Dietzia_sp_B32_aed_coverage                         0.930502   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_coverage        0.995902   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_coverage           0.899628   \n",
       "A_Rhodococcus_sp_B50_aed_coverage                     0.899628   \n",
       "A_Tomitella_gaofuii_HY172_aed_coverage                0.914729   \n",
       "PositiveControl_Min_coverage                          0.899628   \n",
       "\n",
       "                                                A_aedE_RS26375  \\\n",
       "A_Dietzia_sp_B32_aed_coverage                         0.954984   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_coverage        0.973856   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_coverage           0.993333   \n",
       "A_Rhodococcus_sp_B50_aed_coverage                     0.990033   \n",
       "A_Tomitella_gaofuii_HY172_aed_coverage                0.996656   \n",
       "PositiveControl_Min_coverage                          0.954984   \n",
       "\n",
       "                                                A_aedF_I_RS26380  \\\n",
       "A_Dietzia_sp_B32_aed_coverage                           0.966321   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_coverage          0.997375   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_coverage             0.976440   \n",
       "A_Rhodococcus_sp_B50_aed_coverage                       0.976440   \n",
       "A_Tomitella_gaofuii_HY172_aed_coverage                  0.976440   \n",
       "PositiveControl_Min_coverage                            0.966321   \n",
       "\n",
       "                                                A_aedG_I_RS26390  \\\n",
       "A_Dietzia_sp_B32_aed_coverage                           0.910543   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_coverage          1.000000   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_coverage             0.916933   \n",
       "A_Rhodococcus_sp_B50_aed_coverage                       0.916933   \n",
       "A_Tomitella_gaofuii_HY172_aed_coverage                  0.914013   \n",
       "PositiveControl_Min_coverage                            0.910543   \n",
       "\n",
       "                                                A_aedH_I_RS26400  \\\n",
       "A_Dietzia_sp_B32_aed_coverage                           1.000000   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_coverage          0.988462   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_coverage             0.934783   \n",
       "A_Rhodococcus_sp_B50_aed_coverage                       0.934783   \n",
       "A_Tomitella_gaofuii_HY172_aed_coverage                  0.724719   \n",
       "PositiveControl_Min_coverage                            0.724719   \n",
       "\n",
       "                                                A_aedI_RS26405  \\\n",
       "A_Dietzia_sp_B32_aed_coverage                         0.992308   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_coverage        1.000000   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_coverage           0.992308   \n",
       "A_Rhodococcus_sp_B50_aed_coverage                     0.992308   \n",
       "A_Tomitella_gaofuii_HY172_aed_coverage                0.984733   \n",
       "PositiveControl_Min_coverage                          0.984733   \n",
       "\n",
       "                                                A_aedJ_I_RS26410  \\\n",
       "A_Dietzia_sp_B32_aed_coverage                           0.942164   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_coverage          0.992141   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_coverage             0.988235   \n",
       "A_Rhodococcus_sp_B50_aed_coverage                       0.992126   \n",
       "A_Tomitella_gaofuii_HY172_aed_coverage                  0.970817   \n",
       "PositiveControl_Min_coverage                            0.942164   \n",
       "\n",
       "                                                A_aedK_I_RS26415  \\\n",
       "A_Dietzia_sp_B32_aed_coverage                           0.989770   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_coverage          0.994859   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_coverage             0.989770   \n",
       "A_Rhodococcus_sp_B50_aed_coverage                       0.989770   \n",
       "A_Tomitella_gaofuii_HY172_aed_coverage                  0.989770   \n",
       "PositiveControl_Min_coverage                            0.989770   \n",
       "\n",
       "                                                A_aedL_RS26420  \\\n",
       "A_Dietzia_sp_B32_aed_coverage                         0.904762   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_coverage             NaN   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_coverage           0.917241   \n",
       "A_Rhodococcus_sp_B50_aed_coverage                     0.796407   \n",
       "A_Tomitella_gaofuii_HY172_aed_coverage                0.815951   \n",
       "PositiveControl_Min_coverage                          0.796407   \n",
       "\n",
       "                                                A_aedM_RS26425  \\\n",
       "A_Dietzia_sp_B32_aed_coverage                         0.988679   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_coverage        0.981273   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_coverage           0.988679   \n",
       "A_Rhodococcus_sp_B50_aed_coverage                     0.988679   \n",
       "A_Tomitella_gaofuii_HY172_aed_coverage                0.966292   \n",
       "PositiveControl_Min_coverage                          0.966292   \n",
       "\n",
       "                                                A_aedN_RS26430  \\\n",
       "A_Dietzia_sp_B32_aed_coverage                         1.000000   \n",
       "A_Mycobacteroides_chelonae_S00154_aed_coverage        0.992366   \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_coverage           0.997468   \n",
       "A_Rhodococcus_sp_B50_aed_coverage                     0.997468   \n",
       "A_Tomitella_gaofuii_HY172_aed_coverage                0.997468   \n",
       "PositiveControl_Min_coverage                          0.992366   \n",
       "\n",
       "                                                A_aedO_RS26435  A_aedP_RS26440  \n",
       "A_Dietzia_sp_B32_aed_coverage                         0.663793        0.984615  \n",
       "A_Mycobacteroides_chelonae_S00154_aed_coverage        0.714286        0.929095  \n",
       "A_Rhodococcus_equi_DSSKP_R_001_aed_coverage           0.689150        1.000000  \n",
       "A_Rhodococcus_sp_B50_aed_coverage                     0.899614        1.000000  \n",
       "A_Tomitella_gaofuii_HY172_aed_coverage                0.692308        1.000000  \n",
       "PositiveControl_Min_coverage                          0.663793        0.929095  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n"
     ]
    }
   ],
   "source": [
    "# HMM actino positive control bit-score threshold\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "#------------------------------------------------------------------------\n",
    "# create a function can get the datafarme of each control's best bit-score, e-vale and coverage dataframes\n",
    "# hmm domtblout name should be ...aed_I_... NOT ..._I_aed_....\n",
    "def positive_control_df(control_dir, control_names, BitScore_df, Evalue_df, Coverage_df):\n",
    "    # Create an empty dictionary to store the best bit-scores; best e-vale; best coverage for each hmmsearch\n",
    "    best_bit_scores = {}\n",
    "    evalue_scores = {}\n",
    "    coverage_scores = {}\n",
    "\n",
    "    # Loop over the hmmsearches and parse the corresponding \"domtblout\" file\n",
    "    for control_name in control_names:\n",
    "        # Load the \"domtblout\" file into a pandas DataFrame\n",
    "        file_path = os.path.join(control_dir, control_name + \".domtblout\")\n",
    "        try:\n",
    "            df = pd.read_csv(file_path, comment=\"#\", sep='\\s+', header=None)\n",
    "        except pd.errors.EmptyDataError:\n",
    "            df = pd.DataFrame()\n",
    "\n",
    "        if df.empty:\n",
    "            pattern = r'A_.*_aed'\n",
    "            query_name = re.sub(pattern, 'A_aed', control_name)        \n",
    "            best_bit_scores[query_name] = None\n",
    "            evalue_scores[query_name] = None\n",
    "            coverage_scores[query_name] = None\n",
    "        else:        \n",
    "            # Assign column names to the DataFrame\n",
    "            df.columns = [\"target_name\", \"accession\", \"tlen\", \"query_name\", \"accession2\", \"qlen\", \"E-value\", \"score\", \"bias\",\n",
    "                          \"num_domains_index\", \"num_domains_total\", \"c-Evalue\", \"i-Evalue\", \"score2\", \"bias2\", \"hmm_from\", \"hmm_to\", \"ali_from\", \"ali_to\",\n",
    "                          \"env_from\", \"env_to\", \"acc\", \"description\"]\n",
    "            # Calculate the coverage for each hit\n",
    "            df[\"coverage\"] = (df[\"ali_to\"] - df[\"ali_from\"] + 1) / df[\"tlen\"]\n",
    "\n",
    "            # Filter the DataFrame by E-value and coverage, and sort by bit-score\n",
    "            significant_hits = df[(df[\"E-value\"] < 0.001) & (df[\"coverage\"] > 0.50)].sort_values(by=\"score\", ascending=False)\n",
    "\n",
    "            # replace strains name to aed..\n",
    "            pattern = r'A_.*_aed'\n",
    "            query_name = re.sub(pattern, 'A_aed', control_name)\n",
    "\n",
    "            # Extract the best bit-score and store it in the dictionary\n",
    "            if not significant_hits.empty:\n",
    "                best_bit_score = significant_hits.iloc[0][\"score\"]\n",
    "                best_bit_scores[query_name] = best_bit_score\n",
    "                evalue_score = significant_hits.iloc[0][\"E-value\"]\n",
    "                evalue_scores[query_name] = evalue_score\n",
    "                coverage_score = significant_hits.iloc[0][\"coverage\"]\n",
    "                coverage_scores[query_name] = coverage_score\n",
    "            else:\n",
    "                best_bit_scores[query_name] = None\n",
    "                evalue_scores[query_name] = None\n",
    "                coverage_scores[query_name] = None\n",
    "\n",
    "    # create the index names for three df\n",
    "    # Get the strain name\n",
    "    pattern = r'A_.*_aed'\n",
    "    StrainName = re.findall(pattern, control_names[0])\n",
    "    \n",
    "    # create index of bit score\n",
    "    bitscore_name = StrainName[0] + '_bit_score'\n",
    "    BitScore_Name = [bitscore_name]\n",
    "\n",
    "    # create index of Evalue\n",
    "    evalue_name = StrainName[0] + '_Evalue'\n",
    "    Evalue_Name = [evalue_name]\n",
    "    \n",
    "    # create index of coverage\n",
    "    coverage_name = StrainName[0] + '_coverage'\n",
    "    Coverage_Name = [coverage_name]\n",
    "    \n",
    "    # create a dataframe of bit score\n",
    "    BitScore_df = pd.DataFrame(best_bit_scores, index=BitScore_Name)\n",
    "\n",
    "    # create a dataframe of e-value\n",
    "    Evalue_df = pd.DataFrame(evalue_scores, index=Evalue_Name)\n",
    "\n",
    "    # create a dataframe of coverage\n",
    "    Coverage_df = pd.DataFrame(coverage_scores, index=Coverage_Name)\n",
    "    \n",
    "    return BitScore_df, Evalue_df, Coverage_df\n",
    "\n",
    "\n",
    "#---------------------------------------------------------------\n",
    "# A_Dietzia_sp_B32\n",
    "# Define the directory that contains the \"domtblout\" files\n",
    "control_dir = \"../data/raw/Actino_HMM_Control/Positive/A_Dietzia_sp_B32/\"\n",
    "\n",
    "# Define the names of the control_names (without the file extension)\n",
    "# Do not use 'A_aedC_RS26365', 'A_aedQ_RS26445', 'A_aedR_RS26450' cause their Non-All homologous\n",
    "control_names = ['A_Dietzia_sp_B32_aedA_I_RS26385', 'A_Dietzia_sp_B32_aedB_I_RS26395', 'A_Dietzia_sp_B32_aedD_RS26370', 'A_Dietzia_sp_B32_aedE_RS26375', 'A_Dietzia_sp_B32_aedF_I_RS26380', 'A_Dietzia_sp_B32_aedG_I_RS26390', 'A_Dietzia_sp_B32_aedH_I_RS26400', 'A_Dietzia_sp_B32_aedI_RS26405', 'A_Dietzia_sp_B32_aedJ_I_RS26410', 'A_Dietzia_sp_B32_aedK_I_RS26415', 'A_Dietzia_sp_B32_aedL_RS26420', 'A_Dietzia_sp_B32_aedM_RS26425', 'A_Dietzia_sp_B32_aedN_RS26430', 'A_Dietzia_sp_B32_aedO_RS26435', 'A_Dietzia_sp_B32_aedP_RS26440']\n",
    "\n",
    "# Use positive_control_df function to get the df (control_dir, control_names, BitScore_df, Evalue_df, Coverage_df)\n",
    "B32_BitScore = pd.DataFrame()\n",
    "B32_Evalue = pd.DataFrame()\n",
    "B32_coverage = pd.DataFrame()\n",
    "\n",
    "B32_BitScore, B32_Evalue, B32_coverage = positive_control_df(control_dir, control_names, B32_BitScore, B32_Evalue, B32_coverage)\n",
    "\n",
    "\n",
    "#---------------------------------------------------------------\n",
    "# A_Mycobacteroides_chelonae_S00154\n",
    "# Define the directory that contains the \"domtblout\" files\n",
    "control_dir = \"../data/raw/Actino_HMM_Control/Positive/A_Mycobacteroides_chelonae_S00154/\"\n",
    "\n",
    "# Define the names of the control_names (without the file extension)\n",
    "# Do not use 'A_aedC_RS26365', 'A_aedQ_RS26445', 'A_aedR_RS26450' cause their Non-All homologous\n",
    "control_names = ['A_Mycobacteroides_chelonae_S00154_aedA_I_RS26385', 'A_Mycobacteroides_chelonae_S00154_aedB_I_RS26395', 'A_Mycobacteroides_chelonae_S00154_aedD_RS26370', 'A_Mycobacteroides_chelonae_S00154_aedE_RS26375', 'A_Mycobacteroides_chelonae_S00154_aedF_I_RS26380', 'A_Mycobacteroides_chelonae_S00154_aedG_I_RS26390', 'A_Mycobacteroides_chelonae_S00154_aedH_I_RS26400', 'A_Mycobacteroides_chelonae_S00154_aedI_RS26405', 'A_Mycobacteroides_chelonae_S00154_aedJ_I_RS26410', 'A_Mycobacteroides_chelonae_S00154_aedK_I_RS26415', 'A_Mycobacteroides_chelonae_S00154_aedL_RS26420', 'A_Mycobacteroides_chelonae_S00154_aedM_RS26425', 'A_Mycobacteroides_chelonae_S00154_aedN_RS26430', 'A_Mycobacteroides_chelonae_S00154_aedO_RS26435', 'A_Mycobacteroides_chelonae_S00154_aedP_RS26440']\n",
    "\n",
    "# Use positive_control_df function to get the df (control_dir, control_names, BitScore_df, Evalue_df, Coverage_df)\n",
    "S00154_BitScore = pd.DataFrame()\n",
    "S00154_Evalue = pd.DataFrame()\n",
    "S00154_coverage = pd.DataFrame()\n",
    "\n",
    "S00154_BitScore, S00154_Evalue, S00154_coverage = positive_control_df(control_dir, control_names, S00154_BitScore, S00154_Evalue, S00154_coverage)\n",
    "\n",
    "\n",
    "#---------------------------------------------------------------\n",
    "# A_Rhodococcus_equi_DSSKP_R_001\n",
    "# Define the directory that contains the \"domtblout\" files\n",
    "control_dir = \"../data/raw/Actino_HMM_Control/Positive/A_Rhodococcus_equi_DSSKP_R_001/\"\n",
    "\n",
    "# Define the names of the control_names (without the file extension)\n",
    "# Do not use 'A_aedC_RS26365', 'A_aedQ_RS26445', 'A_aedR_RS26450' cause their Non-All homologous\n",
    "control_names = ['A_Rhodococcus_equi_DSSKP_R_001_aedA_I_RS26385', 'A_Rhodococcus_equi_DSSKP_R_001_aedB_I_RS26395', 'A_Rhodococcus_equi_DSSKP_R_001_aedD_RS26370', 'A_Rhodococcus_equi_DSSKP_R_001_aedE_RS26375', 'A_Rhodococcus_equi_DSSKP_R_001_aedF_I_RS26380', 'A_Rhodococcus_equi_DSSKP_R_001_aedG_I_RS26390', 'A_Rhodococcus_equi_DSSKP_R_001_aedH_I_RS26400', 'A_Rhodococcus_equi_DSSKP_R_001_aedI_RS26405', 'A_Rhodococcus_equi_DSSKP_R_001_aedJ_I_RS26410', 'A_Rhodococcus_equi_DSSKP_R_001_aedK_I_RS26415', 'A_Rhodococcus_equi_DSSKP_R_001_aedL_RS26420', 'A_Rhodococcus_equi_DSSKP_R_001_aedM_RS26425', 'A_Rhodococcus_equi_DSSKP_R_001_aedN_RS26430', 'A_Rhodococcus_equi_DSSKP_R_001_aedO_RS26435', 'A_Rhodococcus_equi_DSSKP_R_001_aedP_RS26440']\n",
    "\n",
    "# Use positive_control_df function to get the df (control_dir, control_names, BitScore_df, Evalue_df, Coverage_df)\n",
    "R_001_BitScore = pd.DataFrame()\n",
    "R_001_Evalue = pd.DataFrame()\n",
    "R_001_coverage = pd.DataFrame()\n",
    "\n",
    "R_001_BitScore, R_001_Evalue, R_001_coverage = positive_control_df(control_dir, control_names, R_001_BitScore, R_001_Evalue, R_001_coverage)\n",
    "\n",
    "\n",
    "#---------------------------------------------------------------\n",
    "# A_Rhodococcus_sp_B50\n",
    "# Define the directory that contains the \"domtblout\" files\n",
    "control_dir = \"../data/raw/Actino_HMM_Control/Positive/A_Rhodococcus_sp_B50/\"\n",
    "\n",
    "# Define the names of the control_names (without the file extension)\n",
    "# Do not use 'A_aedC_RS26365', 'A_aedQ_RS26445', 'A_aedR_RS26450' cause their Non-All homologous\n",
    "control_names = ['A_Rhodococcus_sp_B50_aedA_I_RS26385', 'A_Rhodococcus_sp_B50_aedB_I_RS26395', 'A_Rhodococcus_sp_B50_aedD_RS26370', 'A_Rhodococcus_sp_B50_aedE_RS26375', 'A_Rhodococcus_sp_B50_aedF_I_RS26380', 'A_Rhodococcus_sp_B50_aedG_I_RS26390', 'A_Rhodococcus_sp_B50_aedH_I_RS26400', 'A_Rhodococcus_sp_B50_aedI_RS26405', 'A_Rhodococcus_sp_B50_aedJ_I_RS26410', 'A_Rhodococcus_sp_B50_aedK_I_RS26415', 'A_Rhodococcus_sp_B50_aedL_RS26420', 'A_Rhodococcus_sp_B50_aedM_RS26425', 'A_Rhodococcus_sp_B50_aedN_RS26430', 'A_Rhodococcus_sp_B50_aedO_RS26435', 'A_Rhodococcus_sp_B50_aedP_RS26440']\n",
    "\n",
    "# Use positive_control_df function to get the df (control_dir, control_names, BitScore_df, Evalue_df, Coverage_df)\n",
    "B50_BitScore = pd.DataFrame()\n",
    "B50_Evalue = pd.DataFrame()\n",
    "B50_coverage = pd.DataFrame()\n",
    "\n",
    "B50_BitScore, B50_Evalue, B50_coverage = positive_control_df(control_dir, control_names, B50_BitScore, B50_Evalue, B50_coverage)\n",
    "\n",
    "\n",
    "#---------------------------------------------------------------\n",
    "# A_Tomitella_gaofuii_HY172\n",
    "# Define the directory that contains the \"domtblout\" files\n",
    "control_dir = \"../data/raw/Actino_HMM_Control/Positive/A_Tomitella_gaofuii_HY172/\"\n",
    "\n",
    "# Define the names of the control_names (without the file extension)\n",
    "# Do not use 'A_aedC_RS26365', 'A_aedQ_RS26445', 'A_aedR_RS26450' cause their Non-All homologous\n",
    "control_names = ['A_Tomitella_gaofuii_HY172_aedA_I_RS26385', 'A_Tomitella_gaofuii_HY172_aedB_I_RS26395', 'A_Tomitella_gaofuii_HY172_aedD_RS26370', 'A_Tomitella_gaofuii_HY172_aedE_RS26375', 'A_Tomitella_gaofuii_HY172_aedF_I_RS26380', 'A_Tomitella_gaofuii_HY172_aedG_I_RS26390', 'A_Tomitella_gaofuii_HY172_aedH_I_RS26400', 'A_Tomitella_gaofuii_HY172_aedI_RS26405', 'A_Tomitella_gaofuii_HY172_aedJ_I_RS26410', 'A_Tomitella_gaofuii_HY172_aedK_I_RS26415', 'A_Tomitella_gaofuii_HY172_aedL_RS26420', 'A_Tomitella_gaofuii_HY172_aedM_RS26425', 'A_Tomitella_gaofuii_HY172_aedN_RS26430', 'A_Tomitella_gaofuii_HY172_aedO_RS26435', 'A_Tomitella_gaofuii_HY172_aedP_RS26440']\n",
    "\n",
    "# Use positive_control_df function to get the df (control_dir, control_names, BitScore_df, Evalue_df, Coverage_df)\n",
    "HY172_BitScore = pd.DataFrame()\n",
    "HY172_Evalue = pd.DataFrame()\n",
    "HY172_coverage = pd.DataFrame()\n",
    "\n",
    "HY172_BitScore, HY172_Evalue, HY172_coverage = positive_control_df(control_dir, control_names, HY172_BitScore, HY172_Evalue, HY172_coverage)\n",
    "\n",
    "\n",
    "#-------------------------------------------------------------------------------------------------------\n",
    "# Merge tabel and create a df of max and min\n",
    "\n",
    "# Got the min bit score of the df\n",
    "# merge 5 best bit score df\n",
    "Actino_PositiveBitScore_df = pd.concat([B32_BitScore, S00154_BitScore, R_001_BitScore, B50_BitScore, HY172_BitScore], axis=0)\n",
    "\n",
    "Actino_MiniBitScore_df = Actino_PositiveBitScore_df.min()\n",
    "Actino_MiniBitScore_df = Actino_MiniBitScore_df.iloc[0:]\n",
    "Actino_MiniBitScore_df = Actino_MiniBitScore_df.to_frame()\n",
    "Actino_MiniBitScore_df.columns = ['PositiveControl_Min_BitScore']\n",
    "\n",
    "# transpose the DataFrame\n",
    "Actino_MiniBitScore_df = Actino_MiniBitScore_df.T\n",
    "\n",
    "# merge all nad Min table\n",
    "Actino_PositiveBitScore_df = pd.concat([Actino_PositiveBitScore_df, Actino_MiniBitScore_df], axis=0)\n",
    "Actino_PositiveBitScore_df.to_csv('../data/processed/Final/ForReader/ControlData/Actino_aed/Actino_PositiveBitScore.csv')\n",
    "\n",
    "\n",
    "# Got the max e-value of the df\n",
    "# merge 5 e-value df\n",
    "Actino_PositiveEvalue_df = pd.concat([B32_Evalue, S00154_Evalue, R_001_Evalue, B50_Evalue, HY172_Evalue], axis=0)\n",
    "\n",
    "Actino_MaxEvalue_df = Actino_PositiveEvalue_df.max()\n",
    "Actino_MaxEvalue_df = Actino_MaxEvalue_df.iloc[0:]\n",
    "Actino_MaxEvalue_df = Actino_MaxEvalue_df.to_frame()\n",
    "Actino_MaxEvalue_df.columns = ['PositiveControl_Max_Evalue']\n",
    "\n",
    "# transpose the DataFrame\n",
    "Actino_MaxEvalue_df = Actino_MaxEvalue_df.T\n",
    "\n",
    "# merge all nad Min table\n",
    "Actino_PositiveEvalue_df = pd.concat([Actino_PositiveEvalue_df, Actino_MaxEvalue_df], axis=0)\n",
    "Actino_PositiveEvalue_df.to_csv('../data/processed/Final/ForReader/ControlData/Actino_aed/Actino_PositiveEvalue.csv')\n",
    "\n",
    "\n",
    "# Got the min coverage of the df\n",
    "# merge 5 e-value df\n",
    "Actino_Positivecoverage_df = pd.concat([B32_coverage, S00154_coverage, R_001_coverage, B50_coverage, HY172_coverage], axis=0)\n",
    "\n",
    "Actino_Minicoverage_df = Actino_Positivecoverage_df.min()\n",
    "Actino_Minicoverage_df = Actino_Minicoverage_df.iloc[0:]\n",
    "Actino_Minicoverage_df = Actino_Minicoverage_df.to_frame()\n",
    "Actino_Minicoverage_df.columns = ['PositiveControl_Min_coverage']\n",
    "\n",
    "# transpose the DataFrame\n",
    "Actino_Minicoverage_df = Actino_Minicoverage_df.T\n",
    "\n",
    "# merge all nad Min table\n",
    "Actino_Positivecoverage_df = pd.concat([Actino_Positivecoverage_df, Actino_Minicoverage_df], axis=0)\n",
    "Actino_Positivecoverage_df.to_csv('../data/processed/Final/ForReader/ControlData/Actino_aed/Actino_PositiveCoverage.csv')\n",
    "\n",
    "\n",
    "# Display three table\n",
    "display(Actino_PositiveBitScore_df)\n",
    "display(Actino_PositiveEvalue_df)\n",
    "display(Actino_Positivecoverage_df)\n",
    "print('Done')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90e67cf3-2713-4d48-bf33-341b1a339932",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 1.2 使用 negative control 來獲得合適的bit score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ca98d08a-752f-4878-ba4d-ac2898378b1a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Criteria_Bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A_aedA_I_RS26385</th>\n",
       "      <td>195.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_aedB_I_RS26395</th>\n",
       "      <td>220.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_aedD_RS26370</th>\n",
       "      <td>128.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_aedE_RS26375</th>\n",
       "      <td>232.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_aedF_I_RS26380</th>\n",
       "      <td>617.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_aedG_I_RS26390</th>\n",
       "      <td>176.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_aedH_I_RS26400</th>\n",
       "      <td>270.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_aedI_RS26405</th>\n",
       "      <td>25.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_aedJ_I_RS26410</th>\n",
       "      <td>328.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_aedK_I_RS26415</th>\n",
       "      <td>321.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_aedL_RS26420</th>\n",
       "      <td>106.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_aedM_RS26425</th>\n",
       "      <td>98.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_aedN_RS26430</th>\n",
       "      <td>221.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_aedO_RS26435</th>\n",
       "      <td>58.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_aedP_RS26440</th>\n",
       "      <td>181.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Criteria_Bitscore\n",
       "A_aedA_I_RS26385              195.4\n",
       "A_aedB_I_RS26395              220.3\n",
       "A_aedD_RS26370                128.9\n",
       "A_aedE_RS26375                232.1\n",
       "A_aedF_I_RS26380              617.3\n",
       "A_aedG_I_RS26390              176.2\n",
       "A_aedH_I_RS26400              270.8\n",
       "A_aedI_RS26405                 25.4\n",
       "A_aedJ_I_RS26410              328.3\n",
       "A_aedK_I_RS26415              321.5\n",
       "A_aedL_RS26420                106.7\n",
       "A_aedM_RS26425                 98.3\n",
       "A_aedN_RS26430                221.1\n",
       "A_aedO_RS26435                 58.9\n",
       "A_aedP_RS26440                181.3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A_aedA_I_RS26385</th>\n",
       "      <th>A_aedB_I_RS26395</th>\n",
       "      <th>A_aedD_RS26370</th>\n",
       "      <th>A_aedE_RS26375</th>\n",
       "      <th>A_aedF_I_RS26380</th>\n",
       "      <th>A_aedG_I_RS26390</th>\n",
       "      <th>A_aedH_I_RS26400</th>\n",
       "      <th>A_aedI_RS26405</th>\n",
       "      <th>A_aedJ_I_RS26410</th>\n",
       "      <th>A_aedK_I_RS26415</th>\n",
       "      <th>A_aedL_RS26420</th>\n",
       "      <th>A_aedM_RS26425</th>\n",
       "      <th>A_aedN_RS26430</th>\n",
       "      <th>A_aedO_RS26435</th>\n",
       "      <th>A_aedP_RS26440</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A_Rhodococcus_jostii_RHA1_aed_bit_score</th>\n",
       "      <td>189.0</td>\n",
       "      <td>220.3</td>\n",
       "      <td>126.7</td>\n",
       "      <td>232.1</td>\n",
       "      <td>617.3</td>\n",
       "      <td>176.2</td>\n",
       "      <td>270.8</td>\n",
       "      <td>25.4</td>\n",
       "      <td>328.3</td>\n",
       "      <td>321.5</td>\n",
       "      <td>106.7</td>\n",
       "      <td>98.3</td>\n",
       "      <td>221.1</td>\n",
       "      <td>58.9</td>\n",
       "      <td>181.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_Mycobacterium_tuberculosis_H37Rv_aed_bit_score</th>\n",
       "      <td>195.4</td>\n",
       "      <td>213.7</td>\n",
       "      <td>128.9</td>\n",
       "      <td>223.7</td>\n",
       "      <td>568.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>30.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>252.0</td>\n",
       "      <td>304.2</td>\n",
       "      <td>20.3</td>\n",
       "      <td>94.6</td>\n",
       "      <td>203.3</td>\n",
       "      <td>55.3</td>\n",
       "      <td>149.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Criteria_Bitscore</th>\n",
       "      <td>195.4</td>\n",
       "      <td>220.3</td>\n",
       "      <td>128.9</td>\n",
       "      <td>232.1</td>\n",
       "      <td>617.3</td>\n",
       "      <td>176.2</td>\n",
       "      <td>270.8</td>\n",
       "      <td>25.4</td>\n",
       "      <td>328.3</td>\n",
       "      <td>321.5</td>\n",
       "      <td>106.7</td>\n",
       "      <td>98.3</td>\n",
       "      <td>221.1</td>\n",
       "      <td>58.9</td>\n",
       "      <td>181.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  A_aedA_I_RS26385  \\\n",
       "A_Rhodococcus_jostii_RHA1_aed_bit_score                      189.0   \n",
       "A_Mycobacterium_tuberculosis_H37Rv_aed_bit_score             195.4   \n",
       "Criteria_Bitscore                                            195.4   \n",
       "\n",
       "                                                  A_aedB_I_RS26395  \\\n",
       "A_Rhodococcus_jostii_RHA1_aed_bit_score                      220.3   \n",
       "A_Mycobacterium_tuberculosis_H37Rv_aed_bit_score             213.7   \n",
       "Criteria_Bitscore                                            220.3   \n",
       "\n",
       "                                                  A_aedD_RS26370  \\\n",
       "A_Rhodococcus_jostii_RHA1_aed_bit_score                    126.7   \n",
       "A_Mycobacterium_tuberculosis_H37Rv_aed_bit_score           128.9   \n",
       "Criteria_Bitscore                                          128.9   \n",
       "\n",
       "                                                  A_aedE_RS26375  \\\n",
       "A_Rhodococcus_jostii_RHA1_aed_bit_score                    232.1   \n",
       "A_Mycobacterium_tuberculosis_H37Rv_aed_bit_score           223.7   \n",
       "Criteria_Bitscore                                          232.1   \n",
       "\n",
       "                                                  A_aedF_I_RS26380  \\\n",
       "A_Rhodococcus_jostii_RHA1_aed_bit_score                      617.3   \n",
       "A_Mycobacterium_tuberculosis_H37Rv_aed_bit_score             568.5   \n",
       "Criteria_Bitscore                                            617.3   \n",
       "\n",
       "                                                  A_aedG_I_RS26390  \\\n",
       "A_Rhodococcus_jostii_RHA1_aed_bit_score                      176.2   \n",
       "A_Mycobacterium_tuberculosis_H37Rv_aed_bit_score               0.0   \n",
       "Criteria_Bitscore                                            176.2   \n",
       "\n",
       "                                                  A_aedH_I_RS26400  \\\n",
       "A_Rhodococcus_jostii_RHA1_aed_bit_score                      270.8   \n",
       "A_Mycobacterium_tuberculosis_H37Rv_aed_bit_score              30.5   \n",
       "Criteria_Bitscore                                            270.8   \n",
       "\n",
       "                                                  A_aedI_RS26405  \\\n",
       "A_Rhodococcus_jostii_RHA1_aed_bit_score                     25.4   \n",
       "A_Mycobacterium_tuberculosis_H37Rv_aed_bit_score             0.0   \n",
       "Criteria_Bitscore                                           25.4   \n",
       "\n",
       "                                                  A_aedJ_I_RS26410  \\\n",
       "A_Rhodococcus_jostii_RHA1_aed_bit_score                      328.3   \n",
       "A_Mycobacterium_tuberculosis_H37Rv_aed_bit_score             252.0   \n",
       "Criteria_Bitscore                                            328.3   \n",
       "\n",
       "                                                  A_aedK_I_RS26415  \\\n",
       "A_Rhodococcus_jostii_RHA1_aed_bit_score                      321.5   \n",
       "A_Mycobacterium_tuberculosis_H37Rv_aed_bit_score             304.2   \n",
       "Criteria_Bitscore                                            321.5   \n",
       "\n",
       "                                                  A_aedL_RS26420  \\\n",
       "A_Rhodococcus_jostii_RHA1_aed_bit_score                    106.7   \n",
       "A_Mycobacterium_tuberculosis_H37Rv_aed_bit_score            20.3   \n",
       "Criteria_Bitscore                                          106.7   \n",
       "\n",
       "                                                  A_aedM_RS26425  \\\n",
       "A_Rhodococcus_jostii_RHA1_aed_bit_score                     98.3   \n",
       "A_Mycobacterium_tuberculosis_H37Rv_aed_bit_score            94.6   \n",
       "Criteria_Bitscore                                           98.3   \n",
       "\n",
       "                                                  A_aedN_RS26430  \\\n",
       "A_Rhodococcus_jostii_RHA1_aed_bit_score                    221.1   \n",
       "A_Mycobacterium_tuberculosis_H37Rv_aed_bit_score           203.3   \n",
       "Criteria_Bitscore                                          221.1   \n",
       "\n",
       "                                                  A_aedO_RS26435  \\\n",
       "A_Rhodococcus_jostii_RHA1_aed_bit_score                     58.9   \n",
       "A_Mycobacterium_tuberculosis_H37Rv_aed_bit_score            55.3   \n",
       "Criteria_Bitscore                                           58.9   \n",
       "\n",
       "                                                  A_aedP_RS26440  \n",
       "A_Rhodococcus_jostii_RHA1_aed_bit_score                    181.3  \n",
       "A_Mycobacterium_tuberculosis_H37Rv_aed_bit_score           149.9  \n",
       "Criteria_Bitscore                                          181.3  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "# HMM actino negative control bit-score threshold\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "#------------------------------------------------------------------------\n",
    "# create a function can get the datafarme of each control's best bit-score, e-vale and coverage dataframes\n",
    "# hmm domtblout name should be ...aed_I_... NOT ..._I_aed_....\n",
    "def negative_control_df(control_dir, control_names, BitScore_df):\n",
    "    # Create an empty dictionary to store the best bit-scores; best e-vale; best coverage for each hmmsearch\n",
    "    best_bit_scores = {}\n",
    "\n",
    "    # Loop over the hmmsearches and parse the corresponding \"domtblout\" file\n",
    "    for control_name in control_names:\n",
    "        # Load the \"domtblout\" file into a pandas DataFrame\n",
    "        file_path = os.path.join(control_dir, control_name + \".domtblout\")\n",
    "        try:\n",
    "            df = pd.read_csv(file_path, comment=\"#\", sep='\\s+', header=None)\n",
    "        except pd.errors.EmptyDataError:\n",
    "            df = pd.DataFrame()\n",
    "\n",
    "        if df.empty:\n",
    "            pattern = r'A_.*_aed'\n",
    "            query_name = re.sub(pattern, 'A_aed', control_name)        \n",
    "            best_bit_scores[query_name] = 0\n",
    "        else:        \n",
    "            # Assign column names to the DataFrame\n",
    "            df.columns = [\"target_name\", \"accession\", \"tlen\", \"query_name\", \"accession2\", \"qlen\", \"E-value\", \"score\", \"bias\",\n",
    "                          \"num_domains_index\", \"num_domains_total\", \"c-Evalue\", \"i-Evalue\", \"score2\", \"bias2\", \"hmm_from\", \"hmm_to\", \"ali_from\", \"ali_to\",\n",
    "                          \"env_from\", \"env_to\", \"acc\", \"description\"]\n",
    "            # Calculate the coverage for each hit\n",
    "            df[\"coverage\"] = (df[\"ali_to\"] - df[\"ali_from\"] + 1) / df[\"tlen\"]\n",
    "\n",
    "            # Filter the DataFrame by E-value and coverage, and sort by bit-score\n",
    "            significant_hits = df[(df[\"E-value\"] < 0.001) & (df[\"coverage\"] > 0.50)].sort_values(by=\"score\", ascending=False)\n",
    "\n",
    "            # replace strains name to aed..\n",
    "            pattern = r'A_.*_aed'\n",
    "            query_name = re.sub(pattern, 'A_aed', control_name)\n",
    "\n",
    "            # Extract the best bit-score and store it in the dictionary\n",
    "            if not significant_hits.empty:\n",
    "                best_bit_score = significant_hits.iloc[0][\"score\"]\n",
    "                best_bit_scores[query_name] = best_bit_score\n",
    "            else:\n",
    "                best_bit_scores[query_name] = 0\n",
    "\n",
    "    # create the index names for three df\n",
    "    # Get the strain name\n",
    "    pattern = r'A_.*_aed'\n",
    "    StrainName = re.findall(pattern, control_names[0])\n",
    "    \n",
    "    # create index of bit score\n",
    "    bitscore_name = StrainName[0] + '_bit_score'\n",
    "    BitScore_Name = [bitscore_name]\n",
    "\n",
    "    # create a dataframe of bit score\n",
    "    BitScore_df = pd.DataFrame(best_bit_scores, index=BitScore_Name)\n",
    "\n",
    "    return BitScore_df\n",
    "\n",
    "\n",
    "#---------------------------------------------------------------\n",
    "# A_Rhodococcus_jostii_RHA1\n",
    "# Define the directory that contains the \"domtblout\" files\n",
    "control_dir = \"../data/raw/Actino_HMM_Control/Negative/A_Rhodococcus_jostii_RHA1/domtblout/\"\n",
    "\n",
    "# Define the names of the control_names (without the file extension)\n",
    "# Do not use 'A_aedC_RS26365', 'A_aedQ_RS26445', 'A_aedR_RS26450' cause their Non-All homologous\n",
    "control_names = ['A_Rhodococcus_jostii_RHA1_aedA_I_RS26385', 'A_Rhodococcus_jostii_RHA1_aedB_I_RS26395', 'A_Rhodococcus_jostii_RHA1_aedD_RS26370', 'A_Rhodococcus_jostii_RHA1_aedE_RS26375', 'A_Rhodococcus_jostii_RHA1_aedF_I_RS26380', 'A_Rhodococcus_jostii_RHA1_aedG_I_RS26390', 'A_Rhodococcus_jostii_RHA1_aedH_I_RS26400', 'A_Rhodococcus_jostii_RHA1_aedI_RS26405', 'A_Rhodococcus_jostii_RHA1_aedJ_I_RS26410', 'A_Rhodococcus_jostii_RHA1_aedK_I_RS26415', 'A_Rhodococcus_jostii_RHA1_aedL_RS26420', 'A_Rhodococcus_jostii_RHA1_aedM_RS26425', 'A_Rhodococcus_jostii_RHA1_aedN_RS26430', 'A_Rhodococcus_jostii_RHA1_aedO_RS26435', 'A_Rhodococcus_jostii_RHA1_aedP_RS26440']\n",
    "\n",
    "# Use positive_control_df function to get the df (control_dir, control_names, BitScore_df, Evalue_df, Coverage_df)\n",
    "RHA1_BitScore = pd.DataFrame()\n",
    "RHA1_BitScore = negative_control_df(control_dir, control_names, RHA1_BitScore)\n",
    "\n",
    "\n",
    "#---------------------------------------------------------------\n",
    "# A_Mycobacterium_tuberculosis_H37Rv\n",
    "# Define the directory that contains the \"domtblout\" files\n",
    "control_dir = \"../data/raw/Actino_HMM_Control/Negative/A_Mycobacterium_tuberculosis_H37Rv/domtblout\"\n",
    "\n",
    "# Define the names of the control_names (without the file extension)\n",
    "# Do not use 'A_aedC_RS26365', 'A_aedQ_RS26445', 'A_aedR_RS26450' cause their Non-All homologous\n",
    "control_names = ['A_Mycobacterium_tuberculosis_H37Rv_aedA_I_RS26385', 'A_Mycobacterium_tuberculosis_H37Rv_aedB_I_RS26395', 'A_Mycobacterium_tuberculosis_H37Rv_aedD_RS26370', 'A_Mycobacterium_tuberculosis_H37Rv_aedE_RS26375', 'A_Mycobacterium_tuberculosis_H37Rv_aedF_I_RS26380', 'A_Mycobacterium_tuberculosis_H37Rv_aedG_I_RS26390', 'A_Mycobacterium_tuberculosis_H37Rv_aedH_I_RS26400', 'A_Mycobacterium_tuberculosis_H37Rv_aedI_RS26405', 'A_Mycobacterium_tuberculosis_H37Rv_aedJ_I_RS26410', 'A_Mycobacterium_tuberculosis_H37Rv_aedK_I_RS26415', 'A_Mycobacterium_tuberculosis_H37Rv_aedL_RS26420', 'A_Mycobacterium_tuberculosis_H37Rv_aedM_RS26425', 'A_Mycobacterium_tuberculosis_H37Rv_aedN_RS26430', 'A_Mycobacterium_tuberculosis_H37Rv_aedO_RS26435', 'A_Mycobacterium_tuberculosis_H37Rv_aedP_RS26440']\n",
    "\n",
    "# Use positive_control_df function to get the df (control_dir, control_names, BitScore_df, Evalue_df, Coverage_df)\n",
    "H37Rv_BitScore = pd.DataFrame()\n",
    "H37Rv_BitScore = negative_control_df(control_dir, control_names, H37Rv_BitScore)\n",
    "\n",
    "\n",
    "#-------------------------------------------------------------------------------------------------------\n",
    "# merge two negative best bit score df\n",
    "Actino_Negative_df = pd.concat([RHA1_BitScore, H37Rv_BitScore], axis=0)\n",
    "\n",
    "# Got the highest bit score of the df\n",
    "Actino_BitScore_Criteria = Actino_Negative_df.max()\n",
    "Actino_BitScore_Criteria = Actino_BitScore_Criteria.iloc[0:]\n",
    "Actino_BitScore_Criteria = Actino_BitScore_Criteria.to_frame()\n",
    "Actino_BitScore_Criteria.columns = ['Criteria_Bitscore']\n",
    "\n",
    "# transpose the DataFrame\n",
    "Actino_BitScore_Criteria_T = Actino_BitScore_Criteria.T\n",
    "\n",
    "# merge all nad Min table\n",
    "Actino_Negative_df = pd.concat([Actino_Negative_df, Actino_BitScore_Criteria_T], axis=0)\n",
    "Actino_Negative_df.to_csv('../data/processed/Final/ForReader/ControlData/Actino_aed/Actino_NegativeBitscore.csv')\n",
    "\n",
    "# done\n",
    "display(Actino_BitScore_Criteria)\n",
    "display(Actino_Negative_df)\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf3352cb-e35c-4bc7-b166-9bed92700fd9",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 2. 使用上述的 bit score 來篩選 MAGS 的 outpout data (不含有 hits 的篩選)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "41047cf5-75d7-44fd-9e75-048f5b87b1c4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n",
      "unique query name/numbers:  ['Q_A_aedP_RS26440' 'Q_A_aedO_RS26435' 'Q_A_aedN_RS26430'\n",
      " 'Q_A_aedM_RS26425' 'Q_A_aedL_RS26420' 'Q_A_I_aedK_RS26415'\n",
      " 'Q_A_I_aedJ_RS26410' 'Q_A_aedI_RS26405' 'Q_A_I_aedH_RS26400'\n",
      " 'Q_A_I_aedG_RS26390' 'Q_A_I_aedF_RS26380' 'Q_A_aedE_RS26375'\n",
      " 'Q_A_aedD_RS26370' 'Q_A_I_aedB_RS26395' 'Q_A_I_aedA_RS26385']  /  15\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_name</th>\n",
       "      <th>accession</th>\n",
       "      <th>tlen</th>\n",
       "      <th>query_name</th>\n",
       "      <th>accession2</th>\n",
       "      <th>qlen</th>\n",
       "      <th>E-value</th>\n",
       "      <th>score</th>\n",
       "      <th>bias</th>\n",
       "      <th>num_domains_index</th>\n",
       "      <th>...</th>\n",
       "      <th>bias2</th>\n",
       "      <th>hmm_from</th>\n",
       "      <th>hmm_to</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>ali_to</th>\n",
       "      <th>env_from</th>\n",
       "      <th>env_to</th>\n",
       "      <th>acc</th>\n",
       "      <th>description</th>\n",
       "      <th>coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3300009702.a:Ga0114931_10050376_3</td>\n",
       "      <td>-</td>\n",
       "      <td>376</td>\n",
       "      <td>Q_A_aedP_RS26440</td>\n",
       "      <td>-</td>\n",
       "      <td>385</td>\n",
       "      <td>6.300000e-116</td>\n",
       "      <td>400.0</td>\n",
       "      <td>2.4</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2.4</td>\n",
       "      <td>1</td>\n",
       "      <td>372</td>\n",
       "      <td>1</td>\n",
       "      <td>374</td>\n",
       "      <td>1</td>\n",
       "      <td>375</td>\n",
       "      <td>0.99</td>\n",
       "      <td>-</td>\n",
       "      <td>0.994681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3300009702.a:Ga0114931_10022461_4</td>\n",
       "      <td>-</td>\n",
       "      <td>376</td>\n",
       "      <td>Q_A_aedP_RS26440</td>\n",
       "      <td>-</td>\n",
       "      <td>385</td>\n",
       "      <td>7.400000e-116</td>\n",
       "      <td>399.8</td>\n",
       "      <td>5.1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>5.1</td>\n",
       "      <td>1</td>\n",
       "      <td>372</td>\n",
       "      <td>1</td>\n",
       "      <td>374</td>\n",
       "      <td>1</td>\n",
       "      <td>375</td>\n",
       "      <td>0.99</td>\n",
       "      <td>-</td>\n",
       "      <td>0.994681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3300027968.a:Ga0209061_1000039_348</td>\n",
       "      <td>-</td>\n",
       "      <td>378</td>\n",
       "      <td>Q_A_aedP_RS26440</td>\n",
       "      <td>-</td>\n",
       "      <td>385</td>\n",
       "      <td>2.300000e-112</td>\n",
       "      <td>388.3</td>\n",
       "      <td>3.1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>3.1</td>\n",
       "      <td>1</td>\n",
       "      <td>372</td>\n",
       "      <td>2</td>\n",
       "      <td>375</td>\n",
       "      <td>2</td>\n",
       "      <td>377</td>\n",
       "      <td>0.98</td>\n",
       "      <td>-</td>\n",
       "      <td>0.989418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3300027662.a:Ga0208565_1003173_4</td>\n",
       "      <td>-</td>\n",
       "      <td>380</td>\n",
       "      <td>Q_A_aedP_RS26440</td>\n",
       "      <td>-</td>\n",
       "      <td>385</td>\n",
       "      <td>1.300000e-111</td>\n",
       "      <td>385.8</td>\n",
       "      <td>3.6</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>3.6</td>\n",
       "      <td>1</td>\n",
       "      <td>375</td>\n",
       "      <td>2</td>\n",
       "      <td>378</td>\n",
       "      <td>2</td>\n",
       "      <td>379</td>\n",
       "      <td>0.99</td>\n",
       "      <td>-</td>\n",
       "      <td>0.992105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3300017444.a:Ga0185300_10000551_65</td>\n",
       "      <td>-</td>\n",
       "      <td>375</td>\n",
       "      <td>Q_A_aedP_RS26440</td>\n",
       "      <td>-</td>\n",
       "      <td>385</td>\n",
       "      <td>3.500000e-111</td>\n",
       "      <td>384.4</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1</td>\n",
       "      <td>372</td>\n",
       "      <td>1</td>\n",
       "      <td>374</td>\n",
       "      <td>1</td>\n",
       "      <td>374</td>\n",
       "      <td>0.99</td>\n",
       "      <td>-</td>\n",
       "      <td>0.997333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>3300014092.a:Ga0121367_100110_13</td>\n",
       "      <td>-</td>\n",
       "      <td>405</td>\n",
       "      <td>Q_A_I_aedA_RS26385</td>\n",
       "      <td>-</td>\n",
       "      <td>398</td>\n",
       "      <td>6.200000e-54</td>\n",
       "      <td>196.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6</td>\n",
       "      <td>395</td>\n",
       "      <td>8</td>\n",
       "      <td>398</td>\n",
       "      <td>3</td>\n",
       "      <td>401</td>\n",
       "      <td>0.93</td>\n",
       "      <td>-</td>\n",
       "      <td>0.965432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>3300027968.a:Ga0209061_1000140_88</td>\n",
       "      <td>-</td>\n",
       "      <td>412</td>\n",
       "      <td>Q_A_I_aedA_RS26385</td>\n",
       "      <td>-</td>\n",
       "      <td>398</td>\n",
       "      <td>6.800000e-54</td>\n",
       "      <td>196.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15</td>\n",
       "      <td>396</td>\n",
       "      <td>17</td>\n",
       "      <td>405</td>\n",
       "      <td>4</td>\n",
       "      <td>407</td>\n",
       "      <td>0.89</td>\n",
       "      <td>-</td>\n",
       "      <td>0.944175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>3300017532.a:Ga0185302_10010386_3</td>\n",
       "      <td>-</td>\n",
       "      <td>394</td>\n",
       "      <td>Q_A_I_aedA_RS26385</td>\n",
       "      <td>-</td>\n",
       "      <td>398</td>\n",
       "      <td>7.700000e-54</td>\n",
       "      <td>196.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2</td>\n",
       "      <td>7</td>\n",
       "      <td>394</td>\n",
       "      <td>5</td>\n",
       "      <td>389</td>\n",
       "      <td>2</td>\n",
       "      <td>393</td>\n",
       "      <td>0.94</td>\n",
       "      <td>-</td>\n",
       "      <td>0.977157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>3300025186.a:Ga0208056_100934_7</td>\n",
       "      <td>-</td>\n",
       "      <td>410</td>\n",
       "      <td>Q_A_I_aedA_RS26385</td>\n",
       "      <td>-</td>\n",
       "      <td>398</td>\n",
       "      <td>8.400000e-54</td>\n",
       "      <td>195.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "      <td>394</td>\n",
       "      <td>8</td>\n",
       "      <td>407</td>\n",
       "      <td>5</td>\n",
       "      <td>409</td>\n",
       "      <td>0.90</td>\n",
       "      <td>-</td>\n",
       "      <td>0.975610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>3300014912.a:Ga0121502_100060_10</td>\n",
       "      <td>-</td>\n",
       "      <td>402</td>\n",
       "      <td>Q_A_I_aedA_RS26385</td>\n",
       "      <td>-</td>\n",
       "      <td>398</td>\n",
       "      <td>9.700000e-54</td>\n",
       "      <td>195.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "      <td>397</td>\n",
       "      <td>9</td>\n",
       "      <td>400</td>\n",
       "      <td>6</td>\n",
       "      <td>401</td>\n",
       "      <td>0.93</td>\n",
       "      <td>-</td>\n",
       "      <td>0.975124</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25076 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            target_name accession tlen          query_name  \\\n",
       "0     3300009702.a:Ga0114931_10050376_3         -  376    Q_A_aedP_RS26440   \n",
       "1     3300009702.a:Ga0114931_10022461_4         -  376    Q_A_aedP_RS26440   \n",
       "2    3300027968.a:Ga0209061_1000039_348         -  378    Q_A_aedP_RS26440   \n",
       "3      3300027662.a:Ga0208565_1003173_4         -  380    Q_A_aedP_RS26440   \n",
       "4    3300017444.a:Ga0185300_10000551_65         -  375    Q_A_aedP_RS26440   \n",
       "..                                  ...       ...  ...                 ...   \n",
       "239    3300014092.a:Ga0121367_100110_13         -  405  Q_A_I_aedA_RS26385   \n",
       "240   3300027968.a:Ga0209061_1000140_88         -  412  Q_A_I_aedA_RS26385   \n",
       "241   3300017532.a:Ga0185302_10010386_3         -  394  Q_A_I_aedA_RS26385   \n",
       "242     3300025186.a:Ga0208056_100934_7         -  410  Q_A_I_aedA_RS26385   \n",
       "243    3300014912.a:Ga0121502_100060_10         -  402  Q_A_I_aedA_RS26385   \n",
       "\n",
       "    accession2 qlen        E-value  score  bias num_domains_index  ... bias2  \\\n",
       "0            -  385  6.300000e-116  400.0   2.4                 1  ...   2.4   \n",
       "1            -  385  7.400000e-116  399.8   5.1                 1  ...   5.1   \n",
       "2            -  385  2.300000e-112  388.3   3.1                 1  ...   3.1   \n",
       "3            -  385  1.300000e-111  385.8   3.6                 1  ...   3.6   \n",
       "4            -  385  3.500000e-111  384.4   3.0                 1  ...   3.0   \n",
       "..         ...  ...            ...    ...   ...               ...  ...   ...   \n",
       "239          -  398   6.200000e-54  196.3   0.0                 1  ...   0.0   \n",
       "240          -  398   6.800000e-54  196.1   0.0                 1  ...   0.0   \n",
       "241          -  398   7.700000e-54  196.0   0.2                 1  ...   0.2   \n",
       "242          -  398   8.400000e-54  195.8   0.0                 1  ...   0.0   \n",
       "243          -  398   9.700000e-54  195.6   0.0                 1  ...   0.0   \n",
       "\n",
       "     hmm_from  hmm_to  ali_from  ali_to env_from env_to   acc description  \\\n",
       "0           1     372         1     374        1    375  0.99           -   \n",
       "1           1     372         1     374        1    375  0.99           -   \n",
       "2           1     372         2     375        2    377  0.98           -   \n",
       "3           1     375         2     378        2    379  0.99           -   \n",
       "4           1     372         1     374        1    374  0.99           -   \n",
       "..        ...     ...       ...     ...      ...    ...   ...         ...   \n",
       "239         6     395         8     398        3    401  0.93           -   \n",
       "240        15     396        17     405        4    407  0.89           -   \n",
       "241         7     394         5     389        2    393  0.94           -   \n",
       "242         4     394         8     407        5    409  0.90           -   \n",
       "243         4     397         9     400        6    401  0.93           -   \n",
       "\n",
       "     coverage  \n",
       "0    0.994681  \n",
       "1    0.994681  \n",
       "2    0.989418  \n",
       "3    0.992105  \n",
       "4    0.997333  \n",
       "..        ...  \n",
       "239  0.965432  \n",
       "240  0.944175  \n",
       "241  0.977157  \n",
       "242  0.975610  \n",
       "243  0.975124  \n",
       "\n",
       "[25076 rows x 24 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Actino_HMM_MAGs\n",
    "# 需先進行前步驟的cell (1. 使用 negative control 來獲得合適的bit score)\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "#-------------------------------\n",
    "# Aed Cluster to MAGs\n",
    "\n",
    "# Define the directory that contains the \"domtblout\" files.需要刪除discription\n",
    "domtblout_dir = \"../data/raw/Actino_HMM_MAGs_domtblout/\"\n",
    "\n",
    "# Create an empty dictionary to store the target name for each hmmsearch\n",
    "MAGs_Hits = {}\n",
    "MAGs_Hits_name = []\n",
    "\n",
    "# covert criteria dataframe to serires\n",
    "Actino_BitScore_Criteria_S = Actino_BitScore_Criteria['Criteria_Bitscore']\n",
    "Actino_BitScore_Criteria_S = Actino_BitScore_Criteria_S.astype(float)    \n",
    "\n",
    "# create a dataframe for a all hits \n",
    "columns = [\"target_name\", \"accession\", \"tlen\", \"query_name\", \"accession2\", \"qlen\", \"E-value\", \"score\", \"bias\",\n",
    "           \"num_domains_index\", \"num_domains_total\", \"c-Evalue\", \"i-Evalue\", \"score2\", \"bias2\", \"hmm_from\",\n",
    "           \"hmm_to\", \"ali_from\", \"ali_to\", \"env_from\", \"env_to\", \"acc\", \"description\"]\n",
    "\n",
    "All_aed_Hits_df = pd.DataFrame(columns=columns)\n",
    "\n",
    "# Loop over the HMM DOMTBLOUT files and filter the results based on bit score, e-value and coverage\n",
    "# hmm name and bit score are in the Actino_BitScore_Criteria series\n",
    "for hmmsearch, threshold in Actino_BitScore_Criteria_S.items():\n",
    "    # Load the \"domtblout\" file into a pandas DataFrame\n",
    "    file_path = os.path.join(domtblout_dir, hmmsearch + \".domtblout\")\n",
    "    df = pd.read_csv(file_path, comment=\"#\", sep='\\s+', header=None)    \n",
    "    # Assign column names to the DataFrame\n",
    "    df.columns = [\"target_name\", \"accession\", \"tlen\", \"query_name\", \"accession2\", \"qlen\", \"E-value\", \"score\", \"bias\",\n",
    "                  \"num_domains_index\", \"num_domains_total\", \"c-Evalue\", \"i-Evalue\", \"score2\", \"bias2\", \"hmm_from\", \"hmm_to\", \"ali_from\", \"ali_to\",\n",
    "                  \"env_from\", \"env_to\", \"acc\", \"description\"]\n",
    "\n",
    "    # Calculate the coverage for each hit\n",
    "    df[\"coverage\"] = (df[\"ali_to\"] - df[\"ali_from\"] + 1) / df[\"tlen\"]\n",
    "    \n",
    "    # Filter the DataFrame by E-value, coverage, and bit-score\n",
    "    significant_hits = df[(df[\"E-value\"] < 0.001) & (df[\"coverage\"] > 0.50) & (df[\"score\"] > threshold)]\n",
    "\n",
    "    # Extract Target nmae and store it in the dictionary\n",
    "    if not significant_hits.empty:\n",
    "        MAGs_Hits_name = significant_hits[\"target_name\"].tolist()\n",
    "        MAGs_Hits[hmmsearch] = MAGs_Hits_name\n",
    "    else:\n",
    "        MAGs_Hits_name = None\n",
    "        MAGs_Hits[hmmsearch] = MAGs_Hits_name\n",
    "    \n",
    "    # add hits table to a df\n",
    "    All_aed_Hits_df = pd.concat([significant_hits, All_aed_Hits_df], axis=0)\n",
    "\n",
    "# done\n",
    "print('done')\n",
    "print('unique query name/numbers: ', All_aed_Hits_df['query_name'].unique(), ' / ', len(All_aed_Hits_df['query_name'].unique()))\n",
    "All_aed_Hits_df.to_csv('../data/processed/All_aed_Hits_df_bitscore.csv')    \n",
    "All_aed_Hits_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da71ca5b-906b-4047-bba5-792636f0eef8",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 2.1 獲取positive的target name並連結MAGsID，整理出一個類似heatmap的table，再以寬鬆hits數量進行篩選(>=8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "48ae4ccf-73eb-42ec-9f8c-30839b3d2fe4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Any Null in TargetToMAGsID:  False\n",
      "done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>query_name</th>\n",
       "      <th>genome_id</th>\n",
       "      <th>num_hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3300024344_6</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3300017444_33</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3300027902_34</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3300027902_36</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3300027649_17</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>3300020193_29</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>3300010158_16</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>3300027827_15</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>3300017452_8</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>3300020369_2</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>167 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "query_name      genome_id  num_hits\n",
       "0            3300024344_6        13\n",
       "1           3300017444_33        12\n",
       "2           3300027902_34        12\n",
       "3           3300027902_36        12\n",
       "4           3300027649_17        12\n",
       "..                    ...       ...\n",
       "162         3300020193_29         8\n",
       "163         3300010158_16         8\n",
       "164         3300027827_15         8\n",
       "165          3300017452_8         8\n",
       "166          3300020369_2         8\n",
       "\n",
       "[167 rows x 2 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Got target name and query name\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "# read table\n",
    "All_aed_Hits_df = pd.read_csv('../data/processed/All_aed_Hits_df_bitscore.csv') \n",
    "\n",
    "# Got target name and query name    \n",
    "All_aed_Hits_TargetAndQuery = All_aed_Hits_df[['target_name', 'query_name']]\n",
    "\n",
    "# Load a Dataframe with the lookup values for merge protein id to MAGs and merge them\n",
    "TarToMAGs_aed = pd.read_csv('../data/interim/Actino_aed/aed_All_TarToMAGsID.csv')\n",
    "\n",
    "# use merge() function to join the MAGsID data\n",
    "aed_hits_TargetAndMAGsID = pd.merge(All_aed_Hits_TargetAndQuery, TarToMAGs_aed, on='target_name', how='left')                                                                               \n",
    "\n",
    "# check the null value\n",
    "print('Any Null in TargetToMAGsID: ', aed_hits_TargetAndMAGsID['MAGsID'].isnull().any())\n",
    "\n",
    "# create the crosstab table (like heatmap)\n",
    "aed_hits_heatmap = pd.crosstab(aed_hits_TargetAndMAGsID['query_name'], aed_hits_TargetAndMAGsID['MAGsID'], dropna=False)\n",
    "aed_hits_heatmap = aed_hits_heatmap.transpose()\n",
    "# aed_hits_heatmap\n",
    "# aed_hits_heatmap.to_csv('../data/processed/aed_hits_heatmap_bitscore.csv')\n",
    "\n",
    "# Count the non-zero values in hmm profiles hit row to calculate the number of different HMM profiles that have hits in a given MAG. \n",
    "def count_nonzero(row):\n",
    "    return len(row[row != 0])\n",
    "\n",
    "num_hits = aed_hits_heatmap.apply(count_nonzero, axis=1)\n",
    "aed_hits_heatmap['num_hits'] = num_hits\n",
    "\n",
    "# sort them by hits numer\n",
    "aed_hits_heatmap = aed_hits_heatmap.sort_values(by=\"num_hits\", ascending=False)\n",
    "\n",
    "# extract > 8 hmm profiles hits and the necessary hits (aedA、aedB、aedJ)\n",
    "aed_hits_FinalFilter =  aed_hits_heatmap[(aed_hits_heatmap['num_hits'] >= 8)]\n",
    "# >7 = 326, >8 = 167, >9 = 70, >10 = 17 用8較為恰當 大於一半的query gene\n",
    "\n",
    "# Reset index and move index column to first position\n",
    "aed_hits_FinalFilter.index.name = None\n",
    "aed_hits_FinalFilter = aed_hits_FinalFilter.reset_index()\n",
    "aed_hits_FinalFilter.insert(0, 'index', aed_hits_FinalFilter.pop('index'))\n",
    "\n",
    "# rename MAGsID\n",
    "aed_hits_FinalFilter = aed_hits_FinalFilter.rename(columns={'index': 'genome_id'})\n",
    "\n",
    "# extract the MAGsID and num_hits column\n",
    "aed_Positive_MAGsID = aed_hits_FinalFilter[['genome_id', 'num_hits']]\n",
    "\n",
    "# done\n",
    "print('done')\n",
    "aed_Positive_MAGsID"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d9cdfa1-b439-4070-a30d-76be1893c8af",
   "metadata": {},
   "source": [
    "##### 2.1.1 連結positive MAGs到 metagenomes 總資料中並進行處理成最株能給R繪製 stacket barchart 的資料格式"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "da1c1a6f-a94d-4853-9ea2-47f0711b39db",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['p__Myxococcota' 'p__Actinobacteriota' 'p__UBP10' 'p__Proteobacteria'\n",
      " 'p__Chloroflexota']\n",
      "Any Null in aed_Positive_metagenomes_df:\n",
      " genome_id             False\n",
      "metagenome_id         False\n",
      "num_hits              False\n",
      "Phylum                False\n",
      "Class                 False\n",
      "ecosystem             False\n",
      "ecosystem_category    False\n",
      "longitude              True\n",
      "latitude               True\n",
      "Homologous_cluster    False\n",
      "dtype: bool\n",
      "Number of p__UBP10:  7\n",
      "done\n"
     ]
    }
   ],
   "source": [
    "# 需先執行上一個cell\n",
    "# open metagenome csv files\n",
    "metagenmoes_df = pd.read_csv('../data/external/genome_metadata_editForAnalysis_NotReference.csv')\n",
    "\n",
    "# merge positive MAGs with metagenome\n",
    "aed_Positive_metagenomes_df = pd.merge(aed_Positive_MAGsID, metagenmoes_df, on='genome_id', how='left')\n",
    "\n",
    "# extract certain column\n",
    "aed_Positive_metagenomes_df = aed_Positive_metagenomes_df[['genome_id', 'metagenome_id', 'taxonomy', 'ecosystem', 'ecosystem_category', 'num_hits', 'longitude', 'latitude']]\n",
    "\n",
    "# extract phylum and class from taxonomy column and expand to new column\n",
    "aed_Positive_metagenomes_df['Phylum'] = aed_Positive_metagenomes_df['taxonomy'].str.extract('(p__\\w+)', expand=True)\n",
    "aed_Positive_metagenomes_df['Class'] = aed_Positive_metagenomes_df['taxonomy'].str.extract('(c__\\w+)', expand=True)\n",
    "\n",
    "# # print non-duplicated values in ecosystem_type than check the lable\n",
    "# print(aed_Positive_metagenomes_df['ecosystem_type'].unique())\n",
    "# aed_Positive_metagenomes_df\n",
    "\n",
    "# # check the None value in ecosystem column Create a Boolean mask to identify NaN values\n",
    "# mask = aed_addEco_df.isna()\n",
    "# aed_addEco_df_nan_rows = aed_addEco_df[mask.any(axis=1)]\n",
    "# aed_addEco_df_nan_rows\n",
    "\n",
    "# remove, add and rearrange column\n",
    "aed_Positive_metagenomes_df.drop('taxonomy', axis=1, inplace=True)\n",
    "aed_Positive_metagenomes_df = aed_Positive_metagenomes_df.reindex(columns=['genome_id', 'metagenome_id', 'num_hits', 'Phylum', 'Class', 'ecosystem', 'ecosystem_category', 'longitude', 'latitude'])\n",
    "aed_Positive_metagenomes_df['Homologous_cluster']='Actino_aed_cluster'\n",
    "\n",
    "# check phylum data\n",
    "print(aed_Positive_metagenomes_df['Phylum'].unique())\n",
    "print('Any Null in aed_Positive_metagenomes_df:\\n', aed_Positive_metagenomes_df.isnull().any())\n",
    "\n",
    "# write file\n",
    "aed_Positive_metagenomes_df.to_csv('../data/processed/Final/Actino/aed_PositiveHits_ForR_loose.csv')\n",
    "print('Number of p__UBP10: ', aed_Positive_metagenomes_df[aed_Positive_metagenomes_df['Phylum'] == 'p__UBP10'].shape[0])\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "492cc492-4862-46ce-967c-d89228d3961f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### 不執行:2.2 獲取positive的target name並連結MAGsID，整理出一個類似heatmap的table，再以嚴格hits數量進行篩選(>=8, aedA, aedB, aedJ)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "cea63f54-999b-49bc-8136-fc1f41898589",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Any Null in TargetToMAGsID:  False\n",
      "done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>query_name</th>\n",
       "      <th>genome_id</th>\n",
       "      <th>num_hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3300024344_6</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3300017444_33</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3300027902_34</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3300025886_13</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "query_name      genome_id  num_hits\n",
       "0            3300024344_6        13\n",
       "1           3300017444_33        12\n",
       "2           3300027902_34        12\n",
       "3           3300025886_13        11"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Got target name and query name\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "# read table\n",
    "All_aed_Hits_df = pd.read_csv('../data/processed/All_aed_Hits_df_bitscore.csv') \n",
    "\n",
    "# Got target name and query name    \n",
    "All_aed_Hits_TargetAndQuery = All_aed_Hits_df[['target_name', 'query_name']]\n",
    "\n",
    "# Load a Dataframe with the lookup values for merge protein id to MAGs and merge them\n",
    "TarToMAGs_aed = pd.read_csv('../data/interim/Actino_aed/aed_All_TarToMAGsID.csv')\n",
    "\n",
    "# use merge() function to join the MAGsID data\n",
    "aed_hits_TargetAndMAGsID = pd.merge(All_aed_Hits_TargetAndQuery, TarToMAGs_aed, on='target_name', how='left')                                                                               \n",
    "\n",
    "# check the null value\n",
    "print('Any Null in TargetToMAGsID: ', aed_hits_TargetAndMAGsID['MAGsID'].isnull().any())\n",
    "\n",
    "# create the crosstab table (like heatmap)\n",
    "aed_hits_heatmap = pd.crosstab(aed_hits_TargetAndMAGsID['query_name'], aed_hits_TargetAndMAGsID['MAGsID'], dropna=False)\n",
    "aed_hits_heatmap = aed_hits_heatmap.transpose()\n",
    "# aed_hits_heatmap\n",
    "# aed_hits_heatmap.to_csv('../data/processed/aed_hits_heatmap_bitscore.csv')\n",
    "\n",
    "# Count the non-zero values in hmm profiles hit row to calculate the number of different HMM profiles that have hits in a given MAG. \n",
    "def count_nonzero(row):\n",
    "    return len(row[row != 0])\n",
    "\n",
    "num_hits = aed_hits_heatmap.apply(count_nonzero, axis=1)\n",
    "aed_hits_heatmap['num_hits'] = num_hits\n",
    "\n",
    "# sort them by hits numer\n",
    "aed_hits_heatmap = aed_hits_heatmap.sort_values(by=\"num_hits\", ascending=False)\n",
    "\n",
    "# extract > 8 hmm profiles hits and the necessary hits (aedA、aedB、aedJ)\n",
    "aed_hits_FinalFilter =  aed_hits_heatmap[(aed_hits_heatmap['num_hits'] >= 8) & (aed_hits_heatmap['Q_A_I_aedA_RS26385'] != 0) & (aed_hits_heatmap['Q_A_I_aedB_RS26395'] != 0) & (aed_hits_heatmap['Q_A_I_aedJ_RS26410'] != 0)]\n",
    "# 4hits; 3300024344_6, 3300017444_33, 3300027902_34,3300025886_13\n",
    "\n",
    "# Reset index and move index column to first position\n",
    "aed_hits_FinalFilter.index.name = None\n",
    "aed_hits_FinalFilter = aed_hits_FinalFilter.reset_index()\n",
    "aed_hits_FinalFilter.insert(0, 'index', aed_hits_FinalFilter.pop('index'))\n",
    "\n",
    "# rename MAGsID\n",
    "aed_hits_FinalFilter = aed_hits_FinalFilter.rename(columns={'index': 'genome_id'})\n",
    "\n",
    "# extract the MAGsID and num_hits column\n",
    "aed_Positive_MAGsID = aed_hits_FinalFilter[['genome_id', 'num_hits']]\n",
    "\n",
    "# done\n",
    "print('done')\n",
    "aed_Positive_MAGsID"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ffc860f-cd72-4456-85f6-f930a4f99719",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### 2.2.1 連結critical_positive MAGs到 metagenomes 總資料中並進行處理成最株能給R繪製 stacket barchart 的資料格式"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "a454805c-f4e9-44d1-b7fc-66cbb11bdbc8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['p__Myxococcota' 'p__Actinobacteriota']\n",
      "Any Null in aed_Positive_metagenomes_df:\n",
      " genome_id             False\n",
      "metagenome_id         False\n",
      "num_hits              False\n",
      "Phylum                False\n",
      "Class                 False\n",
      "ecosystem             False\n",
      "ecosystem_category    False\n",
      "longitude             False\n",
      "latitude              False\n",
      "Homologous_cluster    False\n",
      "dtype: bool\n",
      "done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>metagenome_id</th>\n",
       "      <th>num_hits</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>ecosystem</th>\n",
       "      <th>ecosystem_category</th>\n",
       "      <th>longitude</th>\n",
       "      <th>latitude</th>\n",
       "      <th>Homologous_cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3300024344_6</td>\n",
       "      <td>3300024344</td>\n",
       "      <td>13</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__UBA9160</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>25.4868</td>\n",
       "      <td>36.5264</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3300017444_33</td>\n",
       "      <td>3300017444</td>\n",
       "      <td>12</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__UBA9160</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Thermal springs</td>\n",
       "      <td>-125.5130</td>\n",
       "      <td>60.1987</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3300027902_34</td>\n",
       "      <td>3300027902</td>\n",
       "      <td>12</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>-86.2300</td>\n",
       "      <td>41.7000</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3300025886_13</td>\n",
       "      <td>3300025886</td>\n",
       "      <td>11</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__UBA9160</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>7.9000</td>\n",
       "      <td>54.1842</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       genome_id  metagenome_id  num_hits               Phylum  \\\n",
       "0   3300024344_6     3300024344        13       p__Myxococcota   \n",
       "1  3300017444_33     3300017444        12       p__Myxococcota   \n",
       "2  3300027902_34     3300027902        12  p__Actinobacteriota   \n",
       "3  3300025886_13     3300025886        11       p__Myxococcota   \n",
       "\n",
       "               Class ecosystem ecosystem_category  longitude  latitude  \\\n",
       "0         c__UBA9160   Aquatic             Marine    25.4868   36.5264   \n",
       "1         c__UBA9160   Aquatic    Thermal springs  -125.5130   60.1987   \n",
       "2  c__Acidimicrobiia   Aquatic         Freshwater   -86.2300   41.7000   \n",
       "3         c__UBA9160   Aquatic             Marine     7.9000   54.1842   \n",
       "\n",
       "   Homologous_cluster  \n",
       "0  Actino_aed_cluster  \n",
       "1  Actino_aed_cluster  \n",
       "2  Actino_aed_cluster  \n",
       "3  Actino_aed_cluster  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 需先執行上一個cell\n",
    "# open metagenome csv files\n",
    "metagenmoes_df = pd.read_csv('../data/external/genome_metadata_editForAnalysis_NotReference.csv')\n",
    "\n",
    "# merge positive MAGs with metagenome\n",
    "aed_Positive_metagenomes_df = pd.merge(aed_Positive_MAGsID, metagenmoes_df, on='genome_id', how='left')\n",
    "\n",
    "# extract certain column\n",
    "aed_Positive_metagenomes_df = aed_Positive_metagenomes_df[['genome_id', 'metagenome_id', 'taxonomy', 'ecosystem', 'ecosystem_category', 'num_hits', 'longitude', 'latitude']]\n",
    "\n",
    "# extract phylum and class from taxonomy column and expand to new column\n",
    "aed_Positive_metagenomes_df['Phylum'] = aed_Positive_metagenomes_df['taxonomy'].str.extract('(p__\\w+)', expand=True)\n",
    "aed_Positive_metagenomes_df['Class'] = aed_Positive_metagenomes_df['taxonomy'].str.extract('(c__\\w+)', expand=True)\n",
    "\n",
    "# # print non-duplicated values in ecosystem_type than check the lable\n",
    "# print(aed_Positive_metagenomes_df['ecosystem_type'].unique())\n",
    "# aed_Positive_metagenomes_df\n",
    "\n",
    "# # check the None value in ecosystem column Create a Boolean mask to identify NaN values\n",
    "# mask = aed_addEco_df.isna()\n",
    "# aed_addEco_df_nan_rows = aed_addEco_df[mask.any(axis=1)]\n",
    "# aed_addEco_df_nan_rows\n",
    "\n",
    "# remove, add and rearrange column\n",
    "aed_Positive_metagenomes_df.drop('taxonomy', axis=1, inplace=True)\n",
    "aed_Positive_metagenomes_df= aed_Positive_metagenomes_df.reindex(columns=['genome_id', 'metagenome_id', 'num_hits', 'Phylum', 'Class', 'ecosystem', 'ecosystem_category', 'longitude', 'latitude'])\n",
    "aed_Positive_metagenomes_df['Homologous_cluster']='Actino_aed_cluster'\n",
    "\n",
    "# check phylum data\n",
    "print(aed_Positive_metagenomes_df['Phylum'].unique())\n",
    "print('Any Null in aed_Positive_metagenomes_df:\\n', aed_Positive_metagenomes_df.isnull().any())\n",
    "\n",
    "# write file\n",
    "aed_Positive_metagenomes_df.to_csv('../data/processed/Final/Actino/aed_PositiveHits_ForR_critical.csv')\n",
    "print('done')\n",
    "aed_Positive_metagenomes_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43d0d546-6eeb-483e-afe9-f108a6a5e675",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# 不執行"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a44cb69a-b3ba-4a2b-9608-a3fffab5a229",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 2.3 獲取positive的target name並連結MAGsID，整理出一個類似heatmap的table，再以本先討論的hits數量進行篩選來看差異(>=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b9d7734f-dfec-496a-9905-93180fa76c19",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Any Null:  False\n",
      "done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>query_name</th>\n",
       "      <th>genome_id</th>\n",
       "      <th>num_hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3300024344_6</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3300027902_36</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3300017444_33</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3300027902_34</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3300027649_17</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3300005529_56</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3300007722_64</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3300027627_9</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3300025886_13</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3300027983_50</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>3300010396_41</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>3300015202_20</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>3300029901_14</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>3300001683_15</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>3300009540_25</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>3300027834_19</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>3300020193_17</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "query_name      genome_id  num_hits\n",
       "0            3300024344_6        13\n",
       "1           3300027902_36        12\n",
       "2           3300017444_33        12\n",
       "3           3300027902_34        12\n",
       "4           3300027649_17        12\n",
       "5           3300005529_56        11\n",
       "6           3300007722_64        11\n",
       "7            3300027627_9        11\n",
       "8           3300025886_13        11\n",
       "9           3300027983_50        11\n",
       "10          3300010396_41        10\n",
       "11          3300015202_20        10\n",
       "12          3300029901_14        10\n",
       "13          3300001683_15        10\n",
       "14          3300009540_25        10\n",
       "15          3300027834_19        10\n",
       "16          3300020193_17        10"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Got target name and query name\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "# read table\n",
    "All_aed_Hits_df = pd.read_csv('../data/processed/All_aed_Hits_df_bitscore.csv') \n",
    "\n",
    "# Got target name and query name    \n",
    "All_aed_Hits_TargetAndQuery = All_aed_Hits_df[['target_name', 'query_name']]\n",
    "\n",
    "# Load a Dataframe with the lookup values for merge protein id to MAGs and merge them\n",
    "TarToMAGs_aed = pd.read_csv('../data/interim/Actino_aed/aed_All_TarToMAGsID.csv')\n",
    "\n",
    "# use merge() function to join the MAGsID data\n",
    "aed_hits_TargetAndMAGsID = pd.merge(All_aed_Hits_TargetAndQuery, TarToMAGs_aed, on='target_name', how='left')                                                                               \n",
    "\n",
    "# check the null value\n",
    "print('Any Null: ', aed_hits_TargetAndMAGsID['MAGsID'].isnull().any())\n",
    "\n",
    "# create the crosstab table (like heatmap)\n",
    "aed_hits_heatmap = pd.crosstab(aed_hits_TargetAndMAGsID['query_name'], aed_hits_TargetAndMAGsID['MAGsID'], dropna=False)\n",
    "aed_hits_heatmap = aed_hits_heatmap.transpose()\n",
    "# aed_hits_heatmap\n",
    "# aed_hits_heatmap.to_csv('../data/processed/aed_hits_heatmap_bitscore.csv')\n",
    "\n",
    "# Count the non-zero values in hmm profiles hit row to calculate the number of different HMM profiles that have hits in a given MAG. \n",
    "def count_nonzero(row):\n",
    "    return len(row[row != 0])\n",
    "\n",
    "num_hits = aed_hits_heatmap.apply(count_nonzero, axis=1)\n",
    "aed_hits_heatmap['num_hits'] = num_hits\n",
    "\n",
    "# sort them by hits numer\n",
    "aed_hits_heatmap = aed_hits_heatmap.sort_values(by=\"num_hits\", ascending=False)\n",
    "\n",
    "# extract > 10 hmm profiles hits and the necessary hits\n",
    "aed_hits_FinalFilter =  aed_hits_heatmap[(aed_hits_heatmap['num_hits'] >= 10)]\n",
    "# 17個hits\n",
    "\n",
    "# Reset index and move index column to first position\n",
    "aed_hits_FinalFilter.index.name = None\n",
    "aed_hits_FinalFilter = aed_hits_FinalFilter.reset_index()\n",
    "aed_hits_FinalFilter.insert(0, 'index', aed_hits_FinalFilter.pop('index'))\n",
    "\n",
    "# rename MAGsID\n",
    "aed_hits_FinalFilter = aed_hits_FinalFilter.rename(columns={'index': 'genome_id'})\n",
    "\n",
    "# extract the MAGsID and num_hits column\n",
    "aed_Positive_MAGsID = aed_hits_FinalFilter[['genome_id', 'num_hits']]\n",
    "\n",
    "# done\n",
    "print('done')\n",
    "aed_Positive_MAGsID"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81e4791c-1b99-400f-8b88-f3b715247dda",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### 2.3.1 連結 10 Hits positive (>10) MAGs到 metagenomes 總資料中並進行處理成最株能給R繪製 stacket barchart 的資料格式"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "ff4e4677-845b-451d-977f-f5eb1fd9f4e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Marine' 'Freshwater' 'Thermal spring' 'Soil' 'Wastewater']\n",
      "done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>num_hits</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>ecosystem</th>\n",
       "      <th>ecosystem_type</th>\n",
       "      <th>Homologous_cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3300024344_6</td>\n",
       "      <td>13</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__UBA9160</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3300027902_36</td>\n",
       "      <td>12</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3300017444_33</td>\n",
       "      <td>12</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__UBA9160</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Thermal spring</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3300027902_34</td>\n",
       "      <td>12</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3300027649_17</td>\n",
       "      <td>12</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3300005529_56</td>\n",
       "      <td>11</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3300007722_64</td>\n",
       "      <td>11</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3300027627_9</td>\n",
       "      <td>11</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3300025886_13</td>\n",
       "      <td>11</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__UBA9160</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3300027983_50</td>\n",
       "      <td>11</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>3300010396_41</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>3300015202_20</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>3300029901_14</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>3300001683_15</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__UBA9160</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>3300009540_25</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Engineered</td>\n",
       "      <td>Wastewater</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>3300027834_19</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>3300020193_17</td>\n",
       "      <td>10</td>\n",
       "      <td>p__Myxococcota</td>\n",
       "      <td>c__UBA9160</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        genome_id  num_hits               Phylum              Class  \\\n",
       "0    3300024344_6        13       p__Myxococcota         c__UBA9160   \n",
       "1   3300027902_36        12  p__Actinobacteriota  c__Acidimicrobiia   \n",
       "2   3300017444_33        12       p__Myxococcota         c__UBA9160   \n",
       "3   3300027902_34        12  p__Actinobacteriota  c__Acidimicrobiia   \n",
       "4   3300027649_17        12  p__Actinobacteriota  c__Acidimicrobiia   \n",
       "5   3300005529_56        11  p__Actinobacteriota  c__Acidimicrobiia   \n",
       "6   3300007722_64        11  p__Actinobacteriota  c__Acidimicrobiia   \n",
       "7    3300027627_9        11  p__Actinobacteriota  c__Acidimicrobiia   \n",
       "8   3300025886_13        11       p__Myxococcota         c__UBA9160   \n",
       "9   3300027983_50        11  p__Actinobacteriota  c__Acidimicrobiia   \n",
       "10  3300010396_41        10  p__Actinobacteriota  c__Acidimicrobiia   \n",
       "11  3300015202_20        10  p__Actinobacteriota  c__Acidimicrobiia   \n",
       "12  3300029901_14        10  p__Actinobacteriota  c__Acidimicrobiia   \n",
       "13  3300001683_15        10       p__Myxococcota         c__UBA9160   \n",
       "14  3300009540_25        10  p__Actinobacteriota  c__Acidimicrobiia   \n",
       "15  3300027834_19        10  p__Actinobacteriota  c__Acidimicrobiia   \n",
       "16  3300020193_17        10       p__Myxococcota         c__UBA9160   \n",
       "\n",
       "      ecosystem  ecosystem_type  Homologous_cluster  \n",
       "0       Aquatic          Marine  Actino_aed_cluster  \n",
       "1       Aquatic      Freshwater  Actino_aed_cluster  \n",
       "2       Aquatic  Thermal spring  Actino_aed_cluster  \n",
       "3       Aquatic      Freshwater  Actino_aed_cluster  \n",
       "4       Aquatic      Freshwater  Actino_aed_cluster  \n",
       "5   Terrestrial            Soil  Actino_aed_cluster  \n",
       "6       Aquatic      Freshwater  Actino_aed_cluster  \n",
       "7       Aquatic      Freshwater  Actino_aed_cluster  \n",
       "8       Aquatic          Marine  Actino_aed_cluster  \n",
       "9       Aquatic      Freshwater  Actino_aed_cluster  \n",
       "10  Terrestrial            Soil  Actino_aed_cluster  \n",
       "11  Terrestrial            Soil  Actino_aed_cluster  \n",
       "12      Aquatic      Freshwater  Actino_aed_cluster  \n",
       "13      Aquatic          Marine  Actino_aed_cluster  \n",
       "14   Engineered      Wastewater  Actino_aed_cluster  \n",
       "15      Aquatic          Marine  Actino_aed_cluster  \n",
       "16      Aquatic      Freshwater  Actino_aed_cluster  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 需先執行上一個cell\n",
    "# open metagenome csv files\n",
    "metagenmoes_df = pd.read_csv('../data/external/Paper_genome_metadata.csv')\n",
    "metagenmoes_df = metagenmoes_df.rename(columns={'ecosystem': 'taxonomy'})\n",
    "\n",
    "# merge positive MAGs with metagenome\n",
    "aed_Positive_metagenomes_df = pd.merge(aed_Positive_MAGsID, metagenmoes_df, on='genome_id', how='left')\n",
    "\n",
    "# print(aed_Positive_metagenomes_df['ecosystem_category'].unique())\n",
    "# print(aed_Positive_metagenomes_df['ecosystem_type'].unique())\n",
    "# display(aed_Positive_metagenomes_df[aed_Positive_metagenomes_df['ecosystem_type'] == 'Anaerobic'])\n",
    "\n",
    "# extract certain column\n",
    "aed_Positive_metagenomes_df = aed_Positive_metagenomes_df[['genome_id', 'taxonomy', 'ecosystem_type', 'num_hits']]\n",
    "\n",
    "# extract phylum and class from taxonomy column and expand to new column\n",
    "aed_Positive_metagenomes_df['Phylum'] = aed_Positive_metagenomes_df['taxonomy'].str.extract('(p__\\w+)', expand=True)\n",
    "aed_Positive_metagenomes_df['Class'] = aed_Positive_metagenomes_df['taxonomy'].str.extract('(c__\\w+)', expand=True)\n",
    "\n",
    "# # print non-duplicated values in ecosystem_type than check the lable\n",
    "# print(aed_Positive_metagenomes_df['ecosystem_type'].unique())\n",
    "# aed_Positive_metagenomes_df\n",
    "\n",
    "# Load a EcosystemToEco_type dataframe for lookup\n",
    "ForEcoLookup = pd.read_csv(\"../data/interim/EcosystemToEco_type.csv\")\n",
    "\n",
    "# merge them with ecosystem_type\n",
    "aed_addEco_df = pd.merge(aed_Positive_metagenomes_df, ForEcoLookup, on='ecosystem_type', how='left')\n",
    "\n",
    "# # check the None value in ecosystem column Create a Boolean mask to identify NaN values\n",
    "# mask = aed_addEco_df.isna()\n",
    "# aed_addEco_df_nan_rows = aed_addEco_df[mask.any(axis=1)]\n",
    "# aed_addEco_df_nan_rows\n",
    "\n",
    "# !!!! change the ecosystem_type value name\n",
    "aed_addEco_df['ecosystem_type'] = aed_addEco_df['ecosystem_type'].replace({'Rhizoplane': 'Plant', 'Defined media': \"Lab enrichment\", 'Nutrient removal': 'Wastewater', 'Activated Sludge': 'Wastewater'})\n",
    "print(aed_addEco_df['ecosystem_type'].unique())\n",
    "\n",
    "# remove, add and rearrange column\n",
    "aed_addEco_df.drop('taxonomy', axis=1, inplace=True)\n",
    "aed_addEco_df= aed_addEco_df.reindex(columns=['genome_id', 'num_hits', 'Phylum', 'Class', 'ecosystem', 'ecosystem_type'])\n",
    "aed_addEco_df['Homologous_cluster']='Actino_aed_cluster'\n",
    "\n",
    "# # check phylum data\n",
    "# print(aed_addEco_df['Phylum'].unique())\n",
    "\n",
    "# write file\n",
    "aed_addEco_df.to_csv('../data/processed/Final/Actino/aed_PositiveHits_ForR_10Hits.csv')\n",
    "print('done')\n",
    "aed_addEco_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8dc26237-2c9d-4a39-bdbd-6a1cf7ccc480",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 第二部分: 舊版參考前人標準 (e-value coverage)，不含有bit score\n",
    "DOI: https://doi.org/10.1128/mBio.00166-16\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74981990-7cd7-43fc-81a8-08ef6fee0ab0",
   "metadata": {},
   "source": [
    "### 1. 用一致的e-value與coverage進行篩選"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "5bec272d-5c5d-4407-8a85-c466ba9b6dc3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_name</th>\n",
       "      <th>accession</th>\n",
       "      <th>tlen</th>\n",
       "      <th>query_name</th>\n",
       "      <th>accession2</th>\n",
       "      <th>qlen</th>\n",
       "      <th>E-value</th>\n",
       "      <th>score</th>\n",
       "      <th>bias</th>\n",
       "      <th>num_domains_index</th>\n",
       "      <th>...</th>\n",
       "      <th>bias2</th>\n",
       "      <th>hmm_from</th>\n",
       "      <th>hmm_to</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>ali_to</th>\n",
       "      <th>env_from</th>\n",
       "      <th>env_to</th>\n",
       "      <th>acc</th>\n",
       "      <th>description</th>\n",
       "      <th>coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3300024344.a:Ga0209992_10000019_32</td>\n",
       "      <td>-</td>\n",
       "      <td>396</td>\n",
       "      <td>Q_A_I_aedK_RS26415</td>\n",
       "      <td>-</td>\n",
       "      <td>389</td>\n",
       "      <td>2.900000e-179</td>\n",
       "      <td>608.4</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2</td>\n",
       "      <td>384</td>\n",
       "      <td>6</td>\n",
       "      <td>391</td>\n",
       "      <td>5</td>\n",
       "      <td>395</td>\n",
       "      <td>0.98</td>\n",
       "      <td>-</td>\n",
       "      <td>0.974747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3300009540.a:Ga0073899_10003596_14</td>\n",
       "      <td>-</td>\n",
       "      <td>362</td>\n",
       "      <td>Q_A_I_aedK_RS26415</td>\n",
       "      <td>-</td>\n",
       "      <td>389</td>\n",
       "      <td>5.000000e-166</td>\n",
       "      <td>564.9</td>\n",
       "      <td>6.8</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>6.8</td>\n",
       "      <td>26</td>\n",
       "      <td>386</td>\n",
       "      <td>1</td>\n",
       "      <td>361</td>\n",
       "      <td>1</td>\n",
       "      <td>361</td>\n",
       "      <td>1.00</td>\n",
       "      <td>-</td>\n",
       "      <td>0.997238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3300017449.a:Ga0185342_1000806_2</td>\n",
       "      <td>-</td>\n",
       "      <td>388</td>\n",
       "      <td>Q_A_I_aedK_RS26415</td>\n",
       "      <td>-</td>\n",
       "      <td>389</td>\n",
       "      <td>8.600000e-142</td>\n",
       "      <td>485.1</td>\n",
       "      <td>1.2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1.2</td>\n",
       "      <td>2</td>\n",
       "      <td>386</td>\n",
       "      <td>3</td>\n",
       "      <td>387</td>\n",
       "      <td>2</td>\n",
       "      <td>387</td>\n",
       "      <td>0.97</td>\n",
       "      <td>-</td>\n",
       "      <td>0.992268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3300017971.a:Ga0180438_10046985_3</td>\n",
       "      <td>-</td>\n",
       "      <td>388</td>\n",
       "      <td>Q_A_I_aedK_RS26415</td>\n",
       "      <td>-</td>\n",
       "      <td>389</td>\n",
       "      <td>2.100000e-141</td>\n",
       "      <td>483.8</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2</td>\n",
       "      <td>386</td>\n",
       "      <td>3</td>\n",
       "      <td>387</td>\n",
       "      <td>2</td>\n",
       "      <td>387</td>\n",
       "      <td>0.96</td>\n",
       "      <td>-</td>\n",
       "      <td>0.992268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3300007722.a:Ga0105051_10000035_61</td>\n",
       "      <td>-</td>\n",
       "      <td>387</td>\n",
       "      <td>Q_A_I_aedK_RS26415</td>\n",
       "      <td>-</td>\n",
       "      <td>389</td>\n",
       "      <td>4.800000e-141</td>\n",
       "      <td>482.6</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1.5</td>\n",
       "      <td>3</td>\n",
       "      <td>387</td>\n",
       "      <td>5</td>\n",
       "      <td>385</td>\n",
       "      <td>3</td>\n",
       "      <td>386</td>\n",
       "      <td>0.98</td>\n",
       "      <td>-</td>\n",
       "      <td>0.984496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4652</th>\n",
       "      <td>3300026288.a:Ga0209572_1000144_104</td>\n",
       "      <td>-</td>\n",
       "      <td>168</td>\n",
       "      <td>Q_A_aedC_RS26365</td>\n",
       "      <td>-</td>\n",
       "      <td>319</td>\n",
       "      <td>7.100000e-26</td>\n",
       "      <td>103.5</td>\n",
       "      <td>0.4</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>151</td>\n",
       "      <td>319</td>\n",
       "      <td>1</td>\n",
       "      <td>167</td>\n",
       "      <td>1</td>\n",
       "      <td>167</td>\n",
       "      <td>0.94</td>\n",
       "      <td>-</td>\n",
       "      <td>0.994048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4653</th>\n",
       "      <td>3300026303.a:Ga0209793_1000105_103</td>\n",
       "      <td>-</td>\n",
       "      <td>168</td>\n",
       "      <td>Q_A_aedC_RS26365</td>\n",
       "      <td>-</td>\n",
       "      <td>319</td>\n",
       "      <td>7.100000e-26</td>\n",
       "      <td>103.5</td>\n",
       "      <td>0.4</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>151</td>\n",
       "      <td>319</td>\n",
       "      <td>1</td>\n",
       "      <td>167</td>\n",
       "      <td>1</td>\n",
       "      <td>167</td>\n",
       "      <td>0.94</td>\n",
       "      <td>-</td>\n",
       "      <td>0.994048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4654</th>\n",
       "      <td>3300012973.a:Ga0123351_1000371_12</td>\n",
       "      <td>-</td>\n",
       "      <td>345</td>\n",
       "      <td>Q_A_aedC_RS26365</td>\n",
       "      <td>-</td>\n",
       "      <td>319</td>\n",
       "      <td>9.500000e-26</td>\n",
       "      <td>103.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51</td>\n",
       "      <td>298</td>\n",
       "      <td>54</td>\n",
       "      <td>322</td>\n",
       "      <td>31</td>\n",
       "      <td>338</td>\n",
       "      <td>0.82</td>\n",
       "      <td>-</td>\n",
       "      <td>0.779710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4655</th>\n",
       "      <td>3300012833.a:Ga0160489_100006_226</td>\n",
       "      <td>-</td>\n",
       "      <td>357</td>\n",
       "      <td>Q_A_aedC_RS26365</td>\n",
       "      <td>-</td>\n",
       "      <td>319</td>\n",
       "      <td>9.900000e-26</td>\n",
       "      <td>103.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2</td>\n",
       "      <td>37</td>\n",
       "      <td>306</td>\n",
       "      <td>34</td>\n",
       "      <td>337</td>\n",
       "      <td>6</td>\n",
       "      <td>350</td>\n",
       "      <td>0.73</td>\n",
       "      <td>-</td>\n",
       "      <td>0.851541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4656</th>\n",
       "      <td>3300012842.a:Ga0160481_100444_22</td>\n",
       "      <td>-</td>\n",
       "      <td>357</td>\n",
       "      <td>Q_A_aedC_RS26365</td>\n",
       "      <td>-</td>\n",
       "      <td>319</td>\n",
       "      <td>9.900000e-26</td>\n",
       "      <td>103.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2</td>\n",
       "      <td>37</td>\n",
       "      <td>306</td>\n",
       "      <td>34</td>\n",
       "      <td>337</td>\n",
       "      <td>6</td>\n",
       "      <td>350</td>\n",
       "      <td>0.73</td>\n",
       "      <td>-</td>\n",
       "      <td>0.851541</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>505251 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             target_name accession tlen          query_name  \\\n",
       "0     3300024344.a:Ga0209992_10000019_32         -  396  Q_A_I_aedK_RS26415   \n",
       "1     3300009540.a:Ga0073899_10003596_14         -  362  Q_A_I_aedK_RS26415   \n",
       "2       3300017449.a:Ga0185342_1000806_2         -  388  Q_A_I_aedK_RS26415   \n",
       "3      3300017971.a:Ga0180438_10046985_3         -  388  Q_A_I_aedK_RS26415   \n",
       "4     3300007722.a:Ga0105051_10000035_61         -  387  Q_A_I_aedK_RS26415   \n",
       "...                                  ...       ...  ...                 ...   \n",
       "4652  3300026288.a:Ga0209572_1000144_104         -  168    Q_A_aedC_RS26365   \n",
       "4653  3300026303.a:Ga0209793_1000105_103         -  168    Q_A_aedC_RS26365   \n",
       "4654   3300012973.a:Ga0123351_1000371_12         -  345    Q_A_aedC_RS26365   \n",
       "4655   3300012833.a:Ga0160489_100006_226         -  357    Q_A_aedC_RS26365   \n",
       "4656    3300012842.a:Ga0160481_100444_22         -  357    Q_A_aedC_RS26365   \n",
       "\n",
       "     accession2 qlen        E-value  score  bias num_domains_index  ... bias2  \\\n",
       "0             -  389  2.900000e-179  608.4   3.0                 1  ...   3.0   \n",
       "1             -  389  5.000000e-166  564.9   6.8                 1  ...   6.8   \n",
       "2             -  389  8.600000e-142  485.1   1.2                 1  ...   1.2   \n",
       "3             -  389  2.100000e-141  483.8   0.5                 1  ...   0.5   \n",
       "4             -  389  4.800000e-141  482.6   1.5                 1  ...   1.5   \n",
       "...         ...  ...            ...    ...   ...               ...  ...   ...   \n",
       "4652          -  319   7.100000e-26  103.5   0.4                 1  ...   0.4   \n",
       "4653          -  319   7.100000e-26  103.5   0.4                 1  ...   0.4   \n",
       "4654          -  319   9.500000e-26  103.1   0.0                 1  ...   0.0   \n",
       "4655          -  319   9.900000e-26  103.0   0.2                 1  ...   0.2   \n",
       "4656          -  319   9.900000e-26  103.0   0.2                 1  ...   0.2   \n",
       "\n",
       "      hmm_from  hmm_to  ali_from  ali_to env_from env_to   acc description  \\\n",
       "0            2     384         6     391        5    395  0.98           -   \n",
       "1           26     386         1     361        1    361  1.00           -   \n",
       "2            2     386         3     387        2    387  0.97           -   \n",
       "3            2     386         3     387        2    387  0.96           -   \n",
       "4            3     387         5     385        3    386  0.98           -   \n",
       "...        ...     ...       ...     ...      ...    ...   ...         ...   \n",
       "4652       151     319         1     167        1    167  0.94           -   \n",
       "4653       151     319         1     167        1    167  0.94           -   \n",
       "4654        51     298        54     322       31    338  0.82           -   \n",
       "4655        37     306        34     337        6    350  0.73           -   \n",
       "4656        37     306        34     337        6    350  0.73           -   \n",
       "\n",
       "      coverage  \n",
       "0     0.974747  \n",
       "1     0.997238  \n",
       "2     0.992268  \n",
       "3     0.992268  \n",
       "4     0.984496  \n",
       "...        ...  \n",
       "4652  0.994048  \n",
       "4653  0.994048  \n",
       "4654  0.779710  \n",
       "4655  0.851541  \n",
       "4656  0.851541  \n",
       "\n",
       "[505251 rows x 24 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Actino MAGs reference filtering\n",
    "# Define the directory that contains the \"domtblout\" files, 這個genome protein fasta要把discription的文字刪除，因有空格會影響read\n",
    "domtblout_dir = \"../data/raw/Actino_HMM_MAGs_domtblout/\"\n",
    "\n",
    "# create the list of hmmsearches output files name (without the file extension).\n",
    "hmmsearches = [\"MAGs_A_aedC_RS26365\", \"MAGs_A_aedD_RS26370\", \"MAGs_A_aedE_RS26375\", \"MAGs_A_aedI_RS26405\", \"MAGs_A_aedL_RS26420\", \"MAGs_A_aedM_RS26425\", \"MAGs_A_aedN_RS26430\",\n",
    "               \"MAGs_A_aedO_RS26435\", \"MAGs_A_aedP_RS26440\", \"MAGs_A_aedQ_RS26445\", \"MAGs_A_aedR_RS26450\", \"MAGs_A_I_aedA_RS26385\",\n",
    "               \"MAGs_A_I_aedB_RS26395\", \"MAGs_A_I_aedF_RS26380\", \"MAGs_A_I_aedG_RS26390\", \"MAGs_A_I_aedH_RS26400\", \"MAGs_A_I_aedJ_RS26410\", \"MAGs_A_I_aedK_RS26415\"]\n",
    "\n",
    "# create a empty dataframe for hits target \n",
    "columns = [\"target_name\", \"accession\", \"tlen\", \"query_name\", \"accession2\", \"qlen\", \"E-value\", \"score\", \"bias\",\n",
    "           \"num_domains_index\", \"num_domains_total\", \"c-Evalue\", \"i-Evalue\", \"score2\", \"bias2\", \"hmm_from\",\n",
    "           \"hmm_to\", \"ali_from\", \"ali_to\", \"env_from\", \"env_to\", \"acc\", \"description\"]\n",
    "\n",
    "MAGs_Hits = {}\n",
    "MAGs_Hits_name = []\n",
    "\n",
    "All_aed_Hits_df = pd.DataFrame(columns=columns)\n",
    "\n",
    "# Loop over the hmmsearches and parse the corresponding \"domtblout\" file\n",
    "for hmmsearch in hmmsearches:\n",
    "    # Load the \"domtblout\" file into a pandas DataFrame\n",
    "    file_path = os.path.join(domtblout_dir, hmmsearch + \".domtblout\")\n",
    "    df = pd.read_csv(file_path, comment=\"#\", sep='\\s+', header=None)    \n",
    "    # Assign column names to the DataFrame\n",
    "    df.columns = [\"target_name\", \"accession\", \"tlen\", \"query_name\", \"accession2\", \"qlen\", \"E-value\", \"score\", \"bias\",\n",
    "                  \"num_domains_index\", \"num_domains_total\", \"c-Evalue\", \"i-Evalue\", \"score2\", \"bias2\", \"hmm_from\", \"hmm_to\", \"ali_from\", \"ali_to\",\n",
    "                  \"env_from\", \"env_to\", \"acc\", \"description\"]\n",
    "    # Calculate the coverage for each hit\n",
    "    df[\"coverage\"] = (df[\"ali_to\"] - df[\"ali_from\"] + 1) / df[\"tlen\"]\n",
    "\n",
    "    # Filter the DataFrame by E-value and coverage, and sort by bit-score\n",
    "    significant_hits_Ref = df[(df[\"E-value\"] <= 1e-25) & (df[\"coverage\"] > 0.30)]\n",
    "\n",
    "    # Extract Target nmae and store it in the dictionary\n",
    "    if not significant_hits_Ref.empty:\n",
    "        MAGs_Hits_name = significant_hits_Ref[\"target_name\"].tolist()\n",
    "        MAGs_Hits[hmmsearch] = MAGs_Hits_name\n",
    "    else:\n",
    "        MAGs_Hits_name = None\n",
    "        MAGs_Hits[hmmsearch] = MAGs_Hits_name\n",
    "    \n",
    "    # add hits table to a df\n",
    "    All_aed_Hits_df = pd.concat([significant_hits_Ref, All_aed_Hits_df], axis=0)\n",
    "\n",
    "All_aed_Hits_df.to_csv('../data/processed/All_aed_Hits_df_reference.csv')    \n",
    "\n",
    "# done\n",
    "print('done')\n",
    "All_aed_Hits_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "510e7e76-9702-400e-b306-e86864c7ed1e",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 2. 獲取 positive的 target name 並連結 MAGs 並整理出一個類似heatmap的table，再以>= 10 hits數量進行篩選，並有6個必須基因的限制(aedA, aedB, aedG, aedH, aedK, aedF, aedJ)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5810382f-2338-496e-ad69-4b2fcf1b0350",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Any Null:  False\n",
      "done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>query_name</th>\n",
       "      <th>genome_id</th>\n",
       "      <th>num_hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3300027951_10</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3300017448_15</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3300009540_25</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3300018412_3</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3300006913_5</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>3300027708_8</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>3300027965_24</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>3300025519_2</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>3300026167_17</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>3300025216_2</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "query_name      genome_id  num_hits\n",
       "0           3300027951_10        15\n",
       "1           3300017448_15        14\n",
       "2           3300009540_25        14\n",
       "3            3300018412_3        14\n",
       "4            3300006913_5        14\n",
       "..                    ...       ...\n",
       "71           3300027708_8        11\n",
       "72          3300027965_24        11\n",
       "73           3300025519_2        11\n",
       "74          3300026167_17        11\n",
       "75           3300025216_2        11\n",
       "\n",
       "[76 rows x 2 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 須先進行前一個 cell 程式\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "\n",
    "# read table\n",
    "All_aed_Hits_df = pd.read_csv('../data/processed/All_aed_Hits_df_reference.csv') \n",
    "\n",
    "# Got target name and query name    \n",
    "All_aed_Hits_TargetAndQuery = All_aed_Hits_df[['target_name', 'query_name']]\n",
    "\n",
    "# Load a Dataframe with the lookup values for merge protein id to MAGs and merge them\n",
    "TarToMAGs_aed = pd.read_csv('../data/interim/Actino_aed/aed_All_TarToMAGsID.csv')\n",
    "\n",
    "# use merge() function to join the MAGsID data\n",
    "aed_hits_TargetAndMAGsID = pd.merge(All_aed_Hits_TargetAndQuery, TarToMAGs_aed, on='target_name', how='left')\n",
    "\n",
    "# check the null value\n",
    "print('Any Null: ', aed_hits_TargetAndMAGsID['MAGsID'].isnull().any())\n",
    "\n",
    "# got null row\n",
    "null = aed_hits_TargetAndMAGsID[aed_hits_TargetAndMAGsID['MAGsID'].isnull()]\n",
    "\n",
    "# create the crosstab table (like heatmap)\n",
    "aed_hits_heatmap = pd.crosstab(aed_hits_TargetAndMAGsID['query_name'], aed_hits_TargetAndMAGsID['MAGsID'], dropna=False)\n",
    "aed_hits_heatmap = aed_hits_heatmap.transpose()\n",
    "# aed_hits_heatmap\n",
    "# aed_hits_heatmap.to_csv('../data/processed/aed_hits_heatmap_bitscore.csv')\n",
    "\n",
    "# Count the non-zero values in hmm profiles hit row to calculate the number of different HMM profiles that have hits in a given MAG. \n",
    "def count_nonzero(row):\n",
    "    return len(row[row != 0])\n",
    "\n",
    "num_hits = aed_hits_heatmap.apply(count_nonzero, axis=1)\n",
    "aed_hits_heatmap['num_hits'] = num_hits\n",
    "\n",
    "# sort them by hits numer\n",
    "aed_hits_heatmap = aed_hits_heatmap.sort_values(by=\"num_hits\", ascending=False)\n",
    "\n",
    "# extract > 10 hmm profiles hits and the necessary hits (aedA、aedB、aedH、aedG、aedJ、aedF、aedK)\n",
    "aed_hits_FinalFilter =  aed_hits_heatmap[(aed_hits_heatmap['num_hits'] >= 10) & (aed_hits_heatmap['Q_A_I_aedA_RS26385'] != 0) & \n",
    "                                         (aed_hits_heatmap['Q_A_I_aedB_RS26395'] != 0) & (aed_hits_heatmap['Q_A_I_aedJ_RS26410'] != 0) &\n",
    "                                        (aed_hits_heatmap['Q_A_I_aedF_RS26380'] != 0) & (aed_hits_heatmap['Q_A_I_aedG_RS26390'] != 0) &\n",
    "                                         (aed_hits_heatmap['Q_A_I_aedH_RS26400'] != 0) & (aed_hits_heatmap['Q_A_I_aedK_RS26415'] != 0)]\n",
    "#\n",
    "\n",
    "# Reset index and move index column to first position\n",
    "aed_hits_FinalFilter.index.name = None\n",
    "aed_hits_FinalFilter = aed_hits_FinalFilter.reset_index()\n",
    "aed_hits_FinalFilter.insert(0, 'index', aed_hits_FinalFilter.pop('index'))\n",
    "\n",
    "# rename MAGsID\n",
    "aed_hits_FinalFilter = aed_hits_FinalFilter.rename(columns={'index': 'genome_id'})\n",
    "\n",
    "# extract the MAGsID and num_hits column\n",
    "aed_Positive_MAGsID = aed_hits_FinalFilter[['genome_id', 'num_hits']]\n",
    "\n",
    "# done\n",
    "print('done')\n",
    "aed_Positive_MAGsID"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b5fd492-0aae-4694-9438-a26e392d7aa7",
   "metadata": {},
   "source": [
    "### 3. 連結Reference_positive MAGs到 metagenomes 總資料中並進行處理成最株能給R繪製 stacket barchart 的資料格式"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f7327ce1-04bd-48c1-8654-163d6185d429",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Porifera' 'Soil' 'Wastewater' 'Freshwater' 'Annelida' 'Cnidaria'\n",
      " 'Non-marine Saline and Alkaline' 'Marine' 'Fungi' 'Built environment'\n",
      " 'Plant' 'Bioreactor' 'Lab enrichment' 'Bioremediation' 'Sediment']\n",
      "done\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>num_hits</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>ecosystem</th>\n",
       "      <th>ecosystem_type</th>\n",
       "      <th>Homologous_cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3300027951_10</td>\n",
       "      <td>15</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Porifera</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3300017448_15</td>\n",
       "      <td>14</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Actinobacteria</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3300009540_25</td>\n",
       "      <td>14</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Engineered</td>\n",
       "      <td>Wastewater</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3300018412_3</td>\n",
       "      <td>14</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Alphaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3300006913_5</td>\n",
       "      <td>14</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Annelida</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>3300027708_8</td>\n",
       "      <td>11</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Alphaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>3300027965_24</td>\n",
       "      <td>11</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Alphaproteobacteria</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>3300025519_2</td>\n",
       "      <td>11</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Engineered</td>\n",
       "      <td>Lab enrichment</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>3300026167_17</td>\n",
       "      <td>11</td>\n",
       "      <td>p__Actinobacteriota</td>\n",
       "      <td>c__Acidimicrobiia</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>3300025216_2</td>\n",
       "      <td>11</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Actino_aed_cluster</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        genome_id  num_hits               Phylum                   Class  \\\n",
       "0   3300027951_10        15  p__Actinobacteriota       c__Acidimicrobiia   \n",
       "1   3300017448_15        14  p__Actinobacteriota       c__Actinobacteria   \n",
       "2   3300009540_25        14  p__Actinobacteriota       c__Acidimicrobiia   \n",
       "3    3300018412_3        14    p__Proteobacteria  c__Alphaproteobacteria   \n",
       "4    3300006913_5        14    p__Proteobacteria  c__Gammaproteobacteria   \n",
       "..            ...       ...                  ...                     ...   \n",
       "71   3300027708_8        11    p__Proteobacteria  c__Alphaproteobacteria   \n",
       "72  3300027965_24        11    p__Proteobacteria  c__Alphaproteobacteria   \n",
       "73   3300025519_2        11    p__Proteobacteria  c__Gammaproteobacteria   \n",
       "74  3300026167_17        11  p__Actinobacteriota       c__Acidimicrobiia   \n",
       "75   3300025216_2        11    p__Proteobacteria  c__Gammaproteobacteria   \n",
       "\n",
       "          ecosystem  ecosystem_type  Homologous_cluster  \n",
       "0   Host-associated        Porifera  Actino_aed_cluster  \n",
       "1       Terrestrial            Soil  Actino_aed_cluster  \n",
       "2        Engineered      Wastewater  Actino_aed_cluster  \n",
       "3           Aquatic      Freshwater  Actino_aed_cluster  \n",
       "4   Host-associated        Annelida  Actino_aed_cluster  \n",
       "..              ...             ...                 ...  \n",
       "71          Aquatic      Freshwater  Actino_aed_cluster  \n",
       "72      Terrestrial            Soil  Actino_aed_cluster  \n",
       "73       Engineered  Lab enrichment  Actino_aed_cluster  \n",
       "74      Terrestrial            Soil  Actino_aed_cluster  \n",
       "75          Aquatic          Marine  Actino_aed_cluster  \n",
       "\n",
       "[76 rows x 7 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 需先執行上一個cell\n",
    "# open metagenome csv files\n",
    "metagenmoes_df = pd.read_csv('../data/external/Paper_genome_metadata.csv')\n",
    "metagenmoes_df = metagenmoes_df.rename(columns={'ecosystem': 'taxonomy'})\n",
    "\n",
    "# merge positive MAGs with metagenome\n",
    "aed_Positive_metagenomes_df = pd.merge(aed_Positive_MAGsID, metagenmoes_df, on='genome_id', how='left')\n",
    "\n",
    "# print(aed_Positive_metagenomes_df['ecosystem_category'].unique())\n",
    "# print(aed_Positive_metagenomes_df['ecosystem_type'].unique())\n",
    "# display(aed_Positive_metagenomes_df[aed_Positive_metagenomes_df['ecosystem_type'] == 'Anaerobic'])\n",
    "\n",
    "# extract certain column\n",
    "aed_Positive_metagenomes_df = aed_Positive_metagenomes_df[['genome_id', 'taxonomy', 'ecosystem_type', 'num_hits']]\n",
    "\n",
    "# extract phylum and class from taxonomy column and expand to new column\n",
    "aed_Positive_metagenomes_df['Phylum'] = aed_Positive_metagenomes_df['taxonomy'].str.extract('(p__\\w+)', expand=True)\n",
    "aed_Positive_metagenomes_df['Class'] = aed_Positive_metagenomes_df['taxonomy'].str.extract('(c__\\w+)', expand=True)\n",
    "\n",
    "# # print non-duplicated values in ecosystem_type than check the lable\n",
    "# print(aed_Positive_metagenomes_df['ecosystem_type'].unique())\n",
    "# aed_Positive_metagenomes_df\n",
    "\n",
    "# Load a EcosystemToEco_type dataframe for lookup\n",
    "ForEcoLookup = pd.read_csv(\"../data/interim/EcosystemToEco_type.csv\")\n",
    "\n",
    "# merge them with ecosystem_type\n",
    "aed_addEco_df = pd.merge(aed_Positive_metagenomes_df, ForEcoLookup, on='ecosystem_type', how='left')\n",
    "\n",
    "# check the None value in ecosystem column Create a Boolean mask to identify NaN values\n",
    "mask = aed_addEco_df.isna()\n",
    "aed_addEco_df_nan_rows = aed_addEco_df[mask.any(axis=1)]\n",
    "aed_addEco_df_nan_rows\n",
    "\n",
    "# !!!! change the ecosystem_type value name\n",
    "aed_addEco_df['ecosystem_type'] = aed_addEco_df['ecosystem_type'].replace({'Activated Sludge': 'Wastewater', 'City': 'Built environment', 'Rhizoplane': 'Plant', 'Anaerobic': 'Bioreactor',\n",
    "                                                                           'Defined media': 'Lab enrichment', 'Tetrachloroethylene and derivatives': 'Bioremediation', 'Mycelium': 'Fungi'})\n",
    "print(aed_addEco_df['ecosystem_type'].unique())\n",
    "\n",
    "# remove, add and rearrange column\n",
    "aed_addEco_df.drop('taxonomy', axis=1, inplace=True)\n",
    "aed_addEco_df= aed_addEco_df.reindex(columns=['genome_id', 'num_hits', 'Phylum', 'Class', 'ecosystem', 'ecosystem_type'])\n",
    "aed_addEco_df['Homologous_cluster']='Actino_aed_cluster'\n",
    "\n",
    "# # check phylum data\n",
    "# print(aed_addEco_df['Phylum'].unique())\n",
    "\n",
    "# write file\n",
    "aed_addEco_df.to_csv('../data/processed/Final/Actino/aed_PositiveHits_ForR_reference.csv')\n",
    "print('done')\n",
    "aed_addEco_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85877e1f-ee19-45cc-b455-06b7c2a4d4fc",
   "metadata": {},
   "source": [
    "# 補充資料"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "520571a7-3e46-45e0-beb3-b2c060a48785",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Q_A_I_aedA_RS26385', 'Q_A_I_aedB_RS26395', 'Q_A_I_aedF_RS26380',\n",
      "       'Q_A_I_aedG_RS26390', 'Q_A_I_aedH_RS26400', 'Q_A_I_aedJ_RS26410',\n",
      "       'Q_A_I_aedK_RS26415', 'Q_A_aedD_RS26370', 'Q_A_aedE_RS26375',\n",
      "       'Q_A_aedI_RS26405', 'Q_A_aedL_RS26420', 'Q_A_aedM_RS26425',\n",
      "       'Q_A_aedN_RS26430', 'Q_A_aedO_RS26435', 'Q_A_aedP_RS26440', 'num_hits'],\n",
      "      dtype='object', name='query_name')\n"
     ]
    }
   ],
   "source": [
    "# print(aed_hits_heatmap.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fab9fb8c-a2e5-4595-9db8-a5d8882668f1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
